{"cells":[{"cell_type":"markdown","metadata":{"id":"DQBomu0-6mGq"},"source":["# Building LLM-based applications for clinician support\n","\n","In this notebook, we will build an application to provide patient treatment recommendations for clinicians with the help of Large Language Models (LLMs).\n","\n","To overcome LLMs limitations of hallucinations and lack of specific domain knowledge, we will allow the LLM to learn domain knowledge from context (In-Context Learning) by feeding into the prompts relevant texts in published papers about recommendations for disease treatment and maintenance.\n","\n","There are two parts in this tutorial:\n","- Build a simple QARetrievalFromSource Chain, which takes in a user question, retrieve relevant documents from a vectorstore and ground the LLM to answer the question based on given context.\n","\n","- Build an autonomous agent which perform multiple steps (search, calculation, query from database, etc) to provide a recommendation based on a patient's calculated eGFR score. The agent follows ReAct reasoning framework (Think-Act-Observe) and is given a diverse toolkit (MRKL) to choose from."]},{"cell_type":"markdown","metadata":{"id":"TIq5oJSaGnUV"},"source":["# First let's install the packages and dependencies required.\n","\n","- langchain: Main package for building LLM application\n","- openai: For use OpenAI LLM model as engine for application\n","- faiss-cpu: Facebook AI Similarity Search. Vector Database for document retrieval.\n","- tiktoken: Package for counting tokens consumed by LLMs. Support several LLMs including OpenAI and HuggingFace models.\n","- pypdf & PyMuPDF: Package for loading PDF documents."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":11890,"status":"ok","timestamp":1691398772363,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"UHzKF3Rsjjag","outputId":"b06c90ae-0546-4da6-e86a-2ceb20871ed4"},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.4/1.4 MB\u001b[0m \u001b[31m7.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m73.6/73.6 kB\u001b[0m \u001b[31m8.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m17.6/17.6 MB\u001b[0m \u001b[31m57.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.7/1.7 MB\u001b[0m \u001b[31m64.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m270.3/270.3 kB\u001b[0m \u001b[31m26.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.1/14.1 MB\u001b[0m \u001b[31m85.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m90.0/90.0 kB\u001b[0m \u001b[31m9.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m49.4/49.4 kB\u001b[0m \u001b[31m5.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25h"]}],"source":["%pip install --quiet langchain openai faiss-cpu tiktoken pypdf PyMuPDF"]},{"cell_type":"markdown","metadata":{"id":"fBQ9tqmyHh1p"},"source":["We will create a working folder iaim_langchain and set this as the MAIN working directory"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":378,"status":"ok","timestamp":1691398783370,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"GgZl8yRSVsBl","outputId":"dd0d3d05-5e9f-4cf4-afe1-42b50fedaa50"},"outputs":[{"name":"stdout","output_type":"stream","text":["/content/iaim_langchain\n"]}],"source":["%mkdir iaim_langchain\n","%cd /content/iaim_langchain\n","\n","import os\n","MAIN_DIR = os.getcwd()"]},{"cell_type":"markdown","metadata":{"id":"hzqI5JCsHqkf"},"source":["# Let's download and extract the data folder which contains the following:\n","1. PDF files to be used for creating vectorstore.\n","\n","2. Dummy Patient Database file (patient_data.csv).\n","\n","3. Authentification json file: Containing OpenAI API keys and SerpAPI API Key"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":17775,"status":"ok","timestamp":1691398817443,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"gqZFKICpWUWu","outputId":"11d527f3-73b1-4066-cb80-566f2b553076"},"outputs":[{"name":"stdout","output_type":"stream","text":["Requirement already satisfied: gdown in /usr/local/lib/python3.10/dist-packages (4.6.6)\n","Collecting gdown\n","  Downloading gdown-4.7.1-py3-none-any.whl (15 kB)\n","Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from gdown) (3.12.2)\n","Requirement already satisfied: requests[socks] in /usr/local/lib/python3.10/dist-packages (from gdown) (2.27.1)\n","Requirement already satisfied: six in /usr/local/lib/python3.10/dist-packages (from gdown) (1.16.0)\n","Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from gdown) (4.65.0)\n","Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.10/dist-packages (from gdown) (4.11.2)\n","Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.10/dist-packages (from beautifulsoup4->gdown) (2.4.1)\n","Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests[socks]->gdown) (1.26.16)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests[socks]->gdown) (2023.7.22)\n","Requirement already satisfied: charset-normalizer~=2.0.0 in /usr/local/lib/python3.10/dist-packages (from requests[socks]->gdown) (2.0.12)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests[socks]->gdown) (3.4)\n","Requirement already satisfied: PySocks!=1.5.7,>=1.5.6 in /usr/local/lib/python3.10/dist-packages (from requests[socks]->gdown) (1.7.1)\n","Installing collected packages: gdown\n","  Attempting uninstall: gdown\n","    Found existing installation: gdown 4.6.6\n","    Uninstalling gdown-4.6.6:\n","      Successfully uninstalled gdown-4.6.6\n","Successfully installed gdown-4.7.1\n","Downloading...\n","From: https://drive.google.com/uc?id=10TjeOFh7Doq8IvynqcYlCC2LwANUhDL8\n","To: /content/iaim_langchain/iaim_data_files.zip\n","100% 16.7M/16.7M [00:00<00:00, 55.7MB/s]\n","Archive:  iaim_data_files.zip\n","  inflating: api_keys.json           \n","   creating: database/\n","  inflating: database/patients_info.csv  \n","   creating: raw_data/\n","  inflating: raw_data/ACG 2015 Hereditary GI Cancers.pdf  \n","  inflating: raw_data/AGA 2021 IBD CRC Surveillance.pdf  \n","  inflating: raw_data/ASCRS 2017 Lynch Syndrome.pdf  \n","  inflating: raw_data/Chronic kidney disease in the elderly.pdf  \n","  inflating: raw_data/ckd_evaluation_classification_stratification.pdf  \n","  inflating: raw_data/MSTF 2017 CRC Screening.pdf  \n","  inflating: raw_data/nihms-1069237.pdf  \n","  inflating: raw_data/nihms639366.pdf  \n","  inflating: raw_data/umj7603-154.pdf  \n","  inflating: raw_data/USPSTF 2021 CRC Screening.pdf  \n"]}],"source":["!pip3 install --upgrade gdown\n","!gdown https://drive.google.com/uc?id=10TjeOFh7Doq8IvynqcYlCC2LwANUhDL8\n","!unzip iaim_data_files.zip"]},{"cell_type":"markdown","metadata":{"id":"OAQxtxUEILna"},"source":["# Import necessary packages"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"hd85yXDKlORm"},"outputs":[],"source":["import os, json, logging, re\n","from getpass import getpass\n","from pydantic import root_validator\n","from typing import Any, Union, Tuple, Dict, Callable, List, Optional, Literal\n","from pprint import pprint\n","import math, numexpr\n","from random import randint\n","\n","from langchain import OpenAI\n","from langchain.docstore.document import Document\n","from langchain.chains.question_answering import load_qa_chain\n","from langchain.document_loaders import PyMuPDFLoader\n","from langchain.embeddings.openai import OpenAIEmbeddings\n","from langchain.vectorstores import VectorStore, FAISS\n","from langchain.text_splitter import RecursiveCharacterTextSplitter\n","from langchain.chat_models import ChatOpenAI\n","from langchain.prompts import PromptTemplate, StringPromptTemplate\n","from langchain.chains import RetrievalQAWithSourcesChain, LLMMathChain\n","from langchain.chains.combine_documents.stuff import StuffDocumentsChain\n","from langchain.chains.llm import LLMChain\n","from langchain.agents import Tool, AgentType, AgentExecutor, LLMSingleActionAgent, AgentOutputParser, initialize_agent, load_tools, create_csv_agent\n","from langchain.tools import BaseTool\n","from langchain.chains.qa_with_sources.base import BaseQAWithSourcesChain\n","from langchain.base_language import BaseLanguageModel\n","from langchain.utilities import GoogleSerperAPIWrapper\n","from langchain.prompts.chat import ChatPromptTemplate, SystemMessagePromptTemplate, HumanMessagePromptTemplate\n","from langchain.tools.python.tool import PythonAstREPLTool # Need Python 3.9 and above\n","from langchain.agents.mrkl.base import ZeroShotAgent"]},{"cell_type":"markdown","metadata":{"id":"rfkggGbEKpTE"},"source":["- To use OpenAI models, we need a working API-Key. You can register for your own key at https://openai.com/blog/openai-api. <br> On your first registration, you will be given some free credits to try out the API. <br> Afterwards, OpenAI charges users based on tokens consumed. <br> Note that for free version, you can only access gpt-3.5-turbo model. For access to the largest model (gpt-4), you need to pay at least once.\n","\n","- In addition, we will also need SerpAPI key to use it's API for Internet Search. You can register a free API Key here (https://serper.dev/).\n","\n","- In this tutorial, we have provided you with the OPENAI API and SERP API keys so you don't need to worry about it. <br>\n","\n","Note: It is recommended that you register API Keys inside your environment variables instead of hard-coding it inside your codes. This is to prevent others accidentally have access to your paid API keys"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jeySlC0J0pl0"},"outputs":[],"source":["with open(\"api_keys.json\", \"r\") as f:\n","    keys = json.load(f)\n","    os.environ[\"OPENAI_API_KEY\"] = keys[\"OPENAI_API_KEY\"]\n","    os.environ[\"SERPER_API_KEY\"] = keys[\"SERP_API\"]\n","\n","# Alternatively, you can manually assign your API keys to environment variables\n","# os.environ[\"OPENAI_API_KEY\"] = \"Your OpenAI_API_Key\"\n","# os.environ[\"SERPER_API_KEY\"] = \"Your SERPER_API_KEY\""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"JvzZ1-Tnk-c9"},"outputs":[],"source":["def check_documents_token(docs: List[Document], llm = ChatOpenAI(temperature=0, model_name=\"gpt-3.5-turbo\")):\n","    if not isinstance(docs, List):\n","        docs = [docs]\n","    combine_document_chain = StuffDocumentsChain(\n","        llm_chain=LLMChain(llm=llm, prompt=PromptTemplate(template=\"{summaries}\", input_variables=[\"summaries\"]), verbose=False),\n","        verbose=False)\n","    return combine_document_chain.prompt_length(docs)"]},{"cell_type":"markdown","metadata":{"id":"inwpKUd8mdPX"},"source":["# Langchain\n","\n","- Langchain is a Python/JS package for building LLM applications. It provides a modular framework to make interacting with LLM easier, including prompt engineering, processing inputs/outputs for chaining multiple LLM calls together.\n","\n","- Langchain has integrations with many LLM models including OpenAI, Google and Hugging Face models. It also has extensive set of tools, API connectors as well as a variety of vector database (Pinecone, Chroma, FAISS, Redis, Elastic Search, etc)\n","\n","- More information can be found at https://python.langchain.com/docs/get_started/introduction.html"]},{"cell_type":"markdown","metadata":{"id":"ZxrV64_hM-aR"},"source":["# Creating the first Prompt\n","\n","For prompt construction, Langchain used PromptTemplate module to allow reproducability with several LLM calls.\n","\n","The text inside curly brackets {} in a string prompt acts as placeholder variables and will allow user to enter a different query each time to the prompt template while keeping the other parts same.\n","\n","Let's setup a simple custom PromptTemplate to answer general questions. Here we add the sentence \"Do not make up an answer if you dont know\" to reduce the chance that LLM will hallucinate and make up answers."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"YXirZJgcXllv"},"outputs":[],"source":["prompt_template = \"\"\"Answer the question below. Do not make up an answer if you dont know.\n","Question: {question}\n","\"\"\"\n","basic_prompt = PromptTemplate(template=prompt_template, input_variables=[\"question\"])\n","basic_chain = LLMChain(llm=ChatOpenAI(temperature=0, model_name = \"gpt-3.5-turbo\"), prompt=basic_prompt, verbose = True)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2895,"status":"ok","timestamp":1691399179085,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"fjYa4251ZJaY","outputId":"5be9e3bf-2fbd-486a-99de-a4b47b683ecc"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new LLMChain chain...\u001b[0m\n","Prompt after formatting:\n","\u001b[32;1m\u001b[1;3mAnswer the question below. Do not make up an answer if you dont know.\n","Question: Who is the prime minister of Singapore in 2021?\n","\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'question': 'Who is the prime minister of Singapore in 2021?',\n"," 'text': 'As of my last update in September 2021, the Prime Minister of Singapore is Lee Hsien Loong. However, please note that political positions can change, and it is always a good idea to verify the most up-to-date information.'}"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["basic_chain(\"Who is the prime minister of Singapore in 2021?\")"]},{"cell_type":"markdown","metadata":{"id":"gkbaA7WxTAY5"},"source":["As you can see, the text inside curly brackets {question} has been replaced with our query \"Who is the prime minister of Singapore in 2021?\". The LLM managed to answer the general question correctly."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3048,"status":"ok","timestamp":1691399208133,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"N95vmCjp5qVr","outputId":"6311bee4-25a2-466b-f870-59de6d6734dd"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new LLMChain chain...\u001b[0m\n","Prompt after formatting:\n","\u001b[32;1m\u001b[1;3mAnswer the question below. Do not make up an answer if you dont know.\n","Question: Who is the current minister of Singapore in 2023?\n","\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'question': 'Who is the current minister of Singapore in 2023?',\n"," 'text': \"I'm sorry, but as an AI language model, I don't have access to real-time information or the ability to browse the internet. Therefore, I cannot provide you with the current minister of Singapore in 2023. It's best to refer to official government sources or news outlets for the most up-to-date information.\"}"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["basic_chain(\"Who is the current minister of Singapore in 2023?\")"]},{"cell_type":"markdown","metadata":{"id":"erdnmIzPTTDs"},"source":["GPT models are trained on data up until 2021. Hence while it can answer question related to facts in year 2021, it cannot answer question after that. This is a limitation of LLM, as it's knowledge is bounded by the data it's trained on. <br>\n","\n","Let's try again. Here we will ask the LLM to answer question about an older Deep Learning model (BERT - 2017) and a recent model (LLAMA-2 - 2023)."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4863,"status":"ok","timestamp":1691399242843,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"2sfF-nOoZr8f","outputId":"27f6df87-e2bf-4aaf-cef9-5b4fe263fb80"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new LLMChain chain...\u001b[0m\n","Prompt after formatting:\n","\u001b[32;1m\u001b[1;3mAnswer the question below. Do not make up an answer if you dont know.\n","Question: Can you give a brief summary of BERT?\n","\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'question': 'Can you give a brief summary of BERT?',\n"," 'text': 'BERT (Bidirectional Encoder Representations from Transformers) is a pre-trained natural language processing model developed by Google. It is designed to understand the context and meaning of words in a sentence by considering the words that come before and after them. BERT uses a transformer architecture, which allows it to capture long-range dependencies in text. It has been trained on a large amount of unlabeled text data from the internet, enabling it to learn general language representations. BERT has achieved state-of-the-art performance on various natural language processing tasks, such as question answering, sentiment analysis, and text classification.'}"]},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["basic_chain(\"Can you give a brief summary of BERT?\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2945,"status":"ok","timestamp":1691399274069,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"pFl2u9Qn6Q36","outputId":"165097ea-d404-4859-fa4c-7e5923fcf89b"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new LLMChain chain...\u001b[0m\n","Prompt after formatting:\n","\u001b[32;1m\u001b[1;3mAnswer the question below. Do not make up an answer if you dont know.\n","Question: Can you give a brief summary of LLAMA-2?\n","\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'question': 'Can you give a brief summary of LLAMA-2?',\n"," 'text': \"I'm sorry, but as an AI language model, my responses are generated based on a mixture of licensed data, data created by human trainers, and publicly available data. I have not been directly trained on specific proprietary databases or have access to classified information. Therefore, I don't have information on LLAMA-2.\"}"]},"execution_count":11,"metadata":{},"output_type":"execute_result"}],"source":["basic_chain(\"Can you give a brief summary of LLAMA-2?\")"]},{"cell_type":"markdown","metadata":{"id":"w4Um1ogRmjN3"},"source":["# Question Answering (QA) over local documents\n","\n","To overcome the knowledge bound limitations of LLMs, we can:\n","1. Fine-Tune the LLM with latest data\n","2. In-Context Learning: Provide the LLM with the updated information inside the prompt.\n","\n","Fine-Tuning is effective but is also very costly in terms of computation and data preparation. In addition, it doesn't solve the problem of un-updated knowledge base, as models fine-tuned today may be outdated a year later and need to be fine-tuned again.\n","\n","In-Context Learning hence is an attractive solution to allows LLMs to have access to latest data at the point of running."]},{"cell_type":"markdown","metadata":{"id":"7lS5UapMWHmS"},"source":["Next we will build a simple Chain which allows LLM to interact will a document.\n","\n","Before that, let's ask the previous LLM chain a specific medical question."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2880,"status":"ok","timestamp":1691399311363,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"wg-fPZng1bmz","outputId":"2bc49efe-38fe-4723-fef4-1b44daa8c462"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new LLMChain chain...\u001b[0m\n","Prompt after formatting:\n","\u001b[32;1m\u001b[1;3mAnswer the question below. Do not make up an answer if you dont know.\n","Question: What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\n","\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'question': 'What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?',\n"," 'text': \"I'm sorry, but as an AI language model, I don't have access to personal medical records or the ability to provide specific medical advice. It is important to consult with a healthcare professional or a gastroenterologist who can evaluate your specific situation and provide appropriate recommendations for colonoscopy screening based on your medical history and family history.\"}"]},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["basic_chain(\"What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\")"]},{"cell_type":"markdown","metadata":{"id":"gmWbq8RkUuuT"},"source":["LLM inherent knowledge tends to be more generic and hence it is not able of answering very specific question which requires strong domain knowledge. <br> We will provide it with some help from a research paper: **\"Colorectal Cancer Screening: Recommendations for Physicians and Patients From the U.S Multi-Society Task Force on Colorectal Cancer\"** - Douglas et al, 2017\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":326,"status":"ok","timestamp":1691399354341,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"Jv9WEpfA0rUN","outputId":"c14f4c69-2931-466a-d9cf-b81273f22e8c"},"outputs":[{"name":"stdout","output_type":"stream","text":["Number of pages: 17\n","eﬁned\n","inherited syndrome caused by mutations in 1 or more\n","mismatch repair genes. Patients in families that meet the\n","clinical criteria for hereditary nonpolyposis CRC but have\n","microsatellite-stable CRCs have family colon cancer syn-\n","drome X, which has not been genetically deﬁned.122 Persons\n","in families with syndrome X should undergo colonoscopy at\n","least every 3 to 5 years, beginning 10 years before the age at\n","diagnosis of the youngest affected relative.\n","A family history of CRC in a ﬁrst-degree relative in-\n","creases the risk of CRC regardless of the age at diagnosis of\n","Table 4.Multi-Society Task Force Ranking of Current\n","Colorectal Cancer Screening Tests\n","Tier 1\n","Colonoscopy every 10 years\n","Annual fecal immunochemical test\n","Tier 2\n","CT colonography every 5 years\n","FIT–fecal DNA every 3 years\n","Flexible sigmoidoscopy every 10 years (or every 5 years)\n","Tier 3\n","Capsule colonoscopy every 5 years\n","Available tests not currently recommended\n","Septin 9\n","- 2017\n","MSTF Recommendations for CRC Screening\n","9\n","AGA SECTION\n","\n"]}],"source":["sample_document = os.path.join(\".\", \"raw_data\", \"MSTF 2017 CRC Screening.pdf\")\n","pages = PyMuPDFLoader(sample_document).load()\n","print(\"Number of pages:\", len(pages))\n","sample_page = pages[8]\n","print(sample_page.page_content[-1000:])"]},{"cell_type":"markdown","metadata":{"id":"ftQSgeiMXRSd"},"source":["Let's initiate a QA chain to load the document in and answer our question.\n","\n","Here we will use chain_type = \"stuff\", meaning that we simply \"stuff\" all the information inside all documents inside our placeholder **{context}**.\n","\n","Let's investigate the prompt template."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":364,"status":"ok","timestamp":1691399403763,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"TNiB1_7h94C7","outputId":"7774818b-67fa-4017-85f2-6514684048cb"},"outputs":[{"name":"stdout","output_type":"stream","text":["Use the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.\n","\n","{context}\n","\n","Question: {question}\n","Helpful Answer:\n"]}],"source":["qa_stuff_chain = load_qa_chain(llm=OpenAI(temperature=0, max_tokens=512), chain_type=\"stuff\", verbose=True)\n","print(qa_stuff_chain.llm_chain.prompt.template)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1277,"status":"ok","timestamp":1691399447071,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"ftear0at_J15","outputId":"e940acfb-6d81-4ebd-ae68-20c2918cc890"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n","\n","\n","\u001b[1m> Entering new LLMChain chain...\u001b[0m\n","Prompt after formatting:\n","\u001b[32;1m\u001b[1;3mUse the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.\n","\n","opportunistic setting is ensuring repeated annual perfor-\n","mance. Using a tier 1 approach that focuses on 2 tests makes\n","the discussion of CRC screening tests between physician and\n","patient manageable and feasible, and, as noted above,\n","expanding the number of options in the initial discussion\n","beyond 2 did not increase screening rates.19 Colonoscopy\n","and FIT can be adapted readily to either the sequential offer\n","of screening (colonoscopy is offered ﬁrst with FIT reserved\n","for those who decline colonoscopy), the multiple-options\n","approach (colonoscopy and FIT are each discussed with\n","patients, and if both are declined the discussion moves\n","sequentially to tier 2 tests), and the risk-stratiﬁed approach\n","(eg, colonoscopy is offered ﬁrst to men age >60 and women\n","age >65 with no prior screening, and FIT is offered to\n","persons under these ages and persons with negative prior\n","colonoscopy). Risk stratiﬁcation may also take into account\n","factors such as cigarette smoking, diabetes, and obesity.\n","The rationale for placing tests in the tier 2 and 3 cate-\n","gories follows from the discussion above of individual tests.\n","The ranking implies equivalence for tests within each\n","category, but this is not the intent of the MSTF. For example,\n","ﬂexible sigmoidoscopy screening has strong evidence to\n","support its use. However, the steady decline in the use of\n","ﬂexible sigmoidoscopy in the United States9,10 suggests that\n","any strong endorsement of ﬂexible sigmoidoscopy is not\n","consistent with the reality of the test’s lack of popularity\n","among patients and poor reimbursement for physicians. The\n","strong evidence base supporting ﬂexible sigmoidoscopy\n","leads us to place the test in the tier 2 category, although we\n","expect that most practitioners using a sequential approach\n","would move to CT colonography or FIT–fecal DNA in pa-\n","tients who decline colonoscopy and FIT, because these tests\n","are less invasive and survey the entire colon.\n","Performance characteristics alone would place capsule\n","colonoscopy in tier 2. However, offering capsule colonos-\n","copy as a tier 2 test in a sequential methodology currently\n","would often lead to frustration because reimbursement for\n","screening capsule colonoscopy is seldom available at this\n","time, and the test itself is frequently not available. The\n","onerous bowel preparation and the lack of systems to\n","accomplish same-day colonoscopy in most patients with a\n","positive capsule colonoscopy are additional factors placing\n","capsule colonoscopy in the tier 3 category at this time.\n","In summary, we suggest that the tiered system (Table 4)\n","has numerous advantages. Predominant reliance on tier 1\n","tests offers modalities with optimal effectiveness, cost-\n","effectiveness, complete colon screening, proven popularity\n","with patients, and a simpliﬁed discussion (compared with\n","offering 5-7 different tests) and still leaves room to offer\n","other tests in a sequential fashion. If patients decline colo-\n","noscopy and FIT, all tests in tier 2 are acceptable CRC\n","screening tests, but each has deﬁciencies relative to the\n","tier 1 tests.\n","Recommendations\n","1. We recommend colonoscopy every 10 years or annual\n","FIT as ﬁrst-tier options for screening the average-risk\n","persons for colorectal neoplasia (strong recommenda-\n","tion; moderate-quality evidence).\n","2. We recommend that physicians performing screening\n","colonoscopy measure quality, including the adenoma\n","detection rate (strong recommendation, high-quality\n","evidence).\n","3. We recommend that physicians performing FIT monitor\n","quality (strong recommendation, low-quality evidence).\n","The recommended quality measurements for FIT pro-\n","grams are detailed in a prior publication.86\n","4. We recommend CT colonography every 5 years or FIT–\n","fecal DNA every 3 years (strong recommendation, low-\n","quality evidence) or ﬂexible sigmoidoscopy every 5 to\n","10\n","years\n","(strong\n","recommendation,\n","high-quality\n","evidence) in patients who refuse colonoscopy and FIT.\n","5. We suggest that capsule colonoscopy (if available) is an\n","appropriate screening test when patients decline colo-\n","noscopy, FIT, FIT–fecal DNA, CT colonography, and ﬂex-\n","ible sigmoidoscopy (weak recommendation, low-quality\n","evidence).\n","6. We suggest against Septin9 for CRC screening (weak\n","recommendation, low-quality evidence).\n","Family History of CRC and Polyps\n","We recommend that screening in most average-risk\n","persons be initiated at age 50 years. A family history of\n","CRC or certain polyps can modify the recommended starting\n","age and the frequency of screening. The MSTF has previ-\n","ously issued recommendations for screening in persons\n","with Lynch syndrome,34 which is a genetically deﬁned\n","inherited syndrome caused by mutations in 1 or more\n","mismatch repair genes. Patients in families that meet the\n","clinical criteria for hereditary nonpolyposis CRC but have\n","microsatellite-stable CRCs have family colon cancer syn-\n","drome X, which has not been genetically deﬁned.122 Persons\n","in families with syndrome X should undergo colonoscopy at\n","least every 3 to 5 years, beginning 10 years before the age at\n","diagnosis of the youngest affected relative.\n","A family history of CRC in a ﬁrst-degree relative in-\n","creases the risk of CRC regardless of the age at diagnosis of\n","Table 4.Multi-Society Task Force Ranking of Current\n","Colorectal Cancer Screening Tests\n","Tier 1\n","Colonoscopy every 10 years\n","Annual fecal immunochemical test\n","Tier 2\n","CT colonography every 5 years\n","FIT–fecal DNA every 3 years\n","Flexible sigmoidoscopy every 10 years (or every 5 years)\n","Tier 3\n","Capsule colonoscopy every 5 years\n","Available tests not currently recommended\n","Septin 9\n","- 2017\n","MSTF Recommendations for CRC Screening\n","9\n","AGA SECTION\n","\n","\n","Question: What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\n","Helpful Answer:\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","Answer:  Colonoscopy every 3 to 5 years, beginning 10 years before the age at diagnosis of the youngest affected relative.\n"]}],"source":["answer = qa_stuff_chain({\"input_documents\": [sample_page],\n","          \"question\": \"What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\"})\n","\n","print(\"Answer:\", answer[\"output_text\"])"]},{"cell_type":"markdown","metadata":{"id":"yb6tGRUdYF-E"},"source":["Great. It seems that the LLM has find the information we want from the piece of context given. <br>"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2091,"status":"ok","timestamp":1691399494121,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"g-lKOcyYA2S3","outputId":"479bcc8a-2770-42ea-92ec-1c40bed7bfeb"},"outputs":[{"name":"stdout","output_type":"stream","text":["Number of characters for Page 10:  5549\n","Number of tokens for Page 10:  1286\n"]}],"source":["print(\"Number of characters for Page 10: \", len(sample_page.page_content))\n","print(\"Number of tokens for Page 10: \", check_documents_token(sample_page))"]},{"cell_type":"markdown","metadata":{"id":"5tyGT0CWYVdv"},"source":["- The tokens used by OpenAI LMs and most current LLMs are sub-word levels. This helps solve Out-Of-Vocabulary problems and reduce the number of model parameters required.\n","- As LLMs have maximum context length which limits the number of tokens which can be fed inside a single call, we usually should keep track of the number of tokens inside our prompts. Also, tokens cost $$$ so it's good to monitor the usage."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000},"executionInfo":{"elapsed":1148,"status":"error","timestamp":1691399521260,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"5eBkZ8lfC7xA","outputId":"616cb3f6-dccd-4a22-eeaf-f451ddfeaca7"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n","\n","\n","\u001b[1m> Entering new LLMChain chain...\u001b[0m\n","Prompt after formatting:\n","\u001b[32;1m\u001b[1;3mUse the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.\n","\n","Colorectal Cancer Screening: Recommendations for Physicians\n","and Patients From the U.S. Multi-Society Task Force on\n","Colorectal Cancer\n","Douglas K. Rex,1 C. Richard Boland,2 Jason A. Dominitz,3 Francis M. Giardiello,4\n","David A. Johnson,5 Tonya Kaltenbach,6 Theodore R. Levin,7 David Lieberman,8 and\n","Douglas J. Robertson9\n","1Indiana University School of Medicine, Indianapolis, Indiana; 2University of California San Diego, San Diego, California;\n","3VA Puget Sound Health Care System, University of Washington, Seattle, Washington; 4Johns Hopkins University School of\n","Medicine, Baltimore, Maryland; 5Eastern Virginia Medical School, Norfolk, Virginia; 6San Francisco Veterans Affairs Medical\n","Center, San Francisco, California; 7Kaiser Permanente Medical Center, Walnut Creek, California; 8Oregon Health and Science\n","University, Portland, Oregon; 9VA Medical Center, White River Junction, Vermont, and Geisel School of Medicine at Dartmouth,\n","Hanover, New Hampshire\n","This\n","document\n","updates\n","the\n","colorectal\n","cancer\n","(CRC)\n","screening recommendations of the U.S. Multi-Society Task\n","Force of Colorectal Cancer (MSTF), which represents the\n","American\n","College\n","of\n","Gastroenterology,\n","the\n","American\n","Gastroenterological Association, and The American Society\n","for Gastrointestinal Endoscopy. CRC screening tests are\n","ranked in 3 tiers based on performance features, costs, and\n","practical considerations. The ﬁrst-tier tests are colonoscopy\n","every 10 years and annual fecal immunochemical test (FIT).\n","Colonoscopy and FIT are recommended as the cornerstones\n","of screening regardless of how screening is offered. Thus, in\n","a sequential approach based on colonoscopy offered ﬁrst,\n","FIT should be offered to patients who decline colonoscopy.\n","Colonoscopy and FIT are recommended as tests of choice\n","when multiple options are presented as alternatives. A risk-\n","stratiﬁed approach is also appropriate, with FIT screening\n","in populations with an estimated low prevalence of\n","advanced neoplasia and colonoscopy screening in high\n","prevalence populations. The second-tier tests include CT\n","colonography every 5 years, the FIT–fecal DNA test every 3\n","years, and ﬂexible sigmoidoscopy every 5 to 10 years. These\n","tests are appropriate screening tests, but each has disad-\n","vantages relative to the tier 1 tests. Because of limited evi-\n","dence and current obstacles to use, capsule colonoscopy\n","every 5 years is a third-tier test. We suggest that the Septin9\n","serum assay (Epigenomics, Seattle, Wash) not be used for\n","screening. Screening should begin at age 50 years in\n","average-risk persons, except in African Americans in whom\n","limited evidence supports screening at 45 years. CRC inci-\n","dence is rising in persons under age 50, and thorough\n","diagnostic evaluation of young persons with suspected\n","colorectal bleeding is recommended. Discontinuation of\n","screening should be considered when persons up to date\n","with screening, who have prior negative screening (partic-\n","ularly colonoscopy), reach age 75 or have <10 years of life\n","expectancy. Persons without prior screening should be\n","considered for screening up to age 85, depending on age and\n","comorbidities. Persons with a family history of CRC or a\n","documented advanced adenoma in a ﬁrst-degree relative\n","age <60 years or 2 ﬁrst-degree relatives with these ﬁndings\n","at any age are recommended to undergo screening by colo-\n","noscopy every 5 years, beginning 10 years before the age at\n","diagnosis of the youngest affected relative or age 40,\n","whichever is earlier. Persons with a single ﬁrst-degree\n","relative diagnosed at ‡60 years with CRC or an advanced\n","adenoma can be offered average-risk screening options\n","beginning at age 40 years.\n","C\n","olorectal cancer (CRC) screening is the process\n","of detecting early-stage CRCs and precancerous\n","lesions in asymptomatic people with no prior history of\n","cancer or precancerous lesions. The U.S. Multi-Society Task\n","Force of Colorectal Cancer (MSTF) is a panel of expert\n","gastroenterologists representing the American College of\n","Gastroenterology, the American Gastroenterological Associ-\n","ation,\n","and\n","the\n","American\n","Society\n","for\n","Gastrointestinal\n","Endoscopy. The MSTF, like others, has long endorsed sys-\n","tematic offers of CRC screening to average-risk persons\n","(persons without a high-risk family history of colorectal\n","neoplasia) beginning at age 50 years, with general evidence\n","supporting screening reviewed in previous publications.1\n","This publication updates the screening recommendations\n","of the MSTF for screening in average-risk persons.1\n","Screening differs from surveillance. Surveillance refers\n","to the interval use of colonoscopy in patients with previ-\n","ously detected CRC or precancerous lesions and interval\n","colonoscopy in patients performed to detect dysplasia in\n","persons with inﬂammatory bowel disease affecting the co-\n","lon. Surveillance recommendations from the MSTF on sur-\n","veillance after cancer2 and removal of precancerous lesions3\n","are available in other documents. Screening is also distinct\n","from\n","diagnostic\n","examinations,\n","which\n","refer\n","to\n","the\n","Abbreviations used in this paper: CRC, colorectal cancer; FIT, fecal\n","immunochemical test; MSTF, U.S. Multi-Society Task Force on Colorectal\n","Cancer; SSP, sessile serrated polyp.\n","© 2017 by the AGA Institute, American College of Gastroenterology, and\n","the American Society for Gastrointestinal Endoscopy.\n","This article is being published jointly in Gastroenterology, American\n","Journal of Gastroenterology, and Gastrointestinal Endoscopy.\n","0016-5085/$36.00\n","http://dx.doi.org/10.1053/j.gastro.2017.05.013\n","Gastroenterology 2017;-:1–17\n","AGA SECTION\n","\n","\n","investigation\n","of\n","patients\n","with\n","symptoms\n","or\n","positive\n","screening tests other than colonoscopy. Colonoscopy is\n","generally the test of choice for diagnostic examinations.\n","Methods\n","Literature Review\n","The English language medical literature using MEDLINE\n","(2005 to August 1, 2016), EMBASE (2005 to third quarter 2016\n","update), the Database of Abstracts of Reviews and Effects\n","(2005 to third quarter 2016 update), and the Cochrane Data-\n","base of Systematic Reviews (2005 to third quarter 2014 up-\n","date)\n","was\n","searched.\n","In\n","MEDLINE,\n","subject\n","headings\n","for\n","colorectal cancer screening were combined with headings for\n","fecal occult blood test, fecal immunochemical test, colonoscopy,\n","sigmoidoscopy, CT colonoscopy, fecal DNA, serum testing, cost-\n","effectiveness, and quality. Similar searches were performed in\n","EMBASE, the Database of Abstracts of Reviews and Effects, and\n","the Cochrane Database of Systematic Reviews. Case reports and\n","studies performed in patients with inﬂammatory bowel disease,\n","prior CRC or polyps, or hereditary CRC syndromes were\n","excluded.\n","Review\n","papers,\n","meta-analyses,\n","gastroenterology\n","textbooks, and editorials were searched manually for additional\n","pertinent references. The review includes studies published\n","since 2008 but also incorporates older evidence used to draft\n","the 2008 recommendations.1 Evidence-based weighted rec-\n","ommendations are provided with supporting discussion to help\n","guide clinicians in the management of these patients.\n","Process and Levels of Evidence\n","Guidance statements were developed by consensus ob-\n","tained through joint teleconferences. The completed article was\n","reviewed and approved by all 3 gastroenterology societies.\n","The use of GRADE for MSTF guidance papers has been\n","outlined in detail elsewhere.2 GRADE involves comprehensive\n","literature search and summary (often through meta-analysis)\n","and then a separate review of literature quality and develop-\n","ment of recommendations. The MSTF uses a modiﬁed qualita-\n","tive approach based on literature review (as described above\n","for this article) but without formal meta-analysis. GRADE\n","allows for a separate assessment of the quality of the evidence\n","and strength of recommendation. This approach explicitly\n","recognizes the importance of literature in informing clinical\n","recommendations but allows latitude because recommenda-\n","tions may be inﬂuenced by other factors, such as patient\n","preference, cost, and expert consensus. “Strong recommenda-\n","tions” are those that would be chosen by most informed pa-\n","tients. “Weak recommendations” are those where patient\n","values and preferences might play a larger role than the quality\n","of evidence. Within the document we preface strong recom-\n","mendations with phrases such as “we recommend” and weak\n","recommendations with “we suggest.”\n","Approaches to Screening\n","In the United States CRC screening usually results from\n","an ofﬁce-based interaction between a healthcare provider\n","and\n","patient.\n","Screening\n","in\n","this\n","setting\n","is\n","termed\n","opportunistic.4\n","Programmatic screening (sometimes called organized\n","screening) refers to a system-wide, organized approach to\n","offering screening to a population or members of a health-\n","care plan.4 Programmatic screening has potential advan-\n","tages over opportunistic screening, including systematic\n","offers of screening, reduction of overscreening, superior\n","monitoring of quality, and systematic follow-up of testing.\n","National CRC screening programs in Europe5 and Australia6\n","use fecal occult blood testing and include screening colo-\n","noscopy in Germany and Poland.5 The United States has no\n","national program for CRC screening, although several large\n","healthcare plans offer programmatic screening, typically\n","with a fecal immunochemical test (FIT).7 Despite the po-\n","tential advantages of programmatic screening, the United\n","States has achieved the world’s highest rates of CRC\n","screening compliance at 60% and the greatest CRC inci-\n","dence and mortality reduction, using an almost entirely\n","opportunistic approach.8–12 Incidence reductions in the\n","United States were 3% to 4% per year and 30% overall in\n","the ﬁrst decade of this century.11,12 High rates of screening\n","in the United States may reﬂect widespread awareness of\n","CRC and insurance coverage of screening. The MSTF antic-\n","ipates growth of programmatic screening within healthcare\n","systems but expects at least short-term continued reliance\n","on opportunistic screening in the United States. Reliance on\n","opportunistic screening can affect the preference for CRC\n","screening, because achieving compliance with tests that\n","should be repeated at short intervals is more challenging in\n","the opportunistic setting.13\n","In the setting of opportunistic screening, healthcare\n","providers can use several broad strategies to offer screening\n","to patients. One approach is multiple options, in which the\n","beneﬁts, risks, and costs of 2 or more tests are discussed\n","and offered to patients (Table 1).14 Some evidence suggests\n","that when patients are offered both colonoscopy and fecal\n","occult blood testing, more patients undergo screening.15\n","Other data suggest no beneﬁt in overall compliance when\n","multiple options are offered.16–18 In 1 study, offering pa-\n","tients 5 options did not enhance compliance over 2\n","Table 1.Approaches to Offering Screening in the Opportunistic Setting\n","Approach\n","Description\n","Multiple options\n","The relative beneﬁts, risks, and costs of 2 or more options are presented\n","Sequential testing\n","A preferred test is offered ﬁrst. If the patients decline another option(s) is offered\n","Risk stratiﬁed approach\n","Colonoscopy is offered to patients predicted to have a high prevalence of advanced pre-cancerous\n","lesions; other tests are offered to patients predicted at low risk\n","2\n","Rex et al\n","Gastroenterology Vol. -, No. -\n","AGA SECTION\n","\n","\n","options.19 In this regard, at least 9 different screening tests\n","(colonoscopy, FIT, guaiac-based fecal occult blood test, FIT–\n","fecal DNA, sigmoidoscopy, sigmoidoscopy plus fecal occult\n","blood test, CT colonography, barium enema, and the Septin9\n","serum assay [Epigenomics, Seattle, Wash]) are endorsed or\n","discussed in recent major screening guidelines.14,20 Thus,\n","the multiple options discussion may best be limited to 2 or 3\n","preferred options. If patients decline all the offered options,\n","1 or more of the other options can be offered.\n","The sequential approach to screening involves an offer of\n","a ﬁrst test that is usually the provider’s preferred screening\n","option; if the patient declines the ﬁrst option, a second test\n","is offered, and so on. In the United States the sequential\n","approach often involves an offer of colonoscopy, followed by\n","FIT if colonoscopy is declined, or another screening test.9\n","Separate guidelines from the American College of Gastro-\n","enterology21 and the American Society for Gastrointestinal\n","Endoscopy22 recommend a sequential approach with colo-\n","noscopy offered ﬁrst. Sequential testing can maximize\n","compliance overall as well as with the test recommended\n","ﬁrst.23–25 Clinicians using the colonoscopy-ﬁrst sequential\n","approach place emphasis on the high efﬁcacy of colonoscopy\n","in preventing CRC and less emphasis on the risks of colo-\n","noscopy. Indeed, high-quality colonoscopy has both higher\n","single-time testing efﬁcacy and greater risks than any other\n","screening tests but with absolute risk rates that are still\n","very low when performed by skilled operators.26 A variant\n","of sequential testing often used in the programmatic setting\n","is to offer patients FIT as the initial or preferred test and\n","have other options such as colonoscopy available to patients\n","who express interest in alternatives.4\n","A third approach to offering screening to average-risk\n","persons is a risk-stratiﬁed approach. Risk stratiﬁcation\n","uses evidence that the “average-risk” population actually\n","represents a wide range of risk that can be estimated based\n","on demographic and other risk factors. For example, older\n","age, male gender, obesity, diabetes, and cigarette smoking\n","are all associated with colorectal adenomas and cancer and\n","therefore might be used in stratifying risk within the\n","average-risk population.21 The goal is to predict subgroups\n","of patients with a high prevalence of important precancer-\n","ous lesions beneﬁting most from referral directly to colo-\n","noscopy, whereas the subgroups with a predicted lower risk\n","(prevalence) of important precancerous lesions are referred\n","for screening tests with less risk and cost than colonoscopy.\n","Risk stratiﬁcation has been poorly accepted because of\n","limited accuracy in discriminating high- and low-prevalence\n","subgroups.27 However, recent validated models appear to\n","be simple to apply and had substantial accuracy in deﬁning\n","high- and low-risk groups for advanced adenomas.28,29\n","There are no clinical trials comparing compliance or other\n","outcomes using a risk-stratiﬁed approach to the multiple\n","options or sequential approaches. Few data are currently\n","available regarding ease of application of a risk-stratiﬁed\n","approach in clinical practice.\n","The MSTF considers that each of the approaches out-\n","lined above is reasonable when offering screening in the\n","opportunistic setting. There is insufﬁcient evidence to\n","identify one approach as superior. Patients undergoing\n","screening tests other than colonoscopy should understand\n","that colonoscopy is used to evaluate these tests when pos-\n","itive. In some instances insurance coverage of colonoscopy\n","performed to evaluate other positive screening tests may be\n","less than coverage of primary screening colonoscopy.\n","Awareness of the different approaches may assist clinicians\n","in understanding screening literature and in selecting an\n","approach to offering screening that seems to be optimal for\n","their practice or for an individual patient.\n","Recommendations\n","1. We recommend that clinicians offer CRC screening\n","beginning at age 50 (strong recommendation, high-\n","quality evidence). (See below for adjustments in recom-\n","mended age for onset of screening based on race and\n","family history.)\n","2. We suggest that sequential offers of screening tests, of-\n","fering multiple screening options, and risk-stratiﬁed\n","screening are all reasonable approaches to offering\n","screening (weak recommendation, low-quality evidence).\n","Screening Targets\n","The object of screening is to reduce CRC incidence and\n","mortality. To accomplish both aims, tests need to detect\n","early-stage (ie, curable) CRCs and high-risk precancerous\n","lesions.1,21 Detection and removal of precancerous lesions\n","prevents CRC.30,31 The 2 main classes of precancerous le-\n","sions in the colon are conventional adenomas and serrated\n","class lesions (Table 2). These 2 classes of precancerous le-\n","sions have distinct endoscopic features and histology and\n","different (though overlapping) distributions within the col-\n","orectum. Speciﬁc screening tests sometimes have particular\n","strengths or weaknesses detecting 1 or the other class of\n","precancerous\n","lesions,\n","particularly\n","the\n","serrated\n","class.\n","Therefore, we review here the main clinical features of the 2\n","classes of precancerous lesions.\n","Adenomas, also known as conventional adenomas, are\n","the precursors of perhaps 70% of all CRCs.32,33 The ade-\n","noma–carcinoma sequence is believed to typically take\n","Table 2.Histologic Classiﬁcation of the Two Major Classes of\n","Colorectal Polyps\n","I. Conventional adenomas\n","a. Dysplasia grade\n","i. High grade\n","ii. Low grade\n","b. Villousity\n","i. Tubular\n","ii. Tubulovillous\n","iii. Villous\n","II. Serrated lesions\n","a. Hyperplastic polyps (not considered precancerous)\n","b. Sessile serrated polyp\n","i. Without cytologic dysplasia\n","ii. With cytologic dysplasia\n","c. Traditional serrated adenoma\n","- 2017\n","MSTF Recommendations for CRC Screening\n","3\n","AGA SECTION\n","\n","\n","more than 10 years to complete in sporadic cancers,\n","whereas much shorter intervals occur in Lynch syndrome.34\n","Correspondingly, colonoscopy is recommended at 10-year\n","intervals in average-risk persons and at 1- to 2-year in-\n","tervals in those with Lynch syndrome.1,34 The distribution\n","of adenomas is relatively even throughout the colon,\n","although adenomas with a ﬂat or depressed morphology are\n","distributed more to the proximal colon and pedunculated\n","lesions more to the distal colon.35 Adenomas are by deﬁ-\n","nition dysplastic, with the overwhelming majority being low\n","grade. The presence of high-grade dysplasia in an adenoma\n","should be noted by a pathologist. Adenomas can also be\n","characterized by tubular versus villous histology, with the\n","overwhelming majority tubular. Lesions with >25% villous\n","elements are termed tubulovillous and those with >75%\n","villous elements villous. Villous elements and invasive can-\n","cer are associated with increasing size of adenomas. Inva-\n","sive cancer in adenomas �5 mm in size is extremely rare,\n","and the prevalence remains well below 1% in adenomas 6\n","to 9 mm in size.36 Recent colonoscopic studies have iden-\n","tiﬁed lower prevalence rates of cancer in polyps <1 cm in\n","size compared with early studies, probably because im-\n","provements in colonoscope technology and performance\n","have led to routine detection of an array of small, ﬂat, low-\n","volume adenomas.36 Interobserver agreement in differen-\n","tiation of high- versus low-grade dysplasia by pathologists\n","and tubular versus tubulovillous histology is poor to\n","moderate, particularly in adenomas <1 cm in size.37\n","Conversely, interobserver agreement between pathologists\n","is good to excellent in placing lesions within the conven-\n","tional adenomas versus serrated polyps and in identifying\n","invasive cancer.38\n","An important clinical concept is the “advanced” ade-\n","noma, deﬁned as a lesion �1 cm in size or having high-grade\n","dysplasia or villous elements.3 Because nonadvanced ade-\n","nomas have a very low prevalence of cancer and a long\n","adenoma–cancer sequence, screening tests can remain\n","useful if they target cancer and advanced adenomas and not\n","small adenomas. Further, the prevalence of nonadvanced\n","adenomas is so high in modern colonoscopy studies that\n","detection of such lesions by noncolonoscopic screening tests\n","leads to unacceptably low speciﬁcity. Colonoscopy has an\n","important beneﬁt over other screening methods because of\n","its ability to detect and remove both advanced and non-\n","advanced adenomas. Although nonadvanced adenomas have\n","limited clinical importance and are not the target of non-\n","colonoscopic screening methods, colonoscopists strive to\n","identify and remove nonadvanced adenomas. Thus, resect-\n","ing lesions with any precancerous potential during colo-\n","noscopy is safe, seems to be better accepted by patients in\n","the United States, and removes them as a clinical concern.\n","Serrated colorectal lesions (Figure 1) represent an\n","emerging area in the ﬁeld of precancerous colorectal lesions.\n","The serrated class of precursor lesions accounts for up to\n","Figure 1. Endoscopic photographs of conventional adenomas and sessile serrated polyps. (A), Small (8-mm diameter) con-\n","ventional adenoma. The red lines are surface blood vessels. (B), A portion of a 40-mm advanced conventional adenoma; one\n","of the targets of all screening tests. The prominent blood vessel pattern is again visible. (C), A conventional adenoma with a\n","focus of invasive cancer. The prominent blood vessel pattern of a conventional adenoma visible over the lesion except in the\n","ulcerated area. The cancer is located at the ulcer (arrows) (D), A sessile serrated polyp without cytologic dysplasia. Note the\n","absence of blood vessels on the surface. (E), A sessile serrated polyp (visualized in narrow-band imaging) with multiple foci of\n","cytologic dysplasia (yellow arrows). The dysplastic areas have the blood vessel pattern (and the histologic features) of an\n","adenoma. The white arrows point to non-dysplastic portions of this sessile serrated polyp. (F), A sessile serrated polyp with\n","invasive cancer; white arrows designate the residual sessile serrated polyp, whereas yellow arrows indicate the ulcerated\n","malignant portion of the lesion.\n","4\n","Rex et al\n","Gastroenterology Vol. -, No. -\n","AGA SECTION\n","\n","\n","30% of CRCs.33 Within the serrated class, hyperplastic\n","polyps are not currently considered precancerous, whereas\n","sessile serrated polyps (SSPs; also known as sessile serrated\n","adenoma) and traditional serrated adenomas are consid-\n","ered precancerous (Table 2).33 Hyperplastic polyps are\n","usually small lesions and are distributed toward the distal\n","colon.39 SSPs are common (found in 8%-9% of screening\n","colonoscopies performed by expert detectors)40,41 and are\n","distributed toward the proximal colon compared with con-\n","ventional adenomas. SSPs are typically ﬂat or sessile in\n","shape, have few or no surface blood vessels (conventional\n","adenomas by comparison have many surface vessels), and\n","are more difﬁcult to detect at colonoscopy than conven-\n","tional adenomas.33,42,43 Because of their prevalence and\n","precancerous potential, SSPs are the major precancerous\n","serrated lesion. There is poor interobserver agreement be-\n","tween pathologists in the differentiation of hyperplastic\n","polyps from SSPs.44 Consequently, clinicians can see widely\n","varying rates of SSPs in pathology reports, depending on the\n","pathologist or even the center in which they practice.45 Most\n","SSPs are not dysplastic, and the lesions should consistently\n","be designated as “SSP without cytologic dysplasia” or “SSP\n","with cytologic dysplasia.”46 When a dysplastic component is\n","present, it is often evident endoscopically (Figure 1) and\n","histologically is a region of conventional adenoma within an\n","otherwise serrated lesion.47 Microdissection studies indi-\n","cate that the dysplastic area often has microsatellite insta-\n","bility.48 The SSP with cytologic dysplasia is considered a\n","more advanced lesion in the polyp cancer sequence than\n","SSP without cytologic dysplasia.3,33,49,50\n","The traditional serrated adenoma is a rare lesion, often in\n","the left colon, sessile, and uniformly dysplastic.33,46 Because\n","traditional serrated adenoma is rare, dysplastic, and has a\n","villous-like growth pattern histologically, it is often mis-\n","interpreted as a tubulovillous conventional adenoma.33\n","The features of these 2 classes of precancerous lesions\n","are relevant to the available screening tests. Colonoscopy is\n","the criterion standard for the detection of all precancerous\n","colorectal lesions. Colonoscopy achieves its greatest supe-\n","riority relative to other screening tests in the detection of\n","conventional adenomas <1 cm in size and serrated class\n","lesions. Detection of SSPs is a major deﬁciency of ﬂexible\n","sigmoidoscopy because SSPs are predominantly in the\n","proximal colon,51 of CT colonography because the lesions\n","tend to be ﬂat,52 and of FIT53 probably because SSPs have\n","no or few surface blood vessels with less tendency to bleed\n","than conventional adenomas. The combined FIT–fecal DNA\n","test achieves its greatest relative performance compared\n","with FIT alone in the detection of serrated class lesions,\n","related to the poor sensitivity of FIT for these lesions and\n","the inclusion of hypermethylation markers in the DNA\n","panel.53 Hypermethylation is a feature of serrated lesions.33\n","Speciﬁc Screening Tests\n","Colonoscopy\n","The advantages of colonoscopy include high sensitivity\n","for\n","cancer\n","and\n","all\n","classes\n","of\n","precancerous\n","lesions,\n","single-session diagnosis and treatment, and long intervals\n","between examinations (10 years) in subjects with normal\n","examinations. One or 2 negative examinations may signal\n","lifetime protection against CRC.54 Patients who value the\n","highest level of sensitivity in detection of precancerous le-\n","sions and are willing to undergo invasive screening should\n","consider choosing colonoscopy. Although no randomized\n","trials of colonoscopy for screening have been completed,\n","extensive evidence from adenoma cohorts,30,31 cohort\n","studies on incidence and mortality,55,56 and case-control\n","studies57–64 support the efﬁcacy of colonoscopy in pre-\n","venting incident CRC and cancer deaths. One cohort study56\n","and\n","3\n","case-control\n","studies58,59,64\n","were\n","performed\n","in\n","screening populations. Reductions in incidence and mor-\n","tality are approximately 80% in the distal colon and 40% to\n","60% in the proximal colon, at least in the United States and\n","Germany.57,59,61,62,64 Furthermore, indirect evidence from\n","randomized trials of fecal occult blood testing65 and\n","sigmoidoscopy,66 as well as studies showing highly variable\n","cancer protection provided by different colonoscopists,67,68\n","also supports a protective effect of colonoscopy against CRC.\n","These ﬁndings are consistent with the observed population\n","trends in the United States.11,12\n","Disadvantages of colonoscopy include the need for\n","thorough bowel cleansing, a higher risk of perforation\n","relative to the other screening tests, higher risk of aspiration\n","pneumonitis (particularly when the procedure is performed\n","with deep sedation),69 a small risk of splenic injury\n","requiring splenectomy, and a greater risk of postprocedural\n","bleeding compared with other screening tests. A meta-\n","analysis of population-based studies found risks of perfo-\n","ration, bleeding, and death of .5 per 1000, 2.6 per 1000, and\n","2.9 per 100,000, respectively.70 Bleeding after colonoscopy\n","is almost entirely related to polypectomy. When electro-\n","cautery is used for resection of all colorectal polyps, most\n","bleeds occur after resection of small lesions. This relates\n","entirely to the high prevalence of these lesions because\n","increasing polyp size and proximal colon location are the\n","major risk factors for bleeding per individual resected\n","polyp.71 Cold resection techniques are effective and nearly\n","devoid of clinically signiﬁcant bleeding risk and can be\n","generally advised for nonpedunculated lesions <1 cm in\n","size.72 Despite these risks, colonoscopy is the preferred\n","approach to management of any benign colorectal polyp\n","regardless of size or location because the alternative of\n","surgical resection has higher mortality and cost compared\n","with colonoscopy.73,74 To the extent that other screening\n","tests effectively identify large lesions, they result in colo-\n","noscopy and do not prevent adverse events related to\n","colonoscopic resection of large lesions.\n","A\n","major\n","disadvantage\n","of\n","colonoscopy\n","is\n","operator\n","dependence in performance. Operator dependence affects\n","detection of cancer,67,68,75 adenomas,76,77 and serrated le-\n","sions40,41,78; selection of appropriate screening and sur-\n","veillance\n","intervals\n","after\n","colonoscopy79;\n","and\n","effective\n","resection of colorectal polyps.80 In general, gastroenterolo-\n","gists performing colonoscopy are more effective than non-\n","gastroenterologists in prevention of cancer62,81–83 and\n","detection of precancerous polyps.84 However, substantial\n","- 2017\n","MSTF Recommendations for CRC Screening\n","5\n","AGA SECTION\n","\n","\n","operator dependence within gastroenterologists is consis-\n","tently observed,42,43,76–78 so that selection of a colono-\n","scopist by specialty is not adequate protection against\n","suboptimal operator performance. Table 3 shows a list of\n","questions that patients can ask potential colonoscopists to\n","judge whether performance is likely to be at a high level.\n","Afterward, the colonoscopy report should contain the items\n","in Table 3 as an additional check on the adequacy of the\n","procedure.\n","Fecal Immunochemical Test\n","Advantages of FIT include its noninvasive nature, 1-time\n","sensitivity for cancer of 79% in 1 meta-analysis,85 fair\n","sensitivity for advanced adenomas (approximately 30%),\n","and low 1-time cost (approximately $20). FIT is recom-\n","mended annually in the United States. The MSTF has\n","recently issued detailed recommendations on the technical\n","performance of FIT86 and considers FIT an essential\n","element of the CRC screening armamentarium for all prac-\n","titioners. FIT is commonly the test of choice in program-\n","matic screening, an excellent second choice for practitioners\n","using sequential testing who offer colonoscopy ﬁrst, and\n","should likely always be one of the tests included in a\n","multiple-options approach. Disadvantages of FIT include the\n","need for repeated testing, which can be problematic in the\n","nonprogrammatic (opportunistic) setting,13 and poor or no\n","sensitivity for serrated class precursor lesions.53 However,\n","there is no evidence that cancers arising through serrated\n","class lesions are less likely to bleed than those arising via\n","adenomas.\n","FIT–Fecal DNA Test\n","The U.S. Food and Drug Administration (FDA) approved\n","a CRC screening test that is a combination of a FIT and\n","markers for abnormal DNA53 (Cologuard; Exact Sciences;\n","Boston, Mass). The Center for Medicaid & Medicare Services\n","approved the test for reimbursement and recommends\n","performance at 3-year intervals. In a large screening colo-\n","noscopy study, patients underwent FIT, the combined FIT–\n","fecal DNA test, and colonoscopy. The FIT–fecal DNA test had\n","a 1-time sensitivity for CRC of 92%. The FIT assay tested in\n","the study had 73.8% sensitivity for cancer, suggesting that\n","most cancer sensitivity of the FIT–fecal DNA test can be\n","achieved without addition of DNA markers. Advantages of\n","the FIT–fecal DNA test include the highest single-time\n","testing sensitivity for cancer of any noninvasive, non-\n","imaging CRC screening test. Also, the study demonstrated\n","40% sensitivity for SSPs >1 cm in size. The sensitivity of FIT\n","for SSPs was equal to the false-positive rate, indicating no\n","sensitivity.\n","The major disadvantages of the FIT–fecal DNA test are a\n","substantial decrease in speciﬁcity (86.6% in persons with\n","normal colonoscopy or nonadvanced lesions and 89.8% in\n","those with normal colonoscopy), compared with 96% for\n","the FIT test alone, and high cost relative to FIT. Speciﬁcity\n","decreased with increasing age and was only 83% in persons\n","aged >65 years. The cost of the FIT–fecal DNA test is\n","approximately $600 for privately insured patients and\n","about $500 for Medicare patients, about 10 times the direct\n","costs of annual FIT. Moreover, there is a further increase in\n","relative costs related to higher numbers of colonoscopies\n","per test. However, speciﬁcity of every 3-year testing with\n","the FIT–fecal DNA test may be approximately equal to the\n","anticipated speciﬁcity over 3 years of annual FIT testing.\n","There is currently no information regarding the program-\n","matic sensitivity of the FIT–fecal DNA test.\n","Annual FIT is more effective and less costly than FIT–\n","fecal DNA every 3 years,87 so the FIT–fecal DNA test is\n","unlikely to replace FIT in large organized screening pro-\n","grams. The FIT–fecal DNA test could be particularly\n","appropriate for patients in the 50- to 65-year age group who\n","seek a noninvasive test with very high sensitivity for cancer,\n","because the test has better speciﬁcity in this age group.\n","Available evidence suggests that asymptomatic patients\n","with a positive FIT–fecal DNA test and a negative high-\n","quality colonoscopy do not need the colonoscopy repeated\n","or evaluation of the remainder of the GI tract.\n","CT Colonography\n","CT colonography has replaced double-contrast barium\n","enema as the test of choice for colorectal imaging for nearly\n","all indications. CT colonography is more effective than\n","barium enema and better tolerated.88,89\n","Advantages of CT colonography include a lower risk of\n","perforation compared with colonoscopy and sensitivity of\n","82% to 92% for adenomas �1 cm in size.88–91 Disadvan-\n","tages of CT colonography include the use of bowel prepa-\n","ration in most centers in the United States. CT colonography\n","can be performed with laxative-free protocols, but this re-\n","sults in clear reductions in sensitivity relative to colonos-\n","copy,91 including for large polyps. The sensitivity of CT\n","Table 3.Tools for Patients to Enhance Colonoscopy Quality\n","Questions for patients to ask prospective colonoscopists to help ensure a high-quality examination\n","1. What is your adenoma detection rate? (should be � 25% overall or � 30% for male patients and � 20% for female patients)\n","2. What is your cecal intubation rate (should be � 95% for screening colonoscopies and � 90% overall)\n","3. Do you use split-dosing of bowel preparations? (effective bowel preparation requires that at least half the preparation be ingested on the\n","day of the colonoscopy)\n","Checks of the endoscopy report after the procedure\n","1. Does the report include photographs of the end of the colon, including the appendiceal oriﬁce and ileocecal valve/terminal ileum?\n","(this demonstrates that the full extent of the colon was examined)\n","2. Is the bowel preparation quality described? (the preparation must be adequate to ensure effective examination)\n","6\n","Rex et al\n","Gastroenterology Vol. -, No. -\n","AGA SECTION\n","\n","\n","colonography for polyps <1 cm is less than colonoscopy,88–91\n","and detection of ﬂat92 and serrated lesions93 are major\n","deﬁciencies of CT colonography. Detection of extracolonic\n","ﬁndings by CT colonography is common, and these ﬁndings\n","have been classiﬁed by the American College of Radiology\n","according to their clinical relevance.94 Radiation exposure is\n","generally viewed as a disadvantage of CT colonography.14\n","Evidence that CT colonography reduces CRC incidence or\n","mortality is lacking.\n","Even in centers where CT colonography has long been\n","available, the impact of CT colonography is limited. At one\n","university, after full development of a CT colonography\n","program, CT colonography accounted for about 10% of\n","colorectal imaging studies, even with the availability of in-\n","surance coverage.95,96 Primary care physicians view the\n","need for frequent follow-up colonoscopy examinations and\n","management of incidental extracolonic ﬁndings as major\n","factors limiting the utility of CT colonography.96 In general,\n","despite an extensive literature investigating the perfor-\n","mance of CT colonography, the test has limited impact on\n","CRC screening compliance.95 However, CT colonography\n","appeals to a niche of patients who are willing to undergo\n","bowel preparation and are concerned about the risks of\n","colonoscopy. When used, the recommended interval is 5\n","years in patients with normal CT colonography. We continue\n","to recommend that patients with polyps �6 mm in size at\n","CT colonography undergo colonoscopy.1\n","Flexible Sigmoidoscopy\n","Randomized controlled trials conﬁrm reductions in\n","distal colon or rectosigmoid cancer incidence and/or mor-\n","tality of 29% to 76% with ﬂexible sigmoidoscopy.66,97–99\n","Flexible sigmoidoscopy can prevent a small fraction (14%)\n","of proximal colon cancers, if liberal criteria are used to\n","indicate colonoscopy based on ﬂexible sigmoidoscopy ﬁnd-\n","ings.66\n","Advantages\n","of\n","ﬂexible\n","sigmoidoscopy\n","include\n","disproportionately lower cost and risk compared with co-\n","lonoscopy, a more limited bowel preparation, and no need\n","for\n","sedation.\n","Disadvantages\n","of\n","ﬂexible\n","sigmoidoscopy\n","include a lower beneﬁt in protection against right-sided\n","colon cancer compared with the level of protection ach-\n","ieved in case-control and cohort studies using colonoscopy.\n","Also, the absence of sedation leads to a low satisfaction\n","experience for patients, such that they are less willing to\n","repeat the examination compared with colonoscopy.100\n","Further, the concept of examining only part of the colon\n","has been unpopular in the United States, so that screening\n","by ﬂexible sigmoidoscopy has almost disappeared from\n","opportunistic\n","screening\n","settings.9\n","Some\n","groups\n","have\n","endorsed the combination of ﬂexible sigmoidoscopy plus\n","FIT for screening,14 but compliance challenges associated\n","with completing 2 screening tests and the dramatic decline\n","in screening ﬂexible sigmoidoscopy make signiﬁcant uptake\n","of this combination unlikely.\n","Flexible sigmoidoscopy when used is often recom-\n","mended at 5-year intervals. However, endoscopic screening\n","in general is more effective in the left than the right side of\n","the colon, and there is no clear reason why ﬂexible\n","sigmoidoscopy should not be recommended at 10-year in-\n","tervals, similar to the recommendation for colonoscopy. The\n","MSTF considers that either 5- or 10-year intervals are\n","acceptable but favors 10-year intervals.\n","Capsule Colonoscopy\n","Capsule colonoscopy has been approved by the FDA for\n","imaging the proximal colon in patients with previous\n","incomplete colonoscopies and more recently for patients\n","who need colorectal imaging but who are not candidates for\n","colonoscopy\n","or\n","sedation.\n","Capsule\n","colonoscopy\n","is\n","not\n","approved by the FDA for screening average-risk persons.\n","Advantages of capsule colonoscopy are the achievement of\n","endoscopic imaging without an invasive procedure and\n","avoiding the risks of colonoscopy. Disadvantages are that\n","the bowel preparation is more extensive than that for co-\n","lonoscopy. Also, because the logistics of performing same-\n","day colonoscopy on patients with positive capsule studies\n","are quite difﬁcult, most patients with positive studies will\n","require re-preparation and colonoscopy on a separate day.\n","In a large screening trial in 884 patients, capsule colonos-\n","copy had 88% sensitivity for detecting patients with a\n","conventional adenoma �6 mm in size but was ineffective for\n","the detection of serrated lesions, and 9% of patients had\n","technically failed examinations for inadequate cleansing or\n","rapid transit of the capsule.101\n","Overall, the burden associated with bowel preparation\n","and the relative superiority of colonoscopy are such that\n","capsule colonoscopy would be expected to appeal to a niche\n","population concerned about the risks of colonoscopy, in a\n","fashion similar to CT colonography. Currently, lack of FDA\n","approval for screening and lack of reimbursement are major\n","obstacles to its use.\n","Septin9 Assay\n","The ﬁrst FDA-approved serum test for CRC screening is\n","the Septin9 assay (Epigenomics, Seattle, Wash). In a large\n","screening colonoscopy study, this test had a sensitivity of\n","48% for detection of CRC and no sensitivity for detection of\n","precancerous polyps.102 The test is expensive relative to\n","FIT.\n","The advantage of the Septin9 test is that it is a serum\n","assay and is at least potentially more convenient for pa-\n","tients. Some patients who refused colonoscopy preferred\n","this test over FIT.103\n","Disadvantages of the Septin9 assay are markedly inferior\n","performance characteristics compared with FIT, including\n","lower sensitivity for cancer, inability to detect advanced\n","adenomas,104 and low cost-effectiveness relative to other\n","screening tests.105 The test appears to have higher sensi-\n","tivity for late-stage compared with early-stage cancer.102\n","The willingness of patients with positive Septin9 tests to\n","undergo colonoscopy remains uncertain. The uncertainties\n","regarding the true clinical utility of Septin 9 makes shared\n","decision-making difﬁcult. Clinicians should inform patients\n","of the uncertain beneﬁts of this test on CRC mortality, the\n","inability of the assay to detect polyps, and the array of su-\n","perior alternatives. The best frequency for performing the\n","- 2017\n","MSTF Recommendations for CRC Screening\n","7\n","AGA SECTION\n","\n","\n","test is uncertain. Given these limitations, the MSTF suggests\n","that Septin9 not be used for screening.\n","Cost Issues\n","A consistent ﬁnding is that CRC screening by any avail-\n","able\n","modality\n","is\n","cost-effective\n","compared\n","with\n","no\n","screening,106,107 and in some models screening results in\n","cost savings. This ﬁnding relates in part to the high costs of\n","CRC treatment. Numerous modeling studies have addressed\n","the relative cost-effectiveness of 2 or more screening tests.\n","The conclusions of the models frequently vary, likely\n","depending in part on the assumptions of the respective\n","models. For example, different models comparing colonos-\n","copy and CT colonography have had variable conclusions.108\n","Consistent trends reveal that FIT performs well compared\n","with other screening tests.106,107,109 Colonoscopy also per-\n","forms well in most models,106–108,110 and, in general, the\n","traditional tests are more cost-effective than the newer\n","modalities, including CT colonography, FIT–fecal DNA,\n","capsule colonoscopy, and the Septin9 assay.105–107,111–113\n","Newer tests could reach cost-effectiveness by substantially\n","increasing compliance,112,113 but evidence of improved\n","compliance is lacking. Some models support the cost-\n","effectiveness of risk-stratiﬁed approaches to screening.114\n","Screening remains cost-effective in patients into their mid-\n","80s if they have few comorbidities and limited prior\n","screening.115,116\n","Quality of Screening\n","Variable performance of screening tests affects at least\n","colonoscopy, sigmoidoscopy, CT colonography, and FIT.\n","Optimal results in CRC screening cannot be achieved\n","without optimizing the technical performance and reporting\n","of tests and ensuring that patients undergo appropriate\n","follow-up after testing. The MSTF has made detailed rec-\n","ommendations regarding the technical performance of FIT86\n","and\n","has\n","previously\n","issued\n","quality\n","recommendations\n","regarding the technical performance of sigmoidoscopy117\n","and colonoscopy.118 The recommendations of the MSTF\n","regarding quality in the technical performance of colonos-\n","copy118 were largely incorporated in quality recommenda-\n","tions\n","from\n","a\n","combined\n","American\n","College\n","of\n","Gastroenterology–American\n","Society\n","for\n","Gastrointestinal\n","Endoscopy Task Force on quality in 2006119 and 2015,120\n","and the MSTF endorses the American College of Gastro-\n","enterology–American Society for Gastrointestinal Endos-\n","copy Task Force recommendations.\n","The burden of performing high-quality FIT falls largely\n","on primary care physicians and/or the healthcare systems\n","in which they work. In the opportunistic setting there may\n","not be resources allocated to systematically ensure that FIT-\n","positive patients are referred for colonoscopy and that FIT-\n","negative patients are offered repeat testing or to monitor\n","whether compliance with quality targets is adequate.13\n","Inability to allocate resources to monitor the quality of\n","FIT testing is a factor favoring reliance on sequential testing\n","with colonoscopy the ﬁrst test offered.\n","Unlike primary care physicians, the main role of gas-\n","troenterologists in the screening process is to perform co-\n","lonoscopy on patients referred for primary colonoscopy\n","screening or for colonoscopy to evaluate other positive\n","screening tests. As such, a primary task of gastroenterolo-\n","gists is to perform high-quality colonoscopy and cost-\n","effective follow-up. The adenoma detection rate, originally\n","proposed by the MSTF in 2002,118 has emerged as the most\n","important and highly variable measure of the quality of\n","mucosal inspection during colonoscopy. Two large studies\n","have validated the adenoma detection rate as a predictor of\n","cancer prevention by colonoscopy.67,68 Measurement of the\n","adenoma detection rate is mandatory to appreciating\n","whether a colonoscopist should be performing screening\n","colonoscopy. Patients should expect a prospective colono-\n","scopist to provide his or her adenoma detection rate, which\n","should meet or exceed recommended minimum thresholds\n","(Table 3).\n","Practical Considerations\n","No published randomized trials have directly compared\n","and reported the relative effects of different tests on CRC\n","incidence or mortality. Several trials are ongoing, but results\n","are not yet available. When compared using simulation\n","models that are dependent on assumptions about natural\n","history of disease, patient acceptance of screening, and test\n","performance, several tests appear to be similarly effec-\n","tive.121 Therefore, practical considerations are important for\n","informing our recommendations.\n","A\n","common\n","statement\n","made\n","with\n","regard\n","to\n","CRC\n","screening is that “the best test is the one that gets done.”\n","The MSTF endorses this concept because it is generally\n","better for any person who is eligible to be screened to un-\n","dergo some screening test rather than not be screened at all.\n","On the other hand, the core concept underlying sequential\n","testing is that offering the “best” test(s) ﬁrst optimizes the\n","sensitivity and/or cost-effectiveness of screening and still\n","leaves the opportunity to offer other tests when patients\n","decline. As an extreme example, the MSTF considers that\n","equating the Septin9 assay with colonoscopy would be a\n","disservice to patients, because the sensitivity of colonoscopy\n","for cancer and advanced lesions exceeds that of Septin9 by a\n","very large margin. Given these disparities in the perfor-\n","mance of individual tests, the MSTF groups the available\n","tests into 3 tiers based on various performance features and\n","costs (Table 4).\n","The tier 1 tests representing the cornerstone of CRC\n","screening are colonoscopy every 10 years and annual FIT\n","(Table 4). The use of these 2 tests as the primary screening\n","measures provides a framework for screening that is simple\n","and accommodates almost every screening setting. In\n","organized programmatic screening FIT will often be offered\n","as the primary screen, but colonoscopy can be considered as\n","an alternative for patients and physicians who prefer or\n","request it. In the opportunistic setting, colonoscopy will\n","often be preferred when the infrastructure to ensure annual\n","performance of FIT is not available. As noted above, 1 of the\n","most\n","challenging\n","aspects\n","of\n","FIT\n","screening\n","in\n","the\n","8\n","Rex et al\n","Gastroenterology Vol. -, No. -\n","AGA SECTION\n","\n","\n","opportunistic setting is ensuring repeated annual perfor-\n","mance. Using a tier 1 approach that focuses on 2 tests makes\n","the discussion of CRC screening tests between physician and\n","patient manageable and feasible, and, as noted above,\n","expanding the number of options in the initial discussion\n","beyond 2 did not increase screening rates.19 Colonoscopy\n","and FIT can be adapted readily to either the sequential offer\n","of screening (colonoscopy is offered ﬁrst with FIT reserved\n","for those who decline colonoscopy), the multiple-options\n","approach (colonoscopy and FIT are each discussed with\n","patients, and if both are declined the discussion moves\n","sequentially to tier 2 tests), and the risk-stratiﬁed approach\n","(eg, colonoscopy is offered ﬁrst to men age >60 and women\n","age >65 with no prior screening, and FIT is offered to\n","persons under these ages and persons with negative prior\n","colonoscopy). Risk stratiﬁcation may also take into account\n","factors such as cigarette smoking, diabetes, and obesity.\n","The rationale for placing tests in the tier 2 and 3 cate-\n","gories follows from the discussion above of individual tests.\n","The ranking implies equivalence for tests within each\n","category, but this is not the intent of the MSTF. For example,\n","ﬂexible sigmoidoscopy screening has strong evidence to\n","support its use. However, the steady decline in the use of\n","ﬂexible sigmoidoscopy in the United States9,10 suggests that\n","any strong endorsement of ﬂexible sigmoidoscopy is not\n","consistent with the reality of the test’s lack of popularity\n","among patients and poor reimbursement for physicians. The\n","strong evidence base supporting ﬂexible sigmoidoscopy\n","leads us to place the test in the tier 2 category, although we\n","expect that most practitioners using a sequential approach\n","would move to CT colonography or FIT–fecal DNA in pa-\n","tients who decline colonoscopy and FIT, because these tests\n","are less invasive and survey the entire colon.\n","Performance characteristics alone would place capsule\n","colonoscopy in tier 2. However, offering capsule colonos-\n","copy as a tier 2 test in a sequential methodology currently\n","would often lead to frustration because reimbursement for\n","screening capsule colonoscopy is seldom available at this\n","time, and the test itself is frequently not available. The\n","onerous bowel preparation and the lack of systems to\n","accomplish same-day colonoscopy in most patients with a\n","positive capsule colonoscopy are additional factors placing\n","capsule colonoscopy in the tier 3 category at this time.\n","In summary, we suggest that the tiered system (Table 4)\n","has numerous advantages. Predominant reliance on tier 1\n","tests offers modalities with optimal effectiveness, cost-\n","effectiveness, complete colon screening, proven popularity\n","with patients, and a simpliﬁed discussion (compared with\n","offering 5-7 different tests) and still leaves room to offer\n","other tests in a sequential fashion. If patients decline colo-\n","noscopy and FIT, all tests in tier 2 are acceptable CRC\n","screening tests, but each has deﬁciencies relative to the\n","tier 1 tests.\n","Recommendations\n","1. We recommend colonoscopy every 10 years or annual\n","FIT as ﬁrst-tier options for screening the average-risk\n","persons for colorectal neoplasia (strong recommenda-\n","tion; moderate-quality evidence).\n","2. We recommend that physicians performing screening\n","colonoscopy measure quality, including the adenoma\n","detection rate (strong recommendation, high-quality\n","evidence).\n","3. We recommend that physicians performing FIT monitor\n","quality (strong recommendation, low-quality evidence).\n","The recommended quality measurements for FIT pro-\n","grams are detailed in a prior publication.86\n","4. We recommend CT colonography every 5 years or FIT–\n","fecal DNA every 3 years (strong recommendation, low-\n","quality evidence) or ﬂexible sigmoidoscopy every 5 to\n","10\n","years\n","(strong\n","recommendation,\n","high-quality\n","evidence) in patients who refuse colonoscopy and FIT.\n","5. We suggest that capsule colonoscopy (if available) is an\n","appropriate screening test when patients decline colo-\n","noscopy, FIT, FIT–fecal DNA, CT colonography, and ﬂex-\n","ible sigmoidoscopy (weak recommendation, low-quality\n","evidence).\n","6. We suggest against Septin9 for CRC screening (weak\n","recommendation, low-quality evidence).\n","Family History of CRC and Polyps\n","We recommend that screening in most average-risk\n","persons be initiated at age 50 years. A family history of\n","CRC or certain polyps can modify the recommended starting\n","age and the frequency of screening. The MSTF has previ-\n","ously issued recommendations for screening in persons\n","with Lynch syndrome,34 which is a genetically deﬁned\n","inherited syndrome caused by mutations in 1 or more\n","mismatch repair genes. Patients in families that meet the\n","clinical criteria for hereditary nonpolyposis CRC but have\n","microsatellite-stable CRCs have family colon cancer syn-\n","drome X, which has not been genetically deﬁned.122 Persons\n","in families with syndrome X should undergo colonoscopy at\n","least every 3 to 5 years, beginning 10 years before the age at\n","diagnosis of the youngest affected relative.\n","A family history of CRC in a ﬁrst-degree relative in-\n","creases the risk of CRC regardless of the age at diagnosis of\n","Table 4.Multi-Society Task Force Ranking of Current\n","Colorectal Cancer Screening Tests\n","Tier 1\n","Colonoscopy every 10 years\n","Annual fecal immunochemical test\n","Tier 2\n","CT colonography every 5 years\n","FIT–fecal DNA every 3 years\n","Flexible sigmoidoscopy every 10 years (or every 5 years)\n","Tier 3\n","Capsule colonoscopy every 5 years\n","Available tests not currently recommended\n","Septin 9\n","- 2017\n","MSTF Recommendations for CRC Screening\n","9\n","AGA SECTION\n","\n","\n","the affected relative.123–125 There is a gradient of risk such\n","that the younger the age of the affected relative, the greater\n","the risk.123–125 The MSTF has previously used age 60 as a\n","threshold of risk elevation, so that a single ﬁrst-degree\n","relative diagnosed with CRC at age <60 years warrants\n","both earlier screening and at more-frequent intervals.1\n","Recent population-based studies123,124 and reviews of risk\n","associated with a positive family history125 support using\n","age at diagnosis of CRC above or below 60 years in the\n","affected ﬁrst-degree relative as a risk stratiﬁer. We continue\n","to recommend that persons with a family history of CRC in a\n","ﬁrst-degree relative diagnosed at <60 years undergo colo-\n","noscopy every 5 years beginning at age 40 years or 10 years\n","before the age the relative was diagnosed, whichever comes\n","ﬁrst (Table 5). In a randomized controlled trial, there was a\n","nonsigniﬁcant trend toward detection of more advanced\n","neoplasia in subjects with a positive family history who\n","underwent colonoscopy compared with FIT.126 Thus, pa-\n","tients with a positive family history who decline colonos-\n","copy should be offered FIT screening.126\n","The greatest relative risk of CRC appears to be in\n","persons <50 years who have a ﬁrst-degree relative with\n","CRC diagnosed at <50 years.123-125 Compliance in young\n","persons with a family history of CRC is suboptimal, and\n","clinicians should make special efforts to ensure that\n","screening occurs. Recent evidence suggests if persons with a\n","single ﬁrst-degree relative with a family history of CRC\n","reach the age of approximately 60 years without manifest-\n","ing signiﬁcant colorectal neoplasia,127 then they are unlikely\n","to be at increased risk of CRC and can be offered the option\n","of expanding the interval between examinations.\n","When ﬁrst-degree relatives have documented advanced\n","serrated lesions (SSPs �10 mm in size, or an SSP with cyto-\n","logic dysplasia, or a traditional serrated adenoma �10 mm in\n","size), there is no clear evidence as to how to proceed, unless\n","the relative meets criteria for serrated polyposis.46 Currently,\n","we recommend that screening for ﬁrst-degree relatives of\n","persons with advanced serrated lesions should be similar to\n","the screening of ﬁrst-degree relatives of persons with\n","advanced conventional adenomas.\n","Persons with a single ﬁrst-degree relative with CRC who\n","was diagnosed at age �60 years are recommended to begin\n","screening at 40 years.128 However, the tests and intervals\n","for testing are the same as the average-risk screening rec-\n","ommendations (Tables 5 and 4).\n","We no longer recommend that persons with a family\n","history of adenomas in a ﬁrst-degree relative undergo early\n","screening, unless there is clear documentation of an\n","advanced adenoma in a ﬁrst-degree relative. In most cases\n","the patient has no information regarding whether the family\n","member’s adenoma was advanced, and in this case we\n","recommend that it be assumed the adenomas or polyps\n","were not advanced. If a colonoscopy and/or pathology re-\n","port(s) is available for a family member that documents an\n","advanced adenoma or there is a report of a polyp requiring\n","surgical resection, an advanced adenoma in a family mem-\n","ber\n","is\n","considered\n","established.\n","These\n","considerations\n","regarding adequate documentation of advanced precancer-\n","ous neoplasms in ﬁrst-degree relatives before intensifying\n","screening apply to documentation of both advanced ade-\n","nomas and advanced serrated lesions. First-degree relatives\n","with advanced adenomas are recommended to be weighted\n","the same as ﬁrst-degree relatives with CRC (Table 5). The\n","yield of colonoscopic screening in ﬁrst-degree relatives of\n","persons\n","with\n","advanced\n","adenomas\n","is\n","substantially\n","increased.129–131\n","Recommendations\n","1. We suggest that persons with 1 ﬁrst-degree relative with\n","CRC or a documented advanced adenoma diagnosed at\n","age <60 years or with 2 ﬁrst-degree relatives with CRC\n","and/or documented advanced adenomas undergo colo-\n","noscopy every 5 years beginning 10 years younger than\n","the age at which the youngest ﬁrst-degree relative was\n","diagnosed or age 40, whichever is earlier (weak recom-\n","mendation, low-quality evidence).\n","2. We suggest that persons with 1 ﬁrst-degree relative\n","diagnosed with CRC or a documented advanced adenoma\n","at age �60 years begin screening at age 40. The options\n","for screening and the recommended intervals are the\n","same as those for average-risk persons (weak recom-\n","mendation, very-low-quality evidence).\n","Table 5.MSTF Recommendations for Persons With High-Risk Family Histories Not Associated With Polyp Syndromes\n","Family history\n","Recommended screening\n","Lynch Syndrome\n","See reference 34\n","Family Colon Cancer Syndrome X\n","Colonoscopy every 3-5 years beginning 10 years before the age at\n","diagnosis of the youngest affected relative\n","Colorectal cancer or an advanced adenoma in two\n","ﬁrst-degree relatives diagnosed at any age\n","OR colorectal cancer or an advanced adenoma\n","in a single ﬁrst-degree relative at age < 60 years\n","Colonoscopy every 5 years beginning 10 years before the age\n","at diagnosis of the youngest affect interval or age 40,\n","whichever is earlier; for those with a single ﬁrst-degree\n","relative with colorectal cancer in whom no signiﬁcant\n","neoplasia appears by age 60 years, physicians can offer\n","expanding the interval between colonoscopies\n","Colorectal cancer or an advanced adenoma in a\n","single ﬁrst-degree relative diagnosed at\n","age � 60 years\n","Begin screening at age 40 years; tests and intervals are as\n","per the average-risk screening recommendations (Table 4)\n","10\n","Rex et al\n","Gastroenterology Vol. -, No. -\n","AGA SECTION\n","\n","\n","3. We suggest that persons with 1 or more ﬁrst-degree\n","relatives with a documented advanced serrated lesion\n","(SSP or traditional serrated adenoma �10 mm in size or\n","an SSP with cytologic dysplasia) should be screened ac-\n","cording to above recommendations for persons with a\n","family history of a documented advanced adenoma\n","(weak recommendation, very-low-quality evidence).\n","4. We recommend that persons with 1 or more ﬁrst-degree\n","relatives with CRC or documented advanced adenomas,\n","for whom we recommend colonoscopy, should be offered\n","annual FIT if they decline colonoscopy (strong recom-\n","mendation, moderate-quality evidence).\n","Considerations Regarding Age\n","and CRC Risk\n","CRC screening is recommended to begin at age 50 years\n","in most average-risk persons, including in prior recom-\n","mendations from the MSTF.1 Recent modeling supports this\n","recommendation.121 Several issues related to age and CRC\n","risk warrant speciﬁc discussion.\n","The incidence of CRC is strongly age related and con-\n","tinues to rise with increasing age. Partly because of wide-\n","spread screening in the United States, the incidence of CRC\n","in falling by 3% to 4% per year in persons age �50 years.12\n","The incidence of CRC in persons under age 50 is increasing\n","in the United States.132,133 Although the reasons for this\n","rising incidence remain unclear and the relative incidence in\n","persons under age 50 remains low, the increasing incidence\n","of CRC in young people is a major public health concern.\n","The best course of action with regard to the rising\n","incidence of CRC in young people is currently not certain.\n","When a young person develops fatal CRC, the loss of life\n","years is great. The ﬁrst step in reducing CRC morbidity and\n","mortality in persons age <50 years is aggressive evaluation\n","(usually colonoscopy) of patients with colorectal symptoms,\n","speciﬁcally those with bleeding symptoms: hematochezia,\n","iron deﬁciency anemia, and/or melena with a negative\n","upper endoscopy. Persons with bleeding symptoms evalu-\n","ated with tests other than colonoscopy (eg, sigmoidoscopy)\n","should have a bleeding source identiﬁed and treated,\n","and the patient should be followed to resolution of the\n","symptom. Patients who have nonbleeding symptoms (eg,\n","abnormal bowel habit, change in bowel habit or shape, or\n","abdominal pain) and who have no evidence of bleeding do\n","not have an increased risk of CRC when they undergo\n","colonoscopy.134,135\n","There is currently insufﬁcient evidence to recommend\n","systematic screening in asymptomatic persons <50 years\n","old who lack speciﬁc risk factors related to family history or\n","Lynch syndrome. The yield of screening colonoscopy in this\n","age group is low in available studies,136 and the biologic\n","reasons for the increasing incidence of CRC in persons un-\n","der age 50 years are uncertain. Additional study of the\n","beneﬁts and harms of screening in persons <50 years is\n","warranted, perhaps particularly in persons with known\n","colorectal risk factors such as cigarette smoking, diabetes\n","mellitus, and obesity.137-139\n","Relative to other races, African Americans have lower\n","screening rates for CRC, higher incidence rates, earlier mean\n","age at onset, worse survival and late-stage presentation, and\n","a higher proportion of cancers before age 50.140–143 These\n","various effects result from both socioeconomic and biologic\n","factors.144 Two of the member organizations of the MSTF\n","endorsed\n","beginning\n","average-risk\n","screening\n","in\n","African\n","Americans at age 45,22,145 and the American College of\n","Physicians recommended beginning at age 40.146 The sci-\n","entiﬁc rationale for beginning screening earlier includes the\n","higher overall incidence rates and younger mean age at\n","onset of CRC in African Americans.140–143 A recent joinpoint\n","regression analysis147and a MISCAN-Colon microsimulation\n","model148 both supported screening African Americans\n","approximately 5 years earlier compared with whites. There\n","are few data on the yield of screening in African Americans\n","before age 50 or whether earlier screening improves out-\n","comes. In persons over age 50, some reports have identiﬁed\n","a higher risk of advanced polyps in African Americans.149\n","The recommendations to begin screening earlier have\n","served an important role in stimulating discussion of and\n","research on CRC in African Americans, increasing awareness\n","in physicians of an important public health problem and\n","racial disparity in health outcomes in the United States, and\n","increasing\n","awareness\n","of\n","CRC\n","in\n","African\n","Americans.\n","Increasing screening rates in African Americans generally is\n","an area of obvious importance. Provider recommendation is\n","key,150 and navigation can improve compliance to colonos-\n","copy screening.151 Additional study of the yield of screening\n","in persons under age 50 is needed, particularly in African\n","Americans.\n","The age to\n","stop screening can be\n","individualized.\n","Screening is potentially beneﬁcial in persons up to age 86 if\n","there has not been previous screening116 but should be\n","considered in the context of comorbidities and life expec-\n","tancy. Persons with previously negative screening tests,\n","particularly negative screening colonoscopy, could consider\n","stopping at age 75 years.14 In a variation of this recom-\n","mendation, the MSTF has recommended that persons with\n","previous negative screening stop when their life expectancy\n","is less than 10 years.1 Thus, the recommendation to stop\n","screening can be reasonably based on patient age and\n","comorbidities.\n","The wishes of the patient should be considered in un-\n","certain cases. These considerations do not necessarily apply\n","in the surveillance setting, where patients with advanced\n","neoplasia may beneﬁt from surveillance colonoscopy even\n","at an advanced age, depending on comorbidities and the\n","conﬁdence in neoplasia clearing at colonoscopy.\n","Recommendations\n","1. We recommend that screening begin in non–African\n","American average-risk persons at age 50 years (strong\n","recommendation; moderate-quality evidence).\n","2. We suggest that screening begin in African Americans at\n","age 45 years (weak recommendation, very-low-quality\n","evidence).\n","- 2017\n","MSTF Recommendations for CRC Screening\n","11\n","AGA SECTION\n","\n","\n","3. We recommend that adults age <50 years with colo-\n","rectal bleeding symptoms (hematochezia, unexplained\n","iron deﬁciency anemia, melena with a negative upper\n","endoscopy) undergo colonoscopy or an evaluation sufﬁ-\n","cient to determine a bleeding cause, initiate treatment,\n","and complete follow-up to determine resolution of\n","bleeding\n","(strong\n","recommendation,\n","moderate-quality\n","evidence).\n","4. We suggest that persons who are up to date with\n","screening and have negative prior screening tests,\n","particularly colonoscopy, consider stopping screening at\n","age 75 years or when life expectancy is less than 10\n","years (weak recommendation, low-quality evidence).\n","5. We suggest that persons without prior screening should\n","be considered for screening up to age 85, depending on\n","consideration of their age and comorbidities (weak\n","recommendation, low-quality evidence).\n","Summary\n","CRC screening should begin at age 50 years in asymp-\n","tomatic persons. Colonoscopy every 10 years and annual\n","FIT are currently the ﬁrst considerations for screening.\n","Colonoscopy every 10 years has advantages in the oppor-\n","tunistic screening setting. Annual FIT is likely to be\n","preferred in organized screening programs. Positioning of\n","the 2 tests can be reasonably based on a sequential offer\n","(colonoscopy ﬁrst with FIT offered to patients who decline\n","colonoscopy, followed by second-tier tests for patients who\n","decline FIT), a multiple-options approach where both tests\n","are discussed with patients (followed by a sequential offer\n","of second-tier tests to patients who decline both colonos-\n","copy and FIT), or a risk-stratiﬁed approach (colonoscopy is\n","offered to patients with a higher pretest probability of\n","neoplasia, and FIT is used in persons with a lower pretest\n","probability of neoplasia).\n","Persons with a history of CRC or a documented\n","advanced adenoma in a ﬁrst-degree relative age <60 years\n","or 2 ﬁrst-degree relatives with these ﬁndings at any age are\n","recommended to undergo screening by colonoscopy every\n","5 years, beginning 10 years before the age at diagnosis of\n","the youngest affected relative, or at age 40, whichever is\n","earlier. Persons with a single ﬁrst-degree relative diag-\n","nosed at �60 years with CRC or an advanced adenoma can\n","be offered average-risk screening options beginning at age\n","40 years.\n","The incidence of CRC is rising in persons under age 50.\n","Patients under age 50 with bleeding symptoms consistent\n","with a colorectal source should be aggressively evaluated\n","and treated. We suggest that screening begin at age 45 in\n","African Americans. Discontinuation of screening should be\n","considered when patients who are up to date with screening\n","and have had negative screening tests, particularly colo-\n","noscopy, reach age 75 years or when life expectancy is <10\n","years. Persons without prior screening should be consid-\n","ered for screening up to age 85 years, depending on co-\n","morbid conditions and life expectancy.\n","References\n","1.Levin B, Lieberman DA, McFarland B, et al. Screening\n","and surveillance for the early detection of colorectal\n","cancer\n","and\n","adenomatous\n","polyps.\n","Gastroenterology\n","2008;134:1570–1595.\n","2.Kahi CJ, Boland CR, Dominitz JA, et al. Colonoscopy\n","surveillance after colorectal cancer resection: recom-\n","mendations of the US multi-society task force on colo-\n","rectal cancer. Gastrointest Endosc 2016;83:489–498.\n","3.Lieberman DA, Rex DK, Winawer SJ, et al. Guidelines for\n","colonoscopy surveillance after screening and poly-\n","pectomy. Gastroenterology 2012;143:844–857.\n","4.Levin TR, Jamieson L, Burley DA, et al. Organized colo-\n","rectal cancer screening in integrated health care sys-\n","tems. Epidemiol Rev 2011;33:101–110.\n","5.Zavoral M, Suchanek S, Zavada F, et al. Colorectal\n","cancer screening in Europe. World J Gastroenterol 2009;\n","15:5907–5915.\n","6.Cenin DR, St John DJB, Ledger MJN, et al. Optimizing\n","the expansion of the National Bowel Cancer Screening\n","Program. Med J Aust 2014;201:456–461.\n","7.Jensen CD, Corley DA, Quinn VP, et al. Fecal immuno-\n","chemical\n","test\n","program\n","over\n","4\n","rounds\n","of\n","annual\n","screening: a retrospective study. Ann Intern Med 2016;\n","164:456–463.\n","8.Siegel RL, Ward EM, Jemal A. Trends in colorectal can-\n","cer incidence rates in the United States by tumor location\n","and stage, 1992-2008. Cancer Epidemiol Biomarkers\n","Prev 2012;21:411–416.\n","9.Klabunde CN, Joseph DA, King JB, et al. Vital signs:\n","colorectal cancer screening test use—United States,\n","2012. MMWR 2013;62:881–888.\n","10.Sabatino SS, White MC, Thompson TT, et al. Cancer\n","screening test use—United States, 2013. MMWR 2015;\n","64:464–468.\n","11.Siegel R, Desantis C, Jemal A. Colorectal cancer statis-\n","tics, 2014. CA Cancer J Clin 2014;64:104–117.\n","12.Edwards BK, Noone AM, Mariotto AB, et al. Annual\n","report to the nation on the status of cancer, 1975-2010,\n","featuring prevalence of comorbidity and impact on sur-\n","vival among persons with lung, colorectal, breast, or\n","prostate cancer. Cancer 2014;120:1290–1314.\n","13.Liang PS, Wheat CL, Abhat A, et al. Adherence to\n","competing strategies for colorectal cancer screening\n","over 3 years. Am J Gastroenterol 2016;111:105–114.\n","14.US Preventive Services Task Force, Bibbins-Domingo K,\n","Grossman DC, Curry SJ, et al. Screening for colorectal\n","cancer: US Preventive Services Task Force recommen-\n","dation statement. JAMA 2016;315:2564–2575.\n","15.Inadomi JM, Vijan S, Janz NK, et al. Adherence to\n","colorectal cancer screening: a randomized clinical trial of\n","competing\n","strategies.\n","Arch\n","Intern\n","Med\n","2012;172:\n","575–582.\n","16.Segnan N, Senore C, Andreoni B, et al. Randomized trial\n","of different screening strategies for colorectal cancer:\n","patient response and detection rates. J Natl Cancer Inst\n","2005;97:147–157.\n","17.Multicentre Australian Colorectal-neoplasia Screening\n","(MACS) Group. A comparison of colorectal neoplasia\n","12\n","Rex et al\n","Gastroenterology Vol. -, No. -\n","AGA SECTION\n","\n","\n","screening tests: a multicentre community-based study of\n","the impact of consumer choice. Med J Aust 2006;\n","184:546–550.\n","18.Scott RG, Edwards JT, Fritschi L, et al. Community-\n","based screening by colonoscopy or computed tomo-\n","graphic colonography in asymptomatic average-risk\n","subjects. Am J Gastroenterol 2004;99:1145–1151.\n","19.Grifﬁth JM, Lewis CL, Brenner ART, et al. The effect of\n","offering different numbers of colorectal cancer screening\n","test options in a decision aid: a pilot randomized trial.\n","BMC Med Inform Dec Mak 2008;8:4.\n","20.Canadian Task Force on Preventive Health Care. Rec-\n","ommendations on screening for colorectal cancer in\n","primary care. CMAJ 2016;188:340–348.\n","21.Rex DK, Johnson DA, Anderson JC, et al. American\n","College of Gastroenterology guidelines for colorectal\n","cancer\n","screening.\n","Am\n","J\n","Gastroenterol\n","2009;104:\n","739–750.\n","22.ASGE Standards of Practice Committee, Davila RE,\n","Rajan E, Baron TH. ASGE guideline: colorectal cancer\n","screening and surveillance. Gastrointest Endosc 2006;\n","63:546–557.\n","23.Senore C, Ederle A, Benazzato L, et al. Offering people a\n","choice for colorectal cancer screening. Gut 2013;62:\n","735–740.\n","24.Symonds EL, Pedersen S, Cole SR, et al. Improving\n","participation in colorectal cancer screening: a random-\n","ized controlled trial of sequential offers of fecal then\n","blood based non-invasive tests. Asian Pac J Cancer Prev\n","2015;16:8455–8460.\n","25.Adler A, Geiger S, Keil A, et al. Improving compliance to\n","colorectal cancer screening using blood and stool based\n","tests in patients refusing screening colonoscopy in\n","Germany. BMC Gastroenterol 2014;14:183.\n","26.Rex D. Colonoscopy: the current king of the hill in the\n","USA. Dig Dis Sci 2015;60:639–646.\n","27.Ma GK, Ladabaum U. Personalizing colorectal cancer\n","screening: a systematic review of models to predict risk\n","of colorectal neoplasia. Clin Gastroenterol Hepatol 2014;\n","12:1624–1634.\n","28.Imperiale TF, Monahan PO, Stump TE, et al. Derivation\n","and validation of a scoring system to stratify risk for\n","advanced colorectal neoplasia in asymptomatic adults: a\n","cross-sectional\n","study.\n","Ann\n","Intern\n","Med\n","2015;163:\n","339–346.\n","29.Chiu HM, Ching JY, Wu KC, et al. A risk-scoring system\n","combined with a fecal immunochemical test is effective\n","in screening high-risk subjects for early colonoscopy to\n","detect advanced colorectal neoplasms. Gastroenter-\n","ology 2016;150:617–625.\n","30.Winawer SJ, Zauber AG, Ho MN. Prevention of colorectal\n","cancer by colonoscopic polypectomy. The National\n","Polyp Study Workgroup. N Engl J Med 1993;329:\n","1977–1981.\n","31.Zauber AG, Winawer SJ, O’Brien MJ, et al. Colonoscopic\n","polypectomy and long-term prevention of colorectal-\n","cancer deaths. N Engl J Med 2012;366:687–696.\n","32.Brenner H, Hoffmeister M, Stegmaier C, et al. Risk of\n","progression of advanced adenomas to colorectal cancer\n","by age and sex: estimates based on 840,149 screening\n","colonoscopies. Gut 2007;56:1585–1589.\n","33.Rex DK, Ahnen DJ, Baron JA, et al. Serrated lesions\n","of the colorectum: review and recommendations from\n","an\n","expert\n","panel.\n","Am\n","J\n","Gastroenterol\n","2012;107:\n","1315–1329.\n","34.Giardiello FM, Allen JI, Axilbund JE, et al. Guidelines on\n","genetic evaluation and management of Lynch syndrome:\n","a consensus statement by the U.S. Multi-Society Task\n","Force on Colorectal Cancer. Gastrointest Endosc 2014;\n","80:197–220.\n","35.Soetikno R, Kaltenbach T, Rouse RV, et al. Prevalence of\n","nonpolypoid (ﬂat and depressed) colorectal neoplasms in\n","asymptomatic and symptomatic adults. JAMA 2008;\n","299:1027–1035.\n","36.Ponugoti PL, Cummings OW, Rex DK. Risk of cancer in\n","small and diminutive colorectal polyps. Dig Liver Dis\n","2017;49:530–534.\n","37.Lasisi F, Mouchli A, Riddell R, et al. Agreement in inter-\n","preting villous elements and dysplasia in adenomas less\n","than one centimeter in size. Dig Liver Dis 2013;45:\n","1049–1055.\n","38.Rex DK, Alikhan M, Cumming O, et al. Accuracy of\n","pathologic interpretation of colorectal polyps by general\n","pathologists in community practice. Gastrointest Endosc\n","1999;50:468–474.\n","39.Ponugoti P, Lin J, Odze R, et al. Prevalence of sessile\n","serrated\n","adenoma/polyp\n","in\n","hyperplastic\n","appearing\n","diminutive rectosigmoid polyps. Gastrointest Endosc\n","2016;85:622–627.\n","40.Abdeljawad K, Vemulapalli KC, Kahi CJ, et al. Sessile\n","serrated polyp prevalence determined by a colono-\n","scopist with a high lesion detection rate and an experi-\n","enced\n","pathologist.\n","Gastrointest\n","Endosc\n","2015;81:\n","517–524.\n","41.Ijspeert JE, de Wit K, van der Vlugt M, et al. Prevalence,\n","distribution and risk of sessile serrated adenomas/polyps\n","at a center with a high adenoma detection rate and\n","experienced pathologists. Endoscopy 2016;48:740–746.\n","42.Hetzel JT, Huang CS, Coukos JA, et al. Variation in the\n","detection of serrated polyps in an average risk colorectal\n","cancer screening cohort. Am J Gastroenterol 2010;\n","105:2656–2664.\n","43.Kahi CJ, Hewett DG, Norton DL, et al. Prevalence and\n","variable detection of proximal colon serrated polyps\n","during\n","screening\n","colonoscopy.\n","Clin\n","Gastroenterol\n","Hepatol 2011;9:42–46.\n","44.Tinmouth J, Henry P, Hsieh E, et al. Sessile serrated\n","polyps at screening colonoscopy: have they been under-\n","diagnosed? Am J Gastroenterol 2014;109:1698–1704.\n","45.Payne SR, Church TR, Wandell M, et al. Endoscopic\n","detection of proximal serrated lesions and pathologic\n","identiﬁcation of sessile serrated adenomas/polyps vary\n","on the basis of center. Clin Gastroenterol Hepatol 2014;\n","12:1119–1126.\n","46.Snover D, Ahnen D, Burt RW, et al. Serrated polyps of the\n","colon and rectum and serrated (“hyperplastic”) polypo-\n","sis. In: Bozman FT, Carneiro F, Hruban RH, et al, eds.\n","WHO classiﬁcation of tumours. Pathology and genetics.\n","- 2017\n","MSTF Recommendations for CRC Screening\n","13\n","AGA SECTION\n","\n","\n","Tumors of the digestive system. Berlin: Springer-Verlag,\n","2010.\n","47.Nanda KS, Tutticci NG, Burgess N, et al. Caught in the\n","act: endoscopic characterization of sessile serrated\n","adenomas with dysplasia. Gastrointest Endosc 2014;\n","79:864–870.\n","48.Sheridan TB, Fenton H, Lewin MR, et al. Sessile serrated\n","adenomas with low and high grade dysplasia and early\n","carcinomas: an immunohistochemical study of serrated\n","lesions “caught in the act.” Am J Clin Pathol 2006;\n","126:564–571.\n","49.Lash RH, Genta RM, Sculer CM. Sessile serrated ade-\n","nomas: prevalence of dysplasia and carcinoma in 2139\n","patients. J Clin Pathol 2010;63:681–686.\n","50.Bettington M, Walker N, Rosty C, et al. Clinicopatho-\n","logical and molecular features of sessile serrated ade-\n","nomas with dysplasia or\n","carcinoma.\n","Gut\n","2017;66:\n","97–106.\n","51.Kahi C, Vemulapalli KC, Snover DC, et al. Findings in the\n","distal colorectum are not associated with proximal\n","advanced serrated lesions. Clin Gastroenterol Hepatol\n","2015;13:345–351.\n","52.IJspeert JE, Tutein Nolthenius CJ, Kuipers EJ, et al. CT-\n","colonography vs. colonoscopy for detection of high-risk\n","sessile serrated polyps. Am J Gastroenterol 2016;\n","111:516–522.\n","53.Imperiale TF, Ransohoff DF, Itzkowitz SH, et al. Multi-\n","target stool DNA testing for colorectal-cancer screening.\n","N Engl J Med 2014;370:1287–1297.\n","54.Brenner H, Chang-Claude J, Seiler CM, et al. Long-term\n","risk of colorectal cancer after negative colonoscopy.\n","J Clin Oncol 2011;29:3761–3767.\n","55.Singh H, Nugent Z, Demers AA, et al. The reduction in\n","colorectal cancer mortality after colonoscopy varies by\n","site\n","of\n","the\n","cancer.\n","Gastroenterology\n","2010;139:\n","1128–1137.\n","56.Kahi CJ, Imperiale TF, Juliar BE, et al. Effect of screening\n","colonoscopy on colorectal cancer incidence and mor-\n","tality. Clin Gastroenterol Hepatol 2009;7:770–775.\n","57.Brenner H, Chang-Claude J, Seiler CM, et al. Protection\n","from colorectal cancer after colonoscopy: a population-\n","based,\n","case-control\n","study.\n","Ann\n","Intern\n","Med\n","2011;\n","154:22–30.\n","58.Doubeni CA, Weinmann S, Adams K, et al. Screening\n","colonoscopy and risk for incident late-stage colorectal\n","cancer diagnosis in average-risk adults: a nested case-\n","control study. Ann Intern Med 2013;158:312–320.\n","59.Brenner H, Chang-Claude J, Jansen L, et al. Reduced\n","risk of colorectal cancer up to 10 years after screening,\n","surveillance or diagnostic colonoscopy. Gastroenter-\n","ology 2014;146:709–717.\n","60.Mulder SA, van Soest EM, Dieleman JP, et al. Exposure\n","to colorectal examinations before a colorectal cancer\n","diagnosis: a case-control study. Eur J Gastroenterol\n","Hepatol 2010;22:437–443.\n","61.Baxter NN, Goldwasser MA, Paszat LF, et al. Association\n","of colonoscopy and death from colorectal cancer. Ann\n","Inern Med 2009;150:1–8.\n","62.Baxter NN, Warren JL, Barrett MJ, et al. Association\n","between colonoscopy and colorectal cancer mortality in\n","a US cohort according to site of cancer and colono-\n","scopist specialty. J Clin Oncol 2012;30:2664–2669.\n","63.Nishihara R, Wu K, Lochhead P, et al. Long-term colo-\n","rectal cancer incidence and mortality after lower endos-\n","copy. N Engl J Med 2013;369:1095–1105.\n","64.Doubeni CA, Corley DA, Quinn VP, et al. Effectiveness of\n","screening colonoscopy in reducing the risk of death from\n","right and left colon cancer: a large community-based\n","study. Gut. Epub 2016, http://dx.doi.org/10.1136/gutjnl-\n","2016-312712.\n","65.Mandel JS, Church TR, Bond JH. The effect of fecal\n","occult-blood screening on the incidence of colorectal\n","cancer. N Engl J Med 2000;343:1603–1607.\n","66.Schoen RE, Pinsky PF, Weissfeld JL, et al. Colorectal-\n","cancer incidence and mortality with ﬂexible sigmoidos-\n","copy. N Engl J Med 2012;366:2345–2357.\n","67.Kaminski MF, Regula J, Kraszewska E, et al. Quality in-\n","dicators for colonoscopy and the risk of interval cancer.\n","N Engl J Med 2010;362:1795–1803.\n","68.Corley DA, Jensen CD, Marks AR, et al. Adenoma\n","detection rate and risk of colorectal cancer and death.\n","N Engl J Med 2014;370:1298–1306.\n","69.Cooper GS, Kou TD, Rex DK. Complications following\n","colonoscopy with anesthesia assistance: a population-\n","based analysis. JAMA Intern Med 2013;173:551–556.\n","70.Reumkens A, Ronadagh EJ, Bakker, et al. Post-colo-\n","noscopy\n","complications:\n","A\n","systematic\n","review,\n","time\n","trends, and meta-analysis of population based studies.\n","Am J Gastroenterol 2016;111:1092–1101.\n","71.Bahin FF, Rasouli KN, Byth K, et al. Prediction of clini-\n","cally signiﬁcant bleeding following wide-ﬁeld endoscopic\n","resection of large sessile and laterally spreading colo-\n","rectal lesions: a clinical risk score. Am J Gastroenterol\n","2016;111:1115–1122.\n","72.Ichise Y, Horiuchi A, Nakayama Y, et al. Prospective\n","randomized comparison of cold snare polypectomy and\n","conventional polypectomy for small colorectal polyps.\n","Digestion 2011;84:78–81.\n","73.Jayanna M, Burgess NG, Singh R, et al. Cost analysis of\n","endoscopic mucosal resection vs surgery for large\n","laterally spreading colorectal lesions. Clin Gastroenterol\n","Hepatol 2016;14:271–278.\n","74.Law R, Das A, Gregory D, et al. Endoscopic resection is\n","cost-effective compared with laparoscopic resection in\n","the management of complex colon polyps: an economic\n","analysis. Gastrointest Endosc 2016;83:1248–1257.\n","75.Shaukat A, Rector TS, Church TR, et al. Longer with-\n","drawal time is associated with a reduced incidence of\n","interval cancer after screening colonoscopy. Gastroen-\n","terology 2015;149:952–957.\n","76.Barclay RL, Vicari JJ, Doughty AS, et al. Colonoscopic\n","withdrawal\n","times\n","and\n","adenoma\n","detection\n","during\n","screening\n","colonoscopy.\n","N\n","Engl\n","J\n","Med\n","2006;355:\n","2533–2541.\n","77.Chen SC, Rex DK. Endoscopist can be more powerful\n","than age and male gender in predicting adenoma\n","detection at colonoscopy. Am J Gastroenterol 2007;\n","102:856–861.\n","78.Butterly L, Robinson CM, Anderson JC, et al. Serrated\n","and adenomatous polyp detection increases with longer\n","14\n","Rex et al\n","Gastroenterology Vol. -, No. -\n","AGA SECTION\n","\n","\n","withdrawal time: results from the New Hampshire Colo-\n","noscopy\n","Registry.\n","Am\n","J\n","Gastroenterol\n","2014;109:\n","417–426.\n","79.Mysliwiec PA, Brown ML, Klabunde CN, Ransohoff DF.\n","Are physicians doing too much colonoscopy? A national\n","survey of colorectal surveillance after polypectomy. Ann\n","Intern Med 2004;141:264–271.\n","80.Pohl H, Srivastava A, Bensen SP, et al. Incomplete polyp\n","resection during colonoscopy-results of the complete\n","adenoma\n","resection\n","(CARE)\n","study.\n","Gastroenterology\n","2013;144:74–80.e1.\n","81.Rex DK, Rahmani EY, Haseman JH, et al. Relative\n","sensitivity of colonoscopy and barium enema for detec-\n","tion of colorectal cancer in clinical practice. Gastroen-\n","terology 1997;112:17–23.\n","82.Rabeneck L, Paszat LF, Saskin R. Endoscopist specialty\n","is associated with incident colorectal cancer after a\n","negative colonoscopy. Clin Gastroenterol Hepatol 2010;\n","8:275–279.\n","83.Hassan C, Rex DK, Zullo A, et al. Loss of efﬁcacy and\n","cost-effectiveness when screening colonoscopy is per-\n","formed\n","by\n","nongastroenterologists.\n","Cancer\n","2012;\n","118:4404–4411.\n","84.Ko CW, Dominitz JA, Green P, et al. Specialty differences\n","in polyp detection, removal, and biopsy during colo-\n","noscopy. Am J Med 2010;123:528–535.\n","85.Lee JK, Liles EG, Bent S, et al. Accuracy of fecal\n","immunochemical tests for colorectal cancer: systematic\n","review\n","and\n","meta-analysis.\n","Ann\n","Intern\n","Med\n","2014;\n","160:171–181.\n","86.Robertson DJ, Lee JK, Boland CR, et al. Recommenda-\n","tions on fecal immunochemical testing to screen for\n","colorectal neoplasia: a consensus statement by the US\n","Multi-Society Task Force on colorectal cancer Gastro-\n","intest Endosc 2016;85:2–21.\n","87.Ladabaum U, Mannalithara A. Comparative effectiveness\n","and cost-effectiveness of a multi-target stool DNA test to\n","screen\n","for\n","colorectal\n","neoplasia.\n","Gastroenterology.\n","2016;151:427-39.e6.\n","88.Rosman AS, Korsten MA. Meta-analysis comparing CT\n","colonography, air contrast barium enema, and colonos-\n","copy meta-analysis comparing CT colonography, air\n","contrast barium enema, and colonoscopy. Am J Med\n","2007;120:203–210.\n","89.Halligan S, Wooldrage K, Dadswell E, et al. Computed\n","tomographic colonography versus barium enema for\n","diagnosis of colorectal cancer or large polyps in symp-\n","tomatic patients (SIGGAR): a multicentre randomised\n","trial. Lancet 2013;381:1185–1193.\n","90.Johnson CD, Herman BA, Chen MH, et al. The National\n","CT Colonography Trial: assessment of accuracy in par-\n","ticipants 65 years of age and older. Radiology 2012;\n","263:401–408.\n","91.Stoop EM, de Haan MC, de Wijkerslooth TR, et al.\n","Participation and yield of colonoscopy versus non-\n","cathartic\n","CT\n","colonography\n","in\n","population-based\n","screening\n","for\n","colorectal\n","cancer:\n","a\n","randomized\n","controlled trial. Lancet Oncol 2012;13:55–64.\n","92.Sakamoto T, Mitsuzaki K, Utsunomiya D, et al. Detection\n","of ﬂat colorectal polyps at screening CT colonography in\n","comparison with conventional polypoid lesions. Acta\n","Radiol 2012;53:714–719.\n","93.IJspeert JE, Tutein Nolthenius CJ, Kuipers EJ, et al.\n","CT-colonography vs. colonoscopy for detection of high-\n","risk sessile serrated polyps. Am J Gastroenterol 2016;\n","111:516–522.\n","94.Zalis ME, Barish MA, Choi JR, et al. CT colonography\n","reporting and data system: a consensus proposal.\n","Radiology 2005;236:3–9.\n","95.Benson M, Pier J, Kraft S, et al. Optical colonoscopy and\n","virtual colonoscopy numbers after initiation of a CT\n","colonography program: long term data. J Gastrointestin\n","Liver Dis 2012;21:391–395.\n","96.Schwartz DC, Dasher KJ, Said A, et al. Impact of a CT\n","colonography screening program on endoscopic colo-\n","noscopy in clinical practice. Am J Gastroenterol 2008;\n","103:346–351.\n","97.Atkin WS, Edwards R, Kralj-Hans I, et al. Once-only\n","ﬂexible sigmoidoscopy screening in prevention of colo-\n","rectal cancer: a multicenter randomized controlled trial.\n","Lancet 2010;375:1624–1633.\n","98.Segnan N, Armaroli P, Bonelli L, et al. Once-only\n","sigmoidoscopy in colorectal cancer screening follow-up\n","ﬁndings of the Italian Randomized Controlled trial-\n","SCORE. JNCI 2011;103:1310–1322.\n","99.Hoff G, Grotmol T, Skovlund E, et al. Risk of colorectal\n","cancer\n","seven\n","years\n","after\n","ﬂexible\n","sigmoidoscopy\n","screening:\n","randomized\n","controlled\n","trial.\n","BMJ\n","2009;\n","338:b1846.\n","100.Zubarik R, Ganguly E, Benway D, et al. Procedure-related\n","abdominal discomfort in patients undergoing colorectal\n","cancer screening: a comparison of colonoscopy and\n","ﬂexible\n","sigmoidoscopy.\n","Am\n","J\n","Gastroenterol\n","2002;\n","97:3056–3061.\n","101.Rex DK, Adler SN, Aisenberg J, et al. Accuracy of\n","capsule colonoscopy in detecting colorectal polyps in a\n","screening\n","population.\n","Gastroenterology\n","2015;148:\n","948–957.e2.\n","102.Church TR, Wandell M, Lofton-Day C, et al. Prospective\n","evaluation of methylated SEPT9 in plasma for detection\n","of\n","asymptomatic\n","colorectal\n","cancer.\n","Gut\n","2014;63:\n","317–325.\n","103.Adler A, Geiger S, Keil A, et al. Improving compliance to\n","colorectal cancer screening using blood and stool based\n","tests in patients refusing screening colonoscopy in\n","Germany. BMC Gastroenterology 2014;14:183.\n","104.Parikh RB, Prasad V. Blood-based screening for colon\n","cancer: a disruptive innovation or simply a disruption?\n","JAMA 2016;315:2519–2520.\n","105.Ladabaum U, Alvarez-Osorio L, Rösch T, et al. Cost-\n","effectiveness of colorectal cancer screening in Germany:\n","current endoscopic and fecal testing strategies versus\n","plasma methylated Septin 9 DNA. Endosc Int Open 2014;\n","2:e96–e104.\n","106.Patel SS, Kilgore ML. Cost effectiveness of colorectal\n","cancer\n","screening\n","strategies.\n","Cancer\n","Control\n","2015;\n","22:248–258.\n","107.Landsdorp-Vogelaar I, Knudsen AB, Brenner H. Cost-\n","effectiveness of colorectal cancer screening. Epidemiol\n","Rev 2011;33:88–100.\n","- 2017\n","MSTF Recommendations for CRC Screening\n","15\n","AGA SECTION\n","\n","\n","108.Hanly P, Skally M, Fenlon H, et al. Cost-effectiveness of\n","computed tomography colonography in colorectal can-\n","cer screening: a systematic review. Int J Technol Assess\n","Health Care 2012;28:415–423.\n","109.Greuter MJ, Berkhof J, Fireman RJ, et al. The potential of\n","imaging techniques as a screening tool for colorectal\n","cancer: a cost-effectiveness analysis. Br J Radiol 2016;\n","89:20150910.\n","110.Wong MC, Ching JY, Chan VC, et al. The comparative\n","cost-effectiveness of colorectal cancer screening using\n","fecal immunochemical test vs colonoscopy. Scientiﬁc\n","Reports 2015; http://dx.doi.org/10.1038/srep13568.\n","111.Skally M, Hanly P, Sharp L, et al. Cost-effectiveness of\n","fecal DNA screening for colorectal cancer: a systematic\n","review and quality appraisal of the literature. Appl Econ\n","Health Policy 2013;11:181–192.\n","112.Knudsen AB, Landsdorp-Vogelaar I, Rutter LM, et al.\n","Cost-effectiveness\n","of\n","computed\n","tomographic\n","colonography screening for colorectal cancer in the\n","medicare population. J Natl Cancer Inst 2010;102:\n","1238–1252.\n","113.Hassan C, Zullo A, Winn S. Cost-effectiveness of capsule\n","endoscopy in screening for colorectal cancer. Endos-\n","copy 2008;40:414–421.\n","114.Dinh T, Ladabaum U, Alperin P, et al. Health beneﬁts and\n","cost-effectiveness of a hybrid screening strategy for\n","colorectal cancer. Clin Gastroenterol Hepatol 2013;\n","11:1158–1166.\n","115.Jenkins M. Colorectal cancer screening is cost-effective\n","in the elderly who have had less intense prior screening,\n","high baseline risk of colorectal cancer and less comor-\n","bidities. Evid Based Med 2016;21:182.\n","116.van Hees F, Habbema JD, Meester RG, et al. Should\n","colorectal cancer screening be considered in elderly\n","persons without previous screening? A cost-effective-\n","ness analysis Ann Intern Med 2014;160:750–759.\n","117.Levin TR, Farraye FA, Schoen RE, et al. Quality in the\n","technical performance of screening ﬂexible sigmoidos-\n","copy: recommendations of an international multi-society\n","task group. Gut 2005;54:807–813.\n","118.Rex DK, Bond JH, Winawer S, et al. Quality in the tech-\n","nical performance of colonoscopy and the continuous\n","quality improvement process for colonoscopy: recom-\n","mendations of the US Multi-Society Task Force in\n","Colorectal\n","Cancer.\n","Am\n","J\n","Gastroenterol\n","2002;97:\n","1296–1308.\n","119.Rex DK, Petrini JL, Baron TH, et al. Quality indicators for\n","colonoscopy. Gastrointest Endosc 2006;63(4 Suppl):\n","S16–S28.\n","120.Rex DK, Schoenfeld PS, Cohen J, et al. Quality indicators\n","for colonoscopy. Am J Gastroenterol 2015;110:72–90.\n","http://dx.doi.org/10.1038/ajg.2014.385.\n","121.Knudsen AB, Zauber AG, Rutter CM, et al. Estimation of\n","beneﬁts,\n","burden,\n","and\n","harms\n","of\n","colorectal\n","cancer\n","screening strategies; modeling study for the US Pre-\n","ventive\n","Services\n","Task\n","Force.\n","JAMA\n","2016;315:\n","2595–2609.\n","122.Lindor NM. Familial colorectal cancer type X: the other\n","half of hereditary nonpolyposis colon cancer syndrome.\n","Surg Oncol Clin N Am 2009:637–645.\n","123.Samadder NJ, Smith KR, Hanson H. Increased risk of\n","colorectal cancer among family members of all ages,\n","regardless of age of index case at diagnosis. Clin Gas-\n","troenterol Hepatol 2015;13:2305–2311.\n","124.Samadder NJ, Curtin K, Tuohy TMF, et al. Increased risk\n","of colorectal neoplasia among family members of pa-\n","tients with colorectal cancer: a population-based study in\n","Utah. Gastroenterology 2014;147:814–821.\n","125.Lowery JT, Ahnen DJ, Schroy PC, et al. Understanding\n","the contribution of family history to colorectal cancer risk\n","and its clinical implications: a state-of-the-science re-\n","view. Cancer 2016;122:2633–2645.\n","126.Quintero E, Carrillo M, Gimeno-Garcia AZ, et al. Equiv-\n","alency of fecal immunochemical tests and colonoscopy\n","in familial colorectal cancer screening. Gastroenterology\n","2014;147:1021–1030.\n","127.Schoen RE, Razzak A, Yu KJ, et al. Incidence and mor-\n","tality of colorectal cancer in individuals with a family\n","history of colorectal cancer. Gastroenterology 2015;\n","149:1438–1445.\n","128.Fuchs\n","CS,\n","Giovannucci\n","EL,\n","Colditz\n","GA,\n","et\n","al.\n","A prospective study of family history and the risk of\n","colorectal cancer. N Engl J Med 1994;331:1669–1674.\n","129.Cottet V, Pariente A, Nalet B, et al. Colonoscopic\n","screening of the ﬁrst-degree relatives of patients with\n","large adenomas; increased risk of colorectal tumors.\n","Gastroenterology 2007;133:1086–1092.\n","130.Wong MC, Ching JY, Chiu HM, et al. Risk of colorectal\n","neoplasia in individuals with self-reported family history:\n","a prospective colonoscopy study from 16 Asia-Paciﬁc\n","regions. Am J Gastroenterol 2016;111:1621–1629.\n","131.Ng SC, Lau JYW, Chan FKL, et al. Risk of advanced\n","adenomas in siblings of individuals with advanced ade-\n","nomas: a cross-sectional study. Gastroenterology 2016;\n","150:608–616.\n","132.Ahnen DJ, Wade SW, Jones WF, et al. The increasing\n","incidence of young-onset colorectal cancer: a call to\n","action. Mayo Clin Proc 2014;89:216–224.\n","133.Printz C. Colorectal cancer incidence increasing in young\n","adults. Cancer 2015;121:1912–1913.\n","134.Rex DK, Mark D, Clarke B, et al. Flexible sigmoidoscopy\n","plus air-contrast barium enema versus colonoscopy for\n","evaluation of symptomatic patients without evidence of\n","bleeding. Gastrointest Endosc 1995;42:132–138.\n","135.Lieberman DA, De Garmo PL, Fleischer DE. Colonic\n","neoplasia in patients with nonspeciﬁc GI symptoms.\n","Gastrointest Endosc 2000;51:647–651.\n","136.Imperiale TF, Wagner DR, Lin CY, et al. Results of\n","screening colonoscopy among persons 40 to 49 years of\n","age. N Engl J Med 2002;346:1781–1785.\n","137.Bardou M, Barkun AN, Martel M. Obesity and colorectal\n","cancer. Gut 2013;62:933–947.\n","138.Hann LM, Jacobs EJ, Thun MJ. The association between\n","cigarette smoking and risk of colorectal cancer in a large\n","prospective cohort from the United States. Cancer\n","Epidemiol Biomarkers Prev 2009;18:3362–3367.\n","139.Guraya SY. Association of type 2 diabetes mellitus and\n","the risk of colorectal cancer: A meta-analysis and sys-\n","tematic\n","review.\n","World\n","J\n","Gastroenterol\n","2015;21:\n","6026–6031.\n","16\n","Rex et al\n","Gastroenterology Vol. -, No. -\n","AGA SECTION\n","\n","\n","140.Williams R, White P, Nieto J, et al. Colorectal cancer in\n","African Americans: an update. Clin Transl Gastroenterol\n","2016;7:e185.\n","141.QuickStats: Colorectal cancer screening among adults\n","aged 50-75 years, by race/ethnicity – National Health\n","interview Study, United States, 2000-2015. MMWR Morb\n","Mortal Wkly Rep 2016;65:1042.\n","142.Carethers\n","JM.\n","Screening\n","for\n","colorectal\n","cancer\n","in\n","African-Americans: determinants and rationale for an\n","earlier age to commence screening. Dig Dis Sci 2015;\n","60:711–721.\n","143.Ashktorab H, Vimenay K, Brim H, et al. Colorectal cancer\n","in young African Americans: is it time to revisit guidelines\n","and prevention? Dig Dis Sci 2016;61:3026–3030.\n","144.Lin J, Qiu M, Xu R, et al. Comparison of survival and\n","clinicopathologic features in colorectal cancer among\n","African American, Caucasian, and Chinese patients\n","treated in the United States: Results from the surveil-\n","lance epidemiology and end results (SEER) database.\n","Oncotarget 2015;6:33935–33943.\n","145.Agrawal S, Bhupinderjit A, Bhutani MS, et al. Colorectal\n","cancer in African Americans. Am J Gastroenterol 2005;\n","100:515–523.\n","146.Qaseem A, Denberg TD. Hopkins RH Screening for\n","colorectal\n","cancer;\n","a\n","guidance\n","statement\n","from\n","the\n","American College of Physicians. Ann Intern Med 2012;\n","156:378–386.\n","147.Paquette IM, Ying J, Shah SA, et al. African Americans\n","should be screened at an earlier age for colorectal can-\n","cer. Gastrointest Endosc 2015;82:878–883.\n","148.Landsdorp-Vogelaar I, van Ballegooijen M, Zauber AG,\n","et al. Individualizing colonoscopy screening by sex and\n","race. Gsatrointest Endosc 2009;70:96–108.\n","149.Lieberman DA, Williams JL, Holub JL, et al. Race,\n","ethnicity, and sex affect risk for polyps > 9 mm in\n","average-risk\n","individuals.\n","Gastroenterology\n","2014;\n","147:351–358.\n","150.Thompson VL, Lander S, Xu S, et al. Identifying key\n","variables in African American adherence to colorectal\n","cancer screening: the application of data mining. BMC\n","Public Health 2014;14:1173. http://dx.doi.org/10.1186/\n","1471-2458-14-1173.\n","151.Horne HN, Phelan-Emrick DF, Pollack CE, et al. Effect of\n","patient navigation on colorectal cancer screening in a\n","community-based randomized controlled trial of urban\n","African American adults. Cancer Causes Control 2015;\n","26:239–246.\n","Reprint requests\n","Address requests for reprints to: Douglas K. Rex, Indiana University Hospital\n","4100, 550 N University Blvd, Indianapolis, IN 46202. e-mail: drex@iu.edu.\n","Acknowledgment\n","The views expressed in this article are those of the authors and do not\n","necessarily reﬂect the position or policy of the Department of Veteran Affairs.\n","Conﬂicts of interest\n","The following authors disclosed ﬁnancial relationships relevant to this\n","publication: D. K. Rex: Consultant for Olympus Corp and Boston Scientiﬁc;\n","research support recipient from Boston Scientiﬁc, Endochoice, EndoAid,\n","Medtronic, and Colonary Solutions. T. Kaltenbach: Consultant for Olympus\n","Corp. D. J. Robertson: Consultant for Medtronic. All other authors disclosed\n","no ﬁnancial relationships relevant to this publication.\n","- 2017\n","MSTF Recommendations for CRC Screening\n","17\n","AGA SECTION\n","\n","\n","Question: What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\n","Helpful Answer:\u001b[0m\n"]},{"ename":"InvalidRequestError","evalue":"ignored","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mInvalidRequestError\u001b[0m                       Traceback (most recent call last)","\u001b[0;32m<ipython-input-17-e865430d1755>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m answer = qa_stuff_chain({\"input_documents\": pages,\n\u001b[0m\u001b[1;32m      2\u001b[0m           \"question\": \"What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\"})\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Answer:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manswer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"output_text\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/base.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, include_run_info)\u001b[0m\n\u001b[1;32m    256\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    257\u001b[0m             \u001b[0mrun_manager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mon_chain_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 258\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    259\u001b[0m         \u001b[0mrun_manager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mon_chain_end\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    260\u001b[0m         final_outputs: Dict[str, Any] = self.prep_outputs(\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/base.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, include_run_info)\u001b[0m\n\u001b[1;32m    250\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    251\u001b[0m             outputs = (\n\u001b[0;32m--> 252\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrun_manager\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrun_manager\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    253\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mnew_arg_supported\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    254\u001b[0m                 \u001b[0;32melse\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/combine_documents/base.py\u001b[0m in \u001b[0;36m_call\u001b[0;34m(self, inputs, run_manager)\u001b[0m\n\u001b[1;32m    104\u001b[0m         \u001b[0;31m# Other keys are assumed to be needed for LLM prediction\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m         \u001b[0mother_keys\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0minputs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mk\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minput_key\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 106\u001b[0;31m         output, extra_return_dict = self.combine_docs(\n\u001b[0m\u001b[1;32m    107\u001b[0m             \u001b[0mdocs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallbacks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0m_run_manager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_child\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mother_keys\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m         )\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/combine_documents/stuff.py\u001b[0m in \u001b[0;36mcombine_docs\u001b[0;34m(self, docs, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    163\u001b[0m         \u001b[0minputs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_inputs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdocs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    164\u001b[0m         \u001b[0;31m# Call predict on the LLM.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 165\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mllm_chain\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcallbacks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallbacks\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    166\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    167\u001b[0m     async def acombine_docs(\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/llm.py\u001b[0m in \u001b[0;36mpredict\u001b[0;34m(self, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    250\u001b[0m                 \u001b[0mcompletion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mllm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madjective\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"funny\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    251\u001b[0m         \"\"\"\n\u001b[0;32m--> 252\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallbacks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallbacks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutput_key\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    253\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    254\u001b[0m     \u001b[0;32masync\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mapredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallbacks\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mCallbacks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mAny\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/base.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, include_run_info)\u001b[0m\n\u001b[1;32m    256\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    257\u001b[0m             \u001b[0mrun_manager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mon_chain_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 258\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    259\u001b[0m         \u001b[0mrun_manager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mon_chain_end\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    260\u001b[0m         final_outputs: Dict[str, Any] = self.prep_outputs(\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/base.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, include_run_info)\u001b[0m\n\u001b[1;32m    250\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    251\u001b[0m             outputs = (\n\u001b[0;32m--> 252\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrun_manager\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrun_manager\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    253\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mnew_arg_supported\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    254\u001b[0m                 \u001b[0;32melse\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/llm.py\u001b[0m in \u001b[0;36m_call\u001b[0;34m(self, inputs, run_manager)\u001b[0m\n\u001b[1;32m     90\u001b[0m         \u001b[0mrun_manager\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mOptional\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mCallbackManagerForChainRun\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     91\u001b[0m     ) -> Dict[str, str]:\n\u001b[0;32m---> 92\u001b[0;31m         \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrun_manager\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrun_manager\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     93\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate_outputs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     94\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/chains/llm.py\u001b[0m in \u001b[0;36mgenerate\u001b[0;34m(self, input_list, run_manager)\u001b[0m\n\u001b[1;32m    100\u001b[0m         \u001b[0;34m\"\"\"Generate LLM result from inputs.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m         \u001b[0mprompts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstop\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprep_prompts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrun_manager\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrun_manager\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 102\u001b[0;31m         return self.llm.generate_prompt(\n\u001b[0m\u001b[1;32m    103\u001b[0m             \u001b[0mprompts\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    104\u001b[0m             \u001b[0mstop\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/llms/base.py\u001b[0m in \u001b[0;36mgenerate_prompt\u001b[0;34m(self, prompts, stop, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    453\u001b[0m     ) -> LLMResult:\n\u001b[1;32m    454\u001b[0m         \u001b[0mprompt_strings\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_string\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mprompts\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 455\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprompt_strings\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstop\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcallbacks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallbacks\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    456\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    457\u001b[0m     async def agenerate_prompt(\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/llms/base.py\u001b[0m in \u001b[0;36mgenerate\u001b[0;34m(self, prompts, stop, callbacks, tags, metadata, **kwargs)\u001b[0m\n\u001b[1;32m    584\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mcallback_manager\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprompt\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcallback_managers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprompts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    585\u001b[0m             ]\n\u001b[0;32m--> 586\u001b[0;31m             output = self._generate_helper(\n\u001b[0m\u001b[1;32m    587\u001b[0m                 \u001b[0mprompts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrun_managers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbool\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_arg_supported\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    588\u001b[0m             )\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/llms/base.py\u001b[0m in \u001b[0;36m_generate_helper\u001b[0;34m(self, prompts, stop, run_managers, new_arg_supported, **kwargs)\u001b[0m\n\u001b[1;32m    490\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mrun_manager\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrun_managers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    491\u001b[0m                 \u001b[0mrun_manager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mon_llm_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 492\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    493\u001b[0m         \u001b[0mflattened_outputs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflatten\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    494\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mmanager\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflattened_output\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_managers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflattened_outputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/llms/base.py\u001b[0m in \u001b[0;36m_generate_helper\u001b[0;34m(self, prompts, stop, run_managers, new_arg_supported, **kwargs)\u001b[0m\n\u001b[1;32m    477\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    478\u001b[0m             output = (\n\u001b[0;32m--> 479\u001b[0;31m                 self._generate(\n\u001b[0m\u001b[1;32m    480\u001b[0m                     \u001b[0mprompts\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    481\u001b[0m                     \u001b[0mstop\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstop\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/llms/openai.py\u001b[0m in \u001b[0;36m_generate\u001b[0;34m(self, prompts, stop, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m    382\u001b[0m                 )\n\u001b[1;32m    383\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 384\u001b[0;31m                 response = completion_with_retry(\n\u001b[0m\u001b[1;32m    385\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprompt\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0m_prompts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrun_manager\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrun_manager\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    386\u001b[0m                 )\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/llms/openai.py\u001b[0m in \u001b[0;36mcompletion_with_retry\u001b[0;34m(llm, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m    114\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mllm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclient\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 116\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_completion_with_retry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    117\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    118\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36mwrapped_f\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    287\u001b[0m         \u001b[0;34m@\u001b[0m\u001b[0mfunctools\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwraps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    288\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mwrapped_f\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 289\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    290\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    291\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mretry_with\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mWrappedFn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[1;32m    377\u001b[0m         \u001b[0mretry_state\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRetryCallState\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mretry_object\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    378\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 379\u001b[0;31m             \u001b[0mdo\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mretry_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mretry_state\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    380\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDoAttempt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36miter\u001b[0;34m(self, retry_state)\u001b[0m\n\u001b[1;32m    312\u001b[0m         \u001b[0mis_explicit_retry\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfut\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfailed\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfut\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTryAgain\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mis_explicit_retry\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mretry_state\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 314\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfut\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    315\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    316\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mafter\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.10/concurrent/futures/_base.py\u001b[0m in \u001b[0;36mresult\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    449\u001b[0m                     \u001b[0;32mraise\u001b[0m \u001b[0mCancelledError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    450\u001b[0m                 \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mFINISHED\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 451\u001b[0;31m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__get_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    452\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    453\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_condition\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.10/concurrent/futures/_base.py\u001b[0m in \u001b[0;36m__get_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    401\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_exception\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    402\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 403\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_exception\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    404\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    405\u001b[0m                 \u001b[0;31m# Break a reference cycle with the exception in self._exception\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[1;32m    380\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDoAttempt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 382\u001b[0;31m                     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    383\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mBaseException\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# noqa: B902\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m                     \u001b[0mretry_state\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexc_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[arg-type]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/llms/openai.py\u001b[0m in \u001b[0;36m_completion_with_retry\u001b[0;34m(**kwargs)\u001b[0m\n\u001b[1;32m    112\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mretry_decorator\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    113\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_completion_with_retry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mAny\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mAny\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 114\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mllm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclient\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    116\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0m_completion_with_retry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_resources/completion.py\u001b[0m in \u001b[0;36mcreate\u001b[0;34m(cls, *args, **kwargs)\u001b[0m\n\u001b[1;32m     23\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mTryAgain\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_resources/abstract/engine_api_resource.py\u001b[0m in \u001b[0;36mcreate\u001b[0;34m(cls, api_key, api_base, api_type, request_id, api_version, organization, **params)\u001b[0m\n\u001b[1;32m    151\u001b[0m         )\n\u001b[1;32m    152\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 153\u001b[0;31m         response, _, api_key = requestor.request(\n\u001b[0m\u001b[1;32m    154\u001b[0m             \u001b[0;34m\"post\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    155\u001b[0m             \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_requestor.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, headers, files, stream, request_id, request_timeout)\u001b[0m\n\u001b[1;32m    296\u001b[0m             \u001b[0mrequest_timeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest_timeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    297\u001b[0m         )\n\u001b[0;32m--> 298\u001b[0;31m         \u001b[0mresp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgot_stream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_interpret_response\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    299\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgot_stream\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapi_key\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    300\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_requestor.py\u001b[0m in \u001b[0;36m_interpret_response\u001b[0;34m(self, result, stream)\u001b[0m\n\u001b[1;32m    698\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    699\u001b[0m             return (\n\u001b[0;32m--> 700\u001b[0;31m                 self._interpret_response_line(\n\u001b[0m\u001b[1;32m    701\u001b[0m                     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"utf-8\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    702\u001b[0m                     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus_code\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_requestor.py\u001b[0m in \u001b[0;36m_interpret_response_line\u001b[0;34m(self, rbody, rcode, rheaders, stream)\u001b[0m\n\u001b[1;32m    761\u001b[0m         \u001b[0mstream_error\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstream\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m\"error\"\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    762\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mstream_error\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mrcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 763\u001b[0;31m             raise self.handle_error_response(\n\u001b[0m\u001b[1;32m    764\u001b[0m                 \u001b[0mrbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrheaders\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstream_error\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstream_error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    765\u001b[0m             )\n","\u001b[0;31mInvalidRequestError\u001b[0m: This model's maximum context length is 4097 tokens, however you requested 26530 tokens (26018 in your prompt; 512 for the completion). Please reduce your prompt; or completion length."]}],"source":["answer = qa_stuff_chain({\"input_documents\": pages,\n","          \"question\": \"What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\"})\n","\n","print(\"Answer:\", answer[\"output_text\"])"]},{"cell_type":"markdown","metadata":{"id":"eZmNRWUgZWIk"},"source":["Ooops!!! When we feed several pages inside our QA chains, the LLM breaks. <br> This is because the prompt length + maximum_tokens_generated exceed the LLM maximum_context_length. <br>\n","\n","Let's explore a few ways which we can engineered around this LLM limitation."]},{"cell_type":"markdown","metadata":{"id":"IFvYxcAyZ_kM"},"source":["# Divide And Conquer.\n","\n","- Since we cannot process everything at once, we can break the task into smaller tasks, which we ask the LLM to answer on each document individually. All the individual answers are then combined to generate the final answer.\n","- Let's try to create a qa_chain with chain_type = **map_reduce**. Let's also ask the chain to return the intermediate answers."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":12907,"status":"ok","timestamp":1691399585823,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"zORjo5Z0D8Ce","outputId":"ff4cbaa6-2d58-40f7-a7f7-ac617a60a21d"},"outputs":[{"name":"stdout","output_type":"stream","text":["Answer:\n"," Persons with a history of CRC or a documented advanced adenoma in a first-degree relative age <60 years or 2 first-degree relatives with these findings at any age are recommended to undergo screening by colonoscopy every 5 years, beginning 10 years before the age at diagnosis of the youngest affected relative, or at age 40, whichever is earlier.\n","\n","Document No 1 Answer:  Persons with a family history of CRC or a documented advanced adenoma in a ﬁrst-degree relative age <60 years or 2 ﬁrst-degree relatives with these ﬁndings at any age are recommended to undergo screening by colonoscopy every 5 years, beginning 10 years before the age at diagnosis of the youngest affected relative or age 40, whichever is earlier.\n","Document No 2 Answer:  No relevant text.\n","Document No 3 Answer:  We recommend that clinicians offer CRC screening beginning at age 50 (strong recommendation, high-quality evidence). (See below for adjustments in recommended age for onset of screening based on race and family history.)\n","Document No 4 Answer:  \"Correspondingly, colonoscopy is recommended at 10-year intervals in average-risk persons and at 1- to 2-year intervals in those with Lynch syndrome.1,34\"\n","Document No 5 Answer:  Colonoscopy is the criterion standard for the detection of all precancerous colorectal lesions. Colonoscopy achieves its greatest superiority relative to other screening tests in the detection of conventional adenomas <1 cm in size and serrated class lesions. Detection of SSPs is a major deficiency of flexible sigmoidoscopy because SSPs are predominantly in the proximal colon,51 of CT colonography because the lesions tend to be flat,52 and of FIT53 probably because SSPs have no or few surface blood vessels with less tendency to bleed than conventional adenomas. The combined FIT–fecal DNA test achieves its greatest relative performance compared with FIT alone in the detection of serrated class lesions, related to the poor sensitivity of FIT for these lesions and the inclusion of hypermethylation markers in the DNA panel.53 Hypermethylation is a feature of serrated lesions.33\n","Document No 6 Answer:  Table 3 shows a list of questions that patients can ask potential colonoscopists to judge whether performance is likely to be at a high level. Afterward, the colonoscopy report should contain the items in Table 3 as an additional check on the adequacy of the procedure. \n","Questions for patients to ask prospective colonoscopists to help ensure a high-quality examination\n","1. What is your adenoma detection rate? (should be � 25% overall or � 30% for male patients and � 20% for female patients)\n","2. What is your cecal intubation rate (should be � 95% for screening colonoscopies and � 90% overall)\n","3. Do you use split-dosing of bowel preparations? (effective bowel preparation requires that at least half the preparation be ingested on the day of the colonoscopy)\n","Checks of the endoscopy report after the procedure\n","1. Does the report include photographs of the end of the colon, including the appendiceal oriﬁce and ileocecal valve/terminal ileum?\n","(this demonstrates that the full extent of the colon was examined)\n","2. Is the bowel preparation quality described? (the preparation must be adequate to ensure effective examination)\n","Document No 7 Answer:  We continue to recommend that patients with polyps �6 mm in size at CT colonography undergo colonoscopy.1\n","Document No 8 Answer:  The tier 1 tests representing the cornerstone of CRC screening are colonoscopy every 10 years and annual FIT (Table 4). The use of these 2 tests as the primary screening measures provides a framework for screening that is simple and accommodates almost every screening setting. In organized programmatic screening FIT will often be offered as the primary screen, but colonoscopy can be considered as an alternative for patients and physicians who prefer or request it.\n","Document No 9 Answer:  \"Persons in families with syndrome X should undergo colonoscopy at least every 3 to 5 years, beginning 10 years before the age at diagnosis of the youngest affected relative.\"\n","Document No 10 Answer:  \n","\"Family Colon Cancer Syndrome X: Colonoscopy every 3-5 years beginning 10 years before the age at diagnosis of the youngest affected relative.\"\n","Document No 11 Answer:  We suggest that persons with 1 or more ﬁrst-degree relatives with a documented advanced serrated lesion (SSP or traditional serrated adenoma -10 mm in size or an SSP with cytologic dysplasia) should be screened according to above recommendations for persons with a family history of a documented advanced adenoma (weak recommendation, very-low-quality evidence). We recommend that persons with 1 or more ﬁrst-degree relatives with CRC or documented advanced adenomas, for whom we recommend colonoscopy, should be offered annual FIT if they decline colonoscopy (strong recommendation, moderate-quality evidence).\n","Document No 12 Answer:  \n","\"Persons with a history of CRC or a documented advanced adenoma in a first-degree relative age <60 years or 2 first-degree relatives with these findings at any age are recommended to undergo screening by colonoscopy every 5 years, beginning 10 years before the age at diagnosis of the youngest affected relative, or at age 40, whichever is earlier.\"\n","Document No 13 Answer:  34.Giardiello FM, Allen JI, Axilbund JE, et al. Guidelines on genetic evaluation and management of Lynch syndrome: a consensus statement by the U.S. Multi-Society Task Force on Colorectal Cancer. Gastrointest Endosc 2014;80:197–220.\n","Document No 14 Answer: \n","No relevant text.\n","Document No 15 Answer:  No relevant text.\n","Document No 16 Answer:  122.Lindor NM. Familial colorectal cancer type X: the other half of hereditary nonpolyposis colon cancer syndrome. Surg Oncol Clin N Am 2009:637–645.\n","Document No 17 Answer:  No relevant text.\n"]}],"source":["qa_mapreduce_chain = load_qa_chain(\n","    llm=OpenAI(temperature=0, max_tokens=512),\n","    chain_type=\"map_reduce\",\n","    return_intermediate_steps=True)\n","\n","answer = qa_mapreduce_chain({\"input_documents\": pages,\n","          \"question\": \"What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\"})\n","\n","print(\"Answer:\\n\" + answer[\"output_text\"] + \"\\n\")\n","\n","for idx, intermediate_step in enumerate(answer[\"intermediate_steps\"]):\n","    print(f\"Document No {idx + 1} Answer:\", intermediate_step)"]},{"cell_type":"markdown","metadata":{"id":"R90s35hka-iz"},"source":["Great. The LLM manages to run without exceeding tokens limit. The answer is also accurate. <br> We can also return the intermediate answers to debug LLM's behaviour."]},{"cell_type":"markdown","metadata":{"id":"uvj9cQWJmqZi"},"source":["# Semantic Search with Vectorstore Database"]},{"cell_type":"markdown","metadata":{"id":"HJEF1kbycD9P"},"source":["Alternatively, we can implement a vector database to store all of our documents and only retrieve the relevant documents to our question. <br>\n","There are several advantages of this approach:\n","- Documents can be stuffed inside a single prompt. This will help preserve the documents granularity as compared to combining answers in a multi-stage process.\n","- Reduce tokens consumption. Divide-And-Conquer technically still runs through all documents, even for completely irrelevant documents.\n","- Higher quality context helps LLM focus on the more important text and reduce hallucinations.\n","- Can also be combined with Divide-And-Conquer strategy."]},{"cell_type":"markdown","metadata":{"id":"G5Un-hA6EjDz"},"source":["## Create vector index database"]},{"cell_type":"markdown","metadata":{"id":"wOTWgtUNdlkO"},"source":["We will create a vector index database using Facebook AI Similarity Search Database. Facebook AI Similarity Search (Faiss) is a library for efficient similarity search and clustering of dense vectors.<br>\n","\n","We use OpenAIEmbeddings model to convert the texts into the vector representations which will be used for cosine similarity search\n","\n","Typically, the loaded documents are further split into smaller text chunks with some overlapping window. The size of the text chunk is a parameter for tuning. Try not to use a very large chunk as you may run into maximum context length exceeding."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":576},"executionInfo":{"elapsed":37769,"status":"error","timestamp":1691399860186,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"7oP2pUvBwK6u","outputId":"d1885552-44f8-4b7d-8bf5-6cc18d55351e"},"outputs":[{"name":"stdout","output_type":"stream","text":["Number of text chunks: 1609\n"]},{"name":"stderr","output_type":"stream","text":["WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-ETbTMTN2JXYOUPDJgxuH2iO6 on tokens per min. Limit: 1000000 / min. Current: 701628 / min. Contact us through our help center at help.openai.com if you continue to have issues..\n","WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-ETbTMTN2JXYOUPDJgxuH2iO6 on tokens per min. Limit: 1000000 / min. Current: 912807 / min. Contact us through our help center at help.openai.com if you continue to have issues..\n","WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-ETbTMTN2JXYOUPDJgxuH2iO6 on tokens per min. Limit: 1000000 / min. Current: 834644 / min. Contact us through our help center at help.openai.com if you continue to have issues..\n","WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 8.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-ETbTMTN2JXYOUPDJgxuH2iO6 on tokens per min. Limit: 1000000 / min. Current: 756831 / min. Contact us through our help center at help.openai.com if you continue to have issues..\n","WARNING:langchain.embeddings.openai:Retrying langchain.embeddings.openai.embed_with_retry.<locals>._embed_with_retry in 10.0 seconds as it raised RateLimitError: Rate limit reached for default-text-embedding-ada-002 in organization org-ETbTMTN2JXYOUPDJgxuH2iO6 on tokens per min. Limit: 1000000 / min. Current: 612774 / min. Contact us through our help center at help.openai.com if you continue to have issues..\n"]},{"ename":"RateLimitError","evalue":"ignored","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mRateLimitError\u001b[0m                            Traceback (most recent call last)","\u001b[0;32m<ipython-input-20-a767420c3f53>\u001b[0m in \u001b[0;36m<cell line: 11>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0membeddings\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mOpenAIEmbeddings\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0mdocsearch\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFAISS\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_documents\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtexts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0membeddings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m \u001b[0mdocsearch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave_local\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mMAIN_DIR\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"vectorstore\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Document Store saved successfully!!!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/vectorstores/base.py\u001b[0m in \u001b[0;36mfrom_documents\u001b[0;34m(cls, documents, embedding, **kwargs)\u001b[0m\n\u001b[1;32m    418\u001b[0m         \u001b[0mtexts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpage_content\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0md\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdocuments\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    419\u001b[0m         \u001b[0mmetadatas\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetadata\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0md\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdocuments\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 420\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_texts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtexts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0membedding\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmetadatas\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmetadatas\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    421\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    422\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/vectorstores/faiss.py\u001b[0m in \u001b[0;36mfrom_texts\u001b[0;34m(cls, texts, embedding, metadatas, ids, **kwargs)\u001b[0m\n\u001b[1;32m    575\u001b[0m                 \u001b[0mfaiss\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFAISS\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfrom_texts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtexts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0membeddings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    576\u001b[0m         \"\"\"\n\u001b[0;32m--> 577\u001b[0;31m         \u001b[0membeddings\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0membedding\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0membed_documents\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtexts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    578\u001b[0m         return cls.__from(\n\u001b[1;32m    579\u001b[0m             \u001b[0mtexts\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/embeddings/openai.py\u001b[0m in \u001b[0;36membed_documents\u001b[0;34m(self, texts, chunk_size)\u001b[0m\n\u001b[1;32m    470\u001b[0m         \u001b[0;31m# NOTE: to keep things simple, we assume the list may contain texts longer\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    471\u001b[0m         \u001b[0;31m#       than the maximum context and use length-safe embedding function.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 472\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_len_safe_embeddings\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtexts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdeployment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    473\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    474\u001b[0m     async def aembed_documents(\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/embeddings/openai.py\u001b[0m in \u001b[0;36m_get_len_safe_embeddings\u001b[0;34m(self, texts, engine, chunk_size)\u001b[0m\n\u001b[1;32m    356\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    357\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_iter\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 358\u001b[0;31m             response = embed_with_retry(\n\u001b[0m\u001b[1;32m    359\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    360\u001b[0m                 \u001b[0minput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtokens\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0mi\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0m_chunk_size\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/embeddings/openai.py\u001b[0m in \u001b[0;36membed_with_retry\u001b[0;34m(embeddings, **kwargs)\u001b[0m\n\u001b[1;32m    105\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_check_response\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    106\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 107\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_embed_with_retry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    108\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    109\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36mwrapped_f\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    287\u001b[0m         \u001b[0;34m@\u001b[0m\u001b[0mfunctools\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwraps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    288\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mwrapped_f\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 289\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    290\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    291\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mretry_with\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAny\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mWrappedFn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[1;32m    377\u001b[0m         \u001b[0mretry_state\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRetryCallState\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mretry_object\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    378\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 379\u001b[0;31m             \u001b[0mdo\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mretry_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mretry_state\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    380\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDoAttempt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36miter\u001b[0;34m(self, retry_state)\u001b[0m\n\u001b[1;32m    323\u001b[0m             \u001b[0mretry_exc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretry_error_cls\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfut\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    324\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreraise\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 325\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mretry_exc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreraise\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    326\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mretry_exc\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mfut\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36mreraise\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    156\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mreraise\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mNoReturn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    157\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlast_attempt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfailed\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 158\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlast_attempt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    159\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    160\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.10/concurrent/futures/_base.py\u001b[0m in \u001b[0;36mresult\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    449\u001b[0m                     \u001b[0;32mraise\u001b[0m \u001b[0mCancelledError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    450\u001b[0m                 \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mFINISHED\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 451\u001b[0;31m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__get_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    452\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    453\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_condition\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.10/concurrent/futures/_base.py\u001b[0m in \u001b[0;36m__get_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    401\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_exception\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    402\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 403\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_exception\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    404\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    405\u001b[0m                 \u001b[0;31m# Break a reference cycle with the exception in self._exception\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tenacity/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[1;32m    380\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDoAttempt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 382\u001b[0;31m                     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    383\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mBaseException\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# noqa: B902\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m                     \u001b[0mretry_state\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexc_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[arg-type]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/langchain/embeddings/openai.py\u001b[0m in \u001b[0;36m_embed_with_retry\u001b[0;34m(**kwargs)\u001b[0m\n\u001b[1;32m    102\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mretry_decorator\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    103\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_embed_with_retry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mAny\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mAny\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 104\u001b[0;31m         \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0membeddings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclient\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    105\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_check_response\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    106\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_resources/embedding.py\u001b[0m in \u001b[0;36mcreate\u001b[0;34m(cls, *args, **kwargs)\u001b[0m\n\u001b[1;32m     31\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m                 \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m                 \u001b[0;31m# If a user specifies base64, we'll just return the encoded string.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_resources/abstract/engine_api_resource.py\u001b[0m in \u001b[0;36mcreate\u001b[0;34m(cls, api_key, api_base, api_type, request_id, api_version, organization, **params)\u001b[0m\n\u001b[1;32m    151\u001b[0m         )\n\u001b[1;32m    152\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 153\u001b[0;31m         response, _, api_key = requestor.request(\n\u001b[0m\u001b[1;32m    154\u001b[0m             \u001b[0;34m\"post\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    155\u001b[0m             \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_requestor.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, headers, files, stream, request_id, request_timeout)\u001b[0m\n\u001b[1;32m    296\u001b[0m             \u001b[0mrequest_timeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest_timeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    297\u001b[0m         )\n\u001b[0;32m--> 298\u001b[0;31m         \u001b[0mresp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgot_stream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_interpret_response\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    299\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgot_stream\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapi_key\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    300\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_requestor.py\u001b[0m in \u001b[0;36m_interpret_response\u001b[0;34m(self, result, stream)\u001b[0m\n\u001b[1;32m    698\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    699\u001b[0m             return (\n\u001b[0;32m--> 700\u001b[0;31m                 self._interpret_response_line(\n\u001b[0m\u001b[1;32m    701\u001b[0m                     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"utf-8\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    702\u001b[0m                     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus_code\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/openai/api_requestor.py\u001b[0m in \u001b[0;36m_interpret_response_line\u001b[0;34m(self, rbody, rcode, rheaders, stream)\u001b[0m\n\u001b[1;32m    761\u001b[0m         \u001b[0mstream_error\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstream\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m\"error\"\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    762\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mstream_error\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mrcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 763\u001b[0;31m             raise self.handle_error_response(\n\u001b[0m\u001b[1;32m    764\u001b[0m                 \u001b[0mrbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrheaders\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstream_error\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstream_error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    765\u001b[0m             )\n","\u001b[0;31mRateLimitError\u001b[0m: Rate limit reached for default-text-embedding-ada-002 in organization org-ETbTMTN2JXYOUPDJgxuH2iO6 on tokens per min. Limit: 1000000 / min. Current: 832054 / min. Contact us through our help center at help.openai.com if you continue to have issues."]}],"source":["document_files = [os.path.join(MAIN_DIR, \"raw_data\", path) for path in os.listdir(os.path.join(MAIN_DIR, \"raw_data\"))]\n","documents = []\n","for doc_file in document_files:\n","     documents.extend(PyMuPDFLoader(doc_file).load()) # Convert all pdf files into list of documents\n","\n","text_splitter = RecursiveCharacterTextSplitter(chunk_size=2000, chunk_overlap=500) # Split document into smaller chunks. Each chunk containing ~600-700 tokens\n","texts = text_splitter.split_documents(documents)\n","print(\"Number of text chunks:\", len(texts))\n","\n","embeddings = OpenAIEmbeddings()\n","docsearch = FAISS.from_documents(texts, embeddings)\n","docsearch.save_local(os.path.join(MAIN_DIR, \"vectorstore\"))\n","print(\"Document Store saved successfully!!!\")"]},{"cell_type":"markdown","metadata":{"id":"FYkR2GUhgMev"},"source":["Creation of new vector database requires running all documents through Embeddings model. This cost tokens & $$$. <br>\n","\n","Once you have the database, simply load it instead of recreating the database again."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":554,"status":"ok","timestamp":1691336462168,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"oGgEsMhMIiXS","outputId":"212b8428-234b-4a6f-cc30-ad38890a27fb"},"outputs":[{"name":"stdout","output_type":"stream","text":["Document Store loaded successfully!!!\n"]}],"source":["docsearch = FAISS.load_local(os.path.join(MAIN_DIR, \"vectorstore\"), OpenAIEmbeddings())\n","print(\"Document Store loaded successfully!!!\")"]},{"cell_type":"markdown","metadata":{"id":"_cObwuaVgmBP"},"source":["Now let's return the top 2 relevant documents w.r.t our previous question"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":831,"status":"ok","timestamp":1691336465913,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"oGAcSQMCf0JC","outputId":"a579abb0-8207-48c8-9dc8-43ee76d8c985"},"outputs":[{"data":{"text/plain":["[Document(page_content='Family history\\nRecommended screening\\nLynch Syndrome\\nSee reference 34\\nFamily Colon Cancer Syndrome X\\nColonoscopy every 3-5 years beginning 10 years before the age at\\ndiagnosis of the youngest affected relative\\nColorectal cancer or an advanced adenoma in two\\nﬁrst-degree relatives diagnosed at any age\\nOR colorectal cancer or an advanced adenoma\\nin a single ﬁrst-degree relative at age < 60 years\\nColonoscopy every 5 years beginning 10 years before the age\\nat diagnosis of the youngest affect interval or age 40,\\nwhichever is earlier; for those with a single ﬁrst-degree\\nrelative with colorectal cancer in whom no signiﬁcant\\nneoplasia appears by age 60 years, physicians can offer\\nexpanding the interval between colonoscopies\\nColorectal cancer or an advanced adenoma in a\\nsingle ﬁrst-degree relative diagnosed at\\nage � 60 years\\nBegin screening at age 40 years; tests and intervals are as\\nper the average-risk screening recommendations (Table 4)\\n10\\nRex et al\\nGastroenterology Vol. -, No. -\\nAGA SECTION', metadata={'source': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'file_path': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'page': 9, 'total_pages': 17, 'format': 'PDF 1.7', 'title': 'Colorectal Cancer Screening: Recommendations for Physicians and Patients From the U.S. Multi-Society Task Force on Colorectal Cancer', 'author': 'Douglas K. Rex', 'subject': 'Gastroenterology, Corrected proof. doi:10.1053/j.gastro.2017.05.013', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170601213054+05'30'\", 'modDate': \"D:20170601213137+05'30'\", 'trapped': ''}),\n"," Document(page_content='of second-tier tests to patients who decline both colonos-\\ncopy and FIT), or a risk-stratiﬁed approach (colonoscopy is\\noffered to patients with a higher pretest probability of\\nneoplasia, and FIT is used in persons with a lower pretest\\nprobability of neoplasia).\\nPersons with a history of CRC or a documented\\nadvanced adenoma in a ﬁrst-degree relative age <60 years\\nor 2 ﬁrst-degree relatives with these ﬁndings at any age are\\nrecommended to undergo screening by colonoscopy every\\n5 years, beginning 10 years before the age at diagnosis of\\nthe youngest affected relative, or at age 40, whichever is\\nearlier. Persons with a single ﬁrst-degree relative diag-\\nnosed at �60 years with CRC or an advanced adenoma can\\nbe offered average-risk screening options beginning at age\\n40 years.\\nThe incidence of CRC is rising in persons under age 50.\\nPatients under age 50 with bleeding symptoms consistent\\nwith a colorectal source should be aggressively evaluated\\nand treated. We suggest that screening begin at age 45 in\\nAfrican Americans. Discontinuation of screening should be\\nconsidered when patients who are up to date with screening\\nand have had negative screening tests, particularly colo-\\nnoscopy, reach age 75 years or when life expectancy is <10\\nyears. Persons without prior screening should be consid-\\nered for screening up to age 85 years, depending on co-\\nmorbid conditions and life expectancy.\\nReferences\\n1.Levin B, Lieberman DA, McFarland B, et al. Screening\\nand surveillance for the early detection of colorectal\\ncancer\\nand\\nadenomatous\\npolyps.\\nGastroenterology\\n2008;134:1570–1595.\\n2.Kahi CJ, Boland CR, Dominitz JA, et al. Colonoscopy\\nsurveillance after colorectal cancer resection: recom-\\nmendations of the US multi-society task force on colo-\\nrectal cancer. Gastrointest Endosc 2016;83:489–498.\\n3.Lieberman DA, Rex DK, Winawer SJ, et al. Guidelines for\\ncolonoscopy surveillance after screening and poly-\\npectomy. Gastroenterology 2012;143:844–857.', metadata={'source': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'file_path': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'page': 11, 'total_pages': 17, 'format': 'PDF 1.7', 'title': 'Colorectal Cancer Screening: Recommendations for Physicians and Patients From the U.S. Multi-Society Task Force on Colorectal Cancer', 'author': 'Douglas K. Rex', 'subject': 'Gastroenterology, Corrected proof. doi:10.1053/j.gastro.2017.05.013', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170601213054+05'30'\", 'modDate': \"D:20170601213137+05'30'\", 'trapped': ''})]"]},"execution_count":36,"metadata":{},"output_type":"execute_result"}],"source":["docsearch.similarity_search(\"What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\",\n","                            k=2)"]},{"cell_type":"markdown","metadata":{"id":"xwoisI2-JtY6"},"source":["## Question Answering with Vectorsearch"]},{"cell_type":"markdown","metadata":{"id":"K7aN6YfXgwB2"},"source":["Now we will create a RetrievalQAWithSourcesChain which will integrate our QA from docs with the vector database. <br>\n","\n","Under the hood, the retriever will retrieve/filter the most relevant documents from the database before stuffing it inside our context placeholder."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"kmge1lwmI57v"},"outputs":[],"source":["qasource_chain_4 = RetrievalQAWithSourcesChain.from_chain_type(\n","    llm=ChatOpenAI(model_name=\"gpt-3.5-turbo\", temperature=0, max_tokens=512),\n","    chain_type=\"stuff\",\n","    retriever=docsearch.as_retriever(search_kwargs={\"k\":4}), # k is the number of documents to retrieve from\n","    return_source_documents=True, # Return Source documents\n","    reduce_k_below_max_tokens=False, # If True, the chain will automatically select the number of documents to below max_tokens. Default to 3375 tokens.\n","    verbose=True\n",")"]},{"cell_type":"markdown","metadata":{"id":"uYlQpb3YhlK2"},"source":["Let's Investigate the Prompt"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":6,"status":"ok","timestamp":1691336711827,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"GSziCSceKsuX","outputId":"4679750d-4cdb-47b2-8e16-de568bcd45e9"},"outputs":[{"name":"stdout","output_type":"stream","text":["Given the following extracted parts of a long document and a question, create a final answer with references (\"SOURCES\"). \n","If you don't know the answer, just say that you don't know. Don't try to make up an answer.\n","ALWAYS return a \"SOURCES\" part in your answer.\n","\n","QUESTION: Which state/country's law governs the interpretation of the contract?\n","=========\n","Content: This Agreement is governed by English law and the parties submit to the exclusive jurisdiction of the English courts in  relation to any dispute (contractual or non-contractual) concerning this Agreement save that either party may apply to any court for an  injunction or other relief to protect its Intellectual Property Rights.\n","Source: 28-pl\n","Content: No Waiver. Failure or delay in exercising any right or remedy under this Agreement shall not constitute a waiver of such (or any other)  right or remedy.\n","\n","11.7 Severability. The invalidity, illegality or unenforceability of any term (or part of a term) of this Agreement shall not affect the continuation  in force of the remainder of the term (if any) and this Agreement.\n","\n","11.8 No Agency. Except as expressly stated otherwise, nothing in this Agreement shall create an agency, partnership or joint venture of any  kind between the parties.\n","\n","11.9 No Third-Party Beneficiaries.\n","Source: 30-pl\n","Content: (b) if Google believes, in good faith, that the Distributor has violated or caused Google to violate any Anti-Bribery Laws (as  defined in Clause 8.5) or that such a violation is reasonably likely to occur,\n","Source: 4-pl\n","=========\n","FINAL ANSWER: This Agreement is governed by English law.\n","SOURCES: 28-pl\n","\n","QUESTION: What did the president say about Michael Jackson?\n","=========\n","Content: Madam Speaker, Madam Vice President, our First Lady and Second Gentleman. Members of Congress and the Cabinet. Justices of the Supreme Court. My fellow Americans.  \n","\n","Last year COVID-19 kept us apart. This year we are finally together again. \n","\n","Tonight, we meet as Democrats Republicans and Independents. But most importantly as Americans. \n","\n","With a duty to one another to the American people to the Constitution. \n","\n","And with an unwavering resolve that freedom will always triumph over tyranny. \n","\n","Six days ago, Russia’s Vladimir Putin sought to shake the foundations of the free world thinking he could make it bend to his menacing ways. But he badly miscalculated. \n","\n","He thought he could roll into Ukraine and the world would roll over. Instead he met a wall of strength he never imagined. \n","\n","He met the Ukrainian people. \n","\n","From President Zelenskyy to every Ukrainian, their fearlessness, their courage, their determination, inspires the world. \n","\n","Groups of citizens blocking tanks with their bodies. Everyone from students to retirees teachers turned soldiers defending their homeland.\n","Source: 0-pl\n","Content: And we won’t stop. \n","\n","We have lost so much to COVID-19. Time with one another. And worst of all, so much loss of life. \n","\n","Let’s use this moment to reset. Let’s stop looking at COVID-19 as a partisan dividing line and see it for what it is: A God-awful disease.  \n","\n","Let’s stop seeing each other as enemies, and start seeing each other for who we really are: Fellow Americans.  \n","\n","We can’t change how divided we’ve been. But we can change how we move forward—on COVID-19 and other issues we must face together. \n","\n","I recently visited the New York City Police Department days after the funerals of Officer Wilbert Mora and his partner, Officer Jason Rivera. \n","\n","They were responding to a 9-1-1 call when a man shot and killed them with a stolen gun. \n","\n","Officer Mora was 27 years old. \n","\n","Officer Rivera was 22. \n","\n","Both Dominican Americans who’d grown up on the same streets they later chose to patrol as police officers. \n","\n","I spoke with their families and told them that we are forever in debt for their sacrifice, and we will carry on their mission to restore the trust and safety every community deserves.\n","Source: 24-pl\n","Content: And a proud Ukrainian people, who have known 30 years  of independence, have repeatedly shown that they will not tolerate anyone who tries to take their country backwards.  \n","\n","To all Americans, I will be honest with you, as I’ve always promised. A Russian dictator, invading a foreign country, has costs around the world. \n","\n","And I’m taking robust action to make sure the pain of our sanctions  is targeted at Russia’s economy. And I will use every tool at our disposal to protect American businesses and consumers. \n","\n","Tonight, I can announce that the United States has worked with 30 other countries to release 60 Million barrels of oil from reserves around the world.  \n","\n","America will lead that effort, releasing 30 Million barrels from our own Strategic Petroleum Reserve. And we stand ready to do more if necessary, unified with our allies.  \n","\n","These steps will help blunt gas prices here at home. And I know the news about what’s happening can seem alarming. \n","\n","But I want you to know that we are going to be okay.\n","Source: 5-pl\n","Content: More support for patients and families. \n","\n","To get there, I call on Congress to fund ARPA-H, the Advanced Research Projects Agency for Health. \n","\n","It’s based on DARPA—the Defense Department project that led to the Internet, GPS, and so much more.  \n","\n","ARPA-H will have a singular purpose—to drive breakthroughs in cancer, Alzheimer’s, diabetes, and more. \n","\n","A unity agenda for the nation. \n","\n","We can do this. \n","\n","My fellow Americans—tonight , we have gathered in a sacred space—the citadel of our democracy. \n","\n","In this Capitol, generation after generation, Americans have debated great questions amid great strife, and have done great things. \n","\n","We have fought for freedom, expanded liberty, defeated totalitarianism and terror. \n","\n","And built the strongest, freest, and most prosperous nation the world has ever known. \n","\n","Now is the hour. \n","\n","Our moment of responsibility. \n","\n","Our test of resolve and conscience, of history itself. \n","\n","It is in this moment that our character is formed. Our purpose is found. Our future is forged. \n","\n","Well I know this nation.\n","Source: 34-pl\n","=========\n","FINAL ANSWER: The president did not mention Michael Jackson.\n","SOURCES:\n","\n","QUESTION: {question}\n","=========\n","{summaries}\n","=========\n","FINAL ANSWER:\n"]}],"source":["print(qasource_chain_4.combine_documents_chain.llm_chain.prompt.template)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":8257,"status":"ok","timestamp":1691336820814,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"bSmowjg-M_Fa","outputId":"bd584ffa-211c-4f97-fd2c-f44a5827d46d"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new RetrievalQAWithSourcesChain chain...\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]}],"source":["answer_from_source = qasource_chain_4(\"What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1691336820814,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"JxIsPC1JNb-A","outputId":"d627d7cb-d4b6-4cf9-8ae2-80769ff5640e"},"outputs":[{"name":"stdout","output_type":"stream","text":["('Answer: For patients with a history of family members having Colon Cancer '\n"," 'Syndrome X, the recommended colonoscopy screening is every 3-5 years '\n"," 'beginning 10 years before the age at diagnosis of the youngest affected '\n"," 'relative. If there is a single first-degree relative with colorectal cancer '\n"," 'at age <60 years, the recommended screening is every 5 years beginning 10 '\n"," 'years before the age at diagnosis of the youngest affected relative or age '\n"," '40, whichever is earlier. If there is a single first-degree relative '\n"," 'diagnosed at age ≥60 years, screening should begin at age 40 years and '\n"," 'follow the average-risk screening recommendations. \\n')\n"]}],"source":["pprint(\"Answer: \"+answer_from_source[\"answer\"])"]},{"cell_type":"markdown","metadata":{"id":"57U_daM4iDgb"},"source":["The answer is accurate. We can also check the document sources which the retriever feed to the LLMs."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":562,"status":"ok","timestamp":1691336827539,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"jmoHHaSuOOS7","outputId":"263b7fa9-3406-4c6f-900f-fbba823aa31a"},"outputs":[{"name":"stdout","output_type":"stream","text":["Document 1: MSTF 2017 CRC Screening.pdf\n","('Family history\\n'\n"," 'Recommended screening\\n'\n"," 'Lynch Syndrome\\n'\n"," 'See reference 34\\n'\n"," 'Family Colon Cancer Syndrome X\\n'\n"," 'Colonoscopy every 3-5 years beginning 10 years before the age at\\n'\n"," 'diagnosis of the youngest affected relative\\n'\n"," 'Colorectal cancer or an advanced adenoma in two\\n'\n"," 'ﬁrst-degree relatives diagnosed at any age\\n'\n"," 'OR colorectal cancer or an advanced adenoma\\n'\n"," 'in a single ﬁrst-degree relative at age < 60 years\\n'\n"," 'Colonoscopy every 5 years beginning 10 years before the age\\n'\n"," 'at diagnosis of the youngest affect interval or age 40,\\n'\n"," 'whichever is earlier; for those with a single ﬁrst-degree\\n'\n"," 'relative with colorectal cancer in whom no signiﬁcant\\n'\n"," 'neoplasia appears by age 60 years, physicians can offer\\n'\n"," 'expanding the interval between colonoscopies\\n'\n"," 'Colorectal cancer or an advanced adenoma in a\\n'\n"," 'single ﬁrst-degree relative diagnosed at\\n'\n"," 'age � 60 years\\n'\n"," 'Begin screening at age 40 years; tests and intervals are as\\n'\n"," 'per the average-risk screening recommendations (Table 4)\\n'\n"," '10\\n'\n"," 'Rex et al\\n'\n"," 'Gastroenterology Vol. -, No. -\\n'\n"," 'AGA SECTION')\n","\n","\n","Document 2: MSTF 2017 CRC Screening.pdf\n","('of second-tier tests to patients who decline both colonos-\\n'\n"," 'copy and FIT), or a risk-stratiﬁed approach (colonoscopy is\\n'\n"," 'offered to patients with a higher pretest probability of\\n'\n"," 'neoplasia, and FIT is used in persons with a lower pretest\\n'\n"," 'probability of neoplasia).\\n'\n"," 'Persons with a history of CRC or a documented\\n'\n"," 'advanced adenoma in a ﬁrst-degree relative age <60 years\\n'\n"," 'or 2 ﬁrst-degree relatives with these ﬁndings at any age are\\n'\n"," 'recommended to undergo screening by colonoscopy every\\n'\n"," '5 years, beginning 10 years before the age at diagnosis of\\n'\n"," 'the youngest affected relative, or at age 40, whichever is\\n'\n"," 'earlier. Persons with a single ﬁrst-degree relative diag-\\n'\n"," 'nosed at �60 years with CRC or an advanced adenoma can\\n'\n"," 'be offered average-risk screening options beginning at age\\n'\n"," '40 years.\\n'\n"," 'The incidence of CRC is rising in persons under age 50.\\n'\n"," 'Patients under age 50 with bleeding symptoms consistent\\n'\n"," 'with a colorectal source should be aggressively evaluated\\n'\n"," 'and treated. We suggest that screening begin at age 45 in\\n'\n"," 'African Americans. Discontinuation of screening should be\\n'\n"," 'considered when patients who are up to date with screening\\n'\n"," 'and have had negative screening tests, particularly colo-\\n'\n"," 'noscopy, reach age 75 years or when life expectancy is <10\\n'\n"," 'years. Persons without prior screening should be consid-\\n'\n"," 'ered for screening up to age 85 years, depending on co-\\n'\n"," 'morbid conditions and life expectancy.\\n'\n"," 'References\\n'\n"," '1.Levin B, Lieberman DA, McFarland B, et al. Screening\\n'\n"," 'and surveillance for the early detection of colorectal\\n'\n"," 'cancer\\n'\n"," 'and\\n'\n"," 'adenomatous\\n'\n"," 'polyps.\\n'\n"," 'Gastroenterology\\n'\n"," '2008;134:1570–1595.\\n'\n"," '2.Kahi CJ, Boland CR, Dominitz JA, et al. Colonoscopy\\n'\n"," 'surveillance after colorectal cancer resection: recom-\\n'\n"," 'mendations of the US multi-society task force on colo-\\n'\n"," 'rectal cancer. Gastrointest Endosc 2016;83:489–498.\\n'\n"," '3.Lieberman DA, Rex DK, Winawer SJ, et al. Guidelines for\\n'\n"," 'colonoscopy surveillance after screening and poly-\\n'\n"," 'pectomy. Gastroenterology 2012;143:844–857.')\n","\n","\n","Document 3: MSTF 2017 CRC Screening.pdf\n","('inherited syndrome caused by mutations in 1 or more\\n'\n"," 'mismatch repair genes. Patients in families that meet the\\n'\n"," 'clinical criteria for hereditary nonpolyposis CRC but have\\n'\n"," 'microsatellite-stable CRCs have family colon cancer syn-\\n'\n"," 'drome X, which has not been genetically deﬁned.122 Persons\\n'\n"," 'in families with syndrome X should undergo colonoscopy at\\n'\n"," 'least every 3 to 5 years, beginning 10 years before the age at\\n'\n"," 'diagnosis of the youngest affected relative.\\n'\n"," 'A family history of CRC in a ﬁrst-degree relative in-\\n'\n"," 'creases the risk of CRC regardless of the age at diagnosis of\\n'\n"," 'Table 4.Multi-Society Task Force Ranking of Current\\n'\n"," 'Colorectal Cancer Screening Tests\\n'\n"," 'Tier 1\\n'\n"," 'Colonoscopy every 10 years\\n'\n"," 'Annual fecal immunochemical test\\n'\n"," 'Tier 2\\n'\n"," 'CT colonography every 5 years\\n'\n"," 'FIT–fecal DNA every 3 years\\n'\n"," 'Flexible sigmoidoscopy every 10 years (or every 5 years)\\n'\n"," 'Tier 3\\n'\n"," 'Capsule colonoscopy every 5 years\\n'\n"," 'Available tests not currently recommended\\n'\n"," 'Septin 9\\n'\n"," '- 2017\\n'\n"," 'MSTF Recommendations for CRC Screening\\n'\n"," '9\\n'\n"," 'AGA SECTION')\n","\n","\n","Document 4: MSTF 2017 CRC Screening.pdf\n","('Hereditary GI Cancer Syndromes\\n'\n"," '© 2015 by the American College of Gastroenterology \\n'\n"," 'The American Journal of GASTROENTEROLOGY\\n'\n"," ' \\n'\n"," '229\\n'\n"," 'surveillance of MMR germline mutation-positive patients, consid-\\n'\n"," 'eration should be given to annual colonoscopy, as several studies \\n'\n"," 'have demonstrated CRC development with surveillance intervals \\n'\n"," 'that are between 1 and 2 years ( 29,47 ). In carriers of deleterious \\n'\n"," ' MSH6 and  PMS2 mutations, the risk of CRC is less and age of diag-\\n'\n"," 'nosis later ( 37,49 ) than in patients with  MLH1 and MSH2 muta-\\n'\n"," 'tions. In these aff ected individuals, consideration could be given \\n'\n"," 'to starting surveillance at age 25–30 in  MSH6 and  PMS2 carriers \\n'\n"," '( 24 ), unless an early-onset cancer exists in a given family; however, \\n'\n"," '( 44 ) reported the results of a prospective observational study of \\n'\n"," 'colonoscopy surveillance of members in hereditary nonpolyposis \\n'\n"," 'colorectal cancer or LS families, revealing a 72% decrease in mor-\\n'\n"," 'tality from CRC in those undergoing screening. In several stud-\\n'\n"," 'ies ( 45–48 ), more frequent colonoscopy screening (≤2 years) was \\n'\n"," 'associated with an earlier stage of CRC at diagnosis and less CRC \\n'\n"," 'than less frequent colonoscopy. At least every 2-year colonoscopic \\n'\n"," 'surveillance of LS patients is supported by the data above and the \\n'\n"," 'rapid adenoma–carcinoma sequence reported in these patients. In \\n'\n"," ' Table 4 .  Amsterdam criteria, revised Bethesda guidelines, and colorectal '\n"," 'cancer risk assessment tool \\n'\n"," ' Amsterdam criteria I ( 24 ) \\n'\n"," ' At least three relatives with colorectal cancer (CRC); all of the following '\n"," 'criteria should be present: \\n'\n"," '  One should be a ﬁ rst-degree relative of the other two; \\n'\n"," '  At least two successive generations must be affected; \\n'\n"," '  At least one of the relatives with CRC must have received the diagnosis '\n"," 'before the age of 50 years; \\n'\n"," '  Familial adenomatous polyposis should be excluded; \\n'\n"," '  Tumors should be veriﬁ ed by pathologic examination. \\n'\n"," ' Amsterdam criteria II ( 24 )')\n","\n","\n"]}],"source":["for idx, document in enumerate(answer_from_source[\"source_documents\"]):\n","    print(f\"Document {idx + 1}:\", answer_from_source[\"source_documents\"][0].metadata[\"source\"].split(\"/\")[-1])\n","    pprint(document.page_content)\n","    print(\"\\n\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1691336901502,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"KiIHyeGhQoBN","outputId":"221924ed-c904-4f83-ef5d-85616e98c81c"},"outputs":[{"name":"stdout","output_type":"stream","text":["Total number of tokens from retrieved docs: 1458\n"]}],"source":["print(\"Total number of tokens from retrieved docs:\", check_documents_token(answer_from_source[\"source_documents\"]))"]},{"cell_type":"markdown","metadata":{"id":"lczom5OXWUyQ"},"source":["Let's try another example"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":6746,"status":"ok","timestamp":1691336912364,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"hQwzTNCqWW8F","outputId":"31b846a9-ca38-4daf-8d06-6fc1720352cf"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new RetrievalQAWithSourcesChain chain...\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]}],"source":["answer_from_source = qasource_chain_4(\"For Lynch syndrome (LS) patients, what is the guidelines for screening for gastric cancer? What is the guidelines for Endometrial cancer?\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1691336913517,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"1gxbmYWrXsdx","outputId":"05bfb760-9a88-42ce-b05f-9b9cb8f18f2b"},"outputs":[{"name":"stdout","output_type":"stream","text":["('Answer: For Lynch syndrome (LS) patients, the guidelines for screening for '\n"," 'gastric cancer recommend baseline esophagogastroduodenoscopy (EGD) with '\n"," 'gastric biopsy at age 30-35 years, and treatment of H. pylori infection when '\n"," 'found. Ongoing surveillance every 3-5 years may be considered if there is a '\n"," 'family history of gastric or duodenal cancer. For endometrial cancer, '\n"," 'screening should be offered to women at risk for or affected with LS by '\n"," 'endometrial biopsy and transvaginal ultrasound annually, starting at age 30 '\n"," 'to 35 years before undergoing surgery or if surgery is deferred.\\n')\n"]}],"source":["pprint(\"Answer: \"+answer_from_source[\"answer\"])"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":348,"status":"ok","timestamp":1691336924503,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"lehVGqgYZlKU","outputId":"49195724-4977-480b-ba72-a3f6f37e072a"},"outputs":[{"name":"stdout","output_type":"stream","text":["Document 1: ACG 2015 Hereditary GI Cancers.pdf\n","('Hereditary GI Cancer Syndromes\\n'\n"," '© 2015 by the American College of Gastroenterology \\n'\n"," 'The American Journal of GASTROENTEROLOGY\\n'\n"," ' \\n'\n"," '225\\n'\n"," ' Table 2  .  Summary of recommendations \\n'\n"," ' Lynch syndrome (LS) \\n'\n"," '  1. In individuals at risk for or affected with LS, screening for '\n"," 'colorectal cancer by colonoscopy should be performed at least every 2 years, '\n"," 'beginning \\n'\n"," 'between ages 20 and 25 years. Annual colonoscopy should be considered in '\n"," 'conﬁ rmed mutation carriers (strong recommendation, moderate quality of \\n'\n"," 'evidence for screening, and very low quality of evidence for annual '\n"," 'surveillance and age of initiation). \\n'\n"," '  2. Colectomy with ileorectal anastomosis (IRA) is the preferred treatment '\n"," 'of patients affected with LS with colon cancer or colonic neoplasia not '\n"," 'controllable \\n'\n"," 'by endoscopy. Segmental colectomy is an option in patients unsuitable for '\n"," 'total colectomy if regular postoperative surveillance is conducted '\n"," '(conditional \\n'\n"," 'recommendation, moderate quality of evidence). \\n'\n"," '  3. Hysterectomy and bilateral salpingo-oophorectomy should be offered to '\n"," 'women who are known LS mutation carriers and who have ﬁ nished child '\n"," 'bearing, \\n'\n"," 'optimally at age 40–45 years (conditional recommendation, low quality of '\n"," 'evidence). \\n'\n"," '  4. Screening for endometrial cancer and ovarian cancer should be offered '\n"," 'to women at risk for or affected with LS by endometrial biopsy and '\n"," 'transvaginal \\n'\n"," 'ultrasound annually, starting at age 30 to 35 years before undergoing '\n"," 'surgery or if surgery is deferred (conditional recommendation, very low '\n"," 'quality of \\n'\n"," 'evidence). \\n'\n"," '  5. Screening for gastric and duodenal cancer can be considered in '\n"," 'individuals at risk for or affected with LS by baseline '\n"," 'esophagogastroduodenoscopy \\n'\n"," '(EGD) with gastric biopsy at age 30–35 years, and treatment of  H. pylori '\n"," 'infection when found. Data for ongoing regular surveillance are limited, '\n"," 'but \\n'\n"," 'ongoing surveillance every 3–5 years may be considered if there is a family '\n"," 'history of gastric or duodenal cancer (conditional recommendation, very low')\n","\n","\n","Document 2: ACG 2015 Hereditary GI Cancers.pdf\n","('evidence). \\n'\n"," '  5. Screening for gastric and duodenal cancer can be considered in '\n"," 'individuals at risk for or affected with LS by baseline '\n"," 'esophagogastroduodenoscopy \\n'\n"," '(EGD) with gastric biopsy at age 30–35 years, and treatment of  H. pylori '\n"," 'infection when found. Data for ongoing regular surveillance are limited, '\n"," 'but \\n'\n"," 'ongoing surveillance every 3–5 years may be considered if there is a family '\n"," 'history of gastric or duodenal cancer (conditional recommendation, very '\n"," 'low \\n'\n"," 'quality of evidence). \\n'\n"," '  6. Screening beyond population-based recommendations for cancers of the '\n"," 'urinary tract, pancreas, prostate, and breast is not recommended unless '\n"," 'there \\n'\n"," 'is a family history of the speciﬁ c cancers (conditional recommendation, low '\n"," 'quality of evidence). \\n'\n"," '  7. Although data suggest that daily aspirin may decrease the risk of '\n"," 'colorectal and extracolonic cancer in LS, currently the evidence is not sufﬁ '\n"," 'ciently robust \\n'\n"," 'or mature to make a recommendation for its standard use (conditional '\n"," 'recommendation, moderate quality of evidence). \\n'\n"," ' Adenomatous polyposis syndromes \\n'\n"," '  Familial adenomatous polyposis (FAP)/MUTYH-associated polyposis '\n"," '(MAP)/attenuated polyposis \\n'\n"," '  8. In individuals at risk for or affected with the classic AP syndromes, '\n"," 'screening for colorectal cancer by annual colonoscopy or ﬂ exible '\n"," 'sigmoidoscopy \\n'\n"," 'should be performed, beginning at puberty. In families with attenuated '\n"," 'familial adenomatous polyposis (AFAP) or MAP, surveillance should be by \\n'\n"," 'colonoscopy (strong recommendation, moderate quality of evidence). \\n'\n"," '  9. Absolute indications for immediate colectomy in FAP, AFAP, and MAP '\n"," 'include: documented or suspected cancer or signiﬁ cant symptoms. Relative \\n'\n"," 'indications for surgery include the presence of multiple adenomas >6 mm, a '\n"," 'signiﬁ cant increase in adenoma number, and inability to adequately survey '\n"," 'the \\n'\n"," 'colon because of multiple diminutive polyps (strong recommendation, low '\n"," 'quality of evidence).')\n","\n","\n","Document 3: ACG 2015 Hereditary GI Cancers.pdf\n","('this malignancy in males and females with MMR gene mutation, \\n'\n"," 'respectively, and lack of familial clustering ( 61 ). Th e majority of \\n'\n"," 'gastric cancers in LS patients appear to be histologically classifi ed \\n'\n"," 'as intestinal type ( 61,62 ), and, consequently, potentially amenable \\n'\n"," 'to endoscopic surveillance. Th ere are no studies that have evalu-\\n'\n"," 'ated the eff ectiveness of screening and surveillance for gastric can-\\n'\n"," 'cer in LS patients.\\n'\n"," ' Th e lifetime risk for small bowel cancer ranges from 0.4 to \\n'\n"," '12.0% ( 31,34,53,54,63,64 ). Th e majority of small bowel cancers \\n'\n"," 'in a LS cohort were located in the duodenum or ileum ( 65 ) and \\n'\n"," 'within the reach of EGD and colonoscopy with dedicated ileal \\n'\n"," 'intubation. Studies of small bowel screening in LS patients are \\n'\n"," 'lacking. However, one screening investigation of 35 gene muta-\\n'\n"," 'tion carriers found that 2 had jejunal adenomas and 1 had a \\n'\n"," 'jejunal cancer ( 66 ) ( Table 8 ). Six additional patients had capsule \\n'\n"," 'endoscopy images of uncertain clinic relevance, prompting fur-\\n'\n"," 'ther invasive investigation in fi ve patients. A recent publication \\n'\n"," 'suggested that routine surveillance of the small bowel in LS was \\n'\n"," 'not cost effi  cient ( 55 ).\\n'\n"," '  Estimates of the lifetime risk of urinary tract cancer in LS range \\n'\n"," 'from 0.2 to 25%, depending on the study and which urinary tract \\n'\n"," ' Summary of evidence  \\n'\n"," '  EC is the second most common cancer occurring in LS. Esti-\\n'\n"," 'mates of the cumulative lifetime risk of EC in LS patients range \\n'\n"," 'from 15 to 71%, with variability depending on specifi c gene \\n'\n"," 'mutation ( 31,33,35–37,49 ) ( Table 7 ); reports of age at diagnosis \\n'\n"," 'of this malignancy are clearly a decade or more younger than \\n'\n"," 'sporadic EC but range from 48 to 54 years ( 31,33,35–37,49 )\\n'\n"," '( Table 7 ). Estimates of the cumulative lifetime risk of ovarian \\n'\n"," 'cancer in LS patients ranges from 3.4 to 22% ( 31,38,53–55 ) \\n'\n"," '( Table 7 ).\\n'\n"," ' Because of the worrisome cumulative risk of EC, several annual \\n'\n"," 'screening modalities have been proposed including pelvic exams,')\n","\n","\n","Document 4: ACG 2015 Hereditary GI Cancers.pdf\n","('Syngal  et al. \\n'\n"," 'The American Journal of GASTROENTEROLOGY \\n'\n"," 'VOLUME 110 | FEBRUARY 2015   www.amjgastro.com\\n'\n"," ' \\n'\n"," '224\\n'\n"," ' Table 1 .  Summary statements \\n'\n"," ' Standard for minimal cancer family history assessment in gastrointestinal '\n"," '(GI) practice \\n'\n"," '  A family history of cancer and premalignant GI conditions that provides '\n"," 'sufﬁ cient information to develop a preliminary determination of the risk of '\n"," 'a familial \\n'\n"," 'predisposition to cancer should be obtained for all patients being evaluated '\n"," 'in outpatient gastroenterology and endoscopy practices. \\n'\n"," '  Essential elements of a family history include presence and type of cancer '\n"," 'diagnoses in ﬁ rst- and second-degree relatives, and presence and (ideally) '\n"," 'type \\n'\n"," 'of polyps in ﬁ rst-degree relatives; age and lineage should be noted for '\n"," 'each diagnosis. \\n'\n"," ' Lynch syndrome (LS) \\n'\n"," ' All newly diagnosed colorectal cancers (CRCs) should be evaluated for '\n"," 'mismatch repair deﬁ ciency. \\n'\n"," '  Analysis may be done by immunohistochemical testing for the  '\n"," 'MLH1/MSH2/MSH6/PMS2 proteins and/or testing for microsatellite instability. '\n"," 'Tumors that \\n'\n"," 'demonstrate loss of  MLH1 should undergo BRAF testing or analysis for  MLH1 '\n"," 'promoter hypermethylation. \\n'\n"," '  Individuals who have a personal history of a tumor showing evidence of '\n"," 'mismatch repair deﬁ ciency (and no demonstrated BRAF mutation or '\n"," 'hypermethylation of \\n'\n"," ' MLH1) , a known family mutation associated with LS, or a risk of ≥5% chance '\n"," 'of LS based on risk prediction models should undergo genetic evaluation for '\n"," 'LS. \\n'\n"," '  Genetic testing of patients with suspected LS should include germline '\n"," 'mutation genetic testing for the  MLH1 ,  MSH2 ,  MSH6 ,  PMS2 , and/or  '\n"," 'EPCAM genes \\n'\n"," 'or the altered gene(s) indicated by immunohistochemical (IHC) testing. \\n'\n"," ' Adenomatous polyposis syndromes \\n'\n"," '  Familial adenomatous polyposis (FAP)/MUTYH-associated polyposis/attenuated '\n"," 'polyposis \\n'\n"," '   Individuals who have a personal history of >10 cumulative colorectal '\n"," 'adenomas, a family history of one of the adenomatous polyposis syndromes, or '\n"," 'a')\n","\n","\n"]}],"source":["for idx, document in enumerate(answer_from_source[\"source_documents\"]):\n","    print(f\"Document {idx + 1}:\", answer_from_source[\"source_documents\"][0].metadata[\"source\"].split(\"/\")[-1])\n","    pprint(document.page_content)\n","    print(\"\\n\")"]},{"cell_type":"markdown","metadata":{"id":"6J5ar_3qmwuv"},"source":["# LLM as Planning Agent"]},{"cell_type":"markdown","metadata":{"id":"jvh-wSH_iuuU"},"source":["In the second part of the tutorial, we will build an autonomous LLM-Agent capable of planning and execute complex tasks which requires multiple steps and multiple tools which extend agent's capability beyond LLM's.<br>\n","\n","We will create an Agent which aims to provide recommendations for patients with Chronic Kidney Diseases (CKD). We want it to search for patients information, perform calculation, search internal knowledge base and external knowledge base (internet) to find information needed.\n","\n","Let's first create the Custom Tools required for the agents to work with.\n","1. QASearchTool: Tool linked to our previously created RetrievalQAChain from last session.\n","2. MathTool: Tool to do math. LLM is known for it's poor capability of doing complex Math.\n","3. CSVSearchTool: Tool to perform query from table/SQL database to get patient's information.\n","4. SearchTool: Tool to search stuffs on the internet\n","5. GeneralKnowledgeTool: This is just an LLM Tool. MRKL agents tend to prefer using tools even in cases where it doesnt need to."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"D5zb0BYtdOZW"},"outputs":[],"source":["class QASearchTool(BaseTool):\n","    name: str = \"Docsearch QA Tool\"\n","    description: str = \"Use this tool to search for document and answer questions related to treatment of disease\"\n","    llm: BaseLanguageModel\n","    k: int = 4\n","    docstore: str = os.path.join(MAIN_DIR, \"vectorstore\")\n","\n","    @root_validator()\n","    def generate_qa_chain(cls, values):\n","        docsearch = FAISS.load_local(values[\"docstore\"], OpenAIEmbeddings())\n","        print(\"Document Store loaded successfully!!!\")\n","        values[\"chain\"] = RetrievalQAWithSourcesChain.from_chain_type(\n","            llm=values[\"llm\"],\n","            chain_type=\"stuff\",\n","            retriever=docsearch.as_retriever(search_kwargs={\"k\":values[\"k\"]}), # k is the number of documents to retrieve from\n","            return_source_documents=True,\n","            reduce_k_below_max_tokens=False, # If True, the chain will automatically select the number of documents to below max_tokens. Default to 3375 tokens.\n","            verbose=True\n","            )\n","        return values\n","\n","    def _run(self, query):\n","        return self.chain(query)\n","\n","    def _arun(self, query):\n","        return NotImplementedError\n","\n","class CustomMathChain(LLMMathChain):\n","    fallback_tool = PythonAstREPLTool()\n","    def _evaluate_expression(self, expression: str) -> str:\n","        try:\n","            local_dict = {\"pi\": math.pi, \"e\": math.e}\n","            output = numexpr.evaluate(\n","                    expression.strip(),\n","                    global_dict={},  # restrict access to globals\n","                    local_dict=local_dict,  # add common mathematical functions\n","                )\n","        except Exception as e:\n","            try: # Fall back to python shell\n","                output = self.fallback_tool(expression.strip())\n","            except Exception as e:\n","                raise ValueError(\n","                    f'LLMMathChain._evaluate(\"{expression}\") raised error: {e}.'\n","                    \" Please try again with a valid numerical expression\"\n","                )\n","\n","        # Remove any leading and trailing brackets from the output\n","        return re.sub(r\"^\\[|\\]$\", \"\", str(output))\n","\n","class CSVDatabaseSearchTool(BaseTool):\n","    name: str = \"CSVDatabase Search\"\n","    description: str = \"Use this tool to search for information in a CSV database\"\n","    llm: BaseLanguageModel\n","    data_path: str\n","\n","    @root_validator()\n","    def generate_agent_chain(cls, values):\n","        values[\"agent\"] = create_csv_agent(\n","            llm = values[\"llm\"],\n","            path = values[\"data_path\"],\n","            verbose = True,\n","            agent_type = AgentType.ZERO_SHOT_REACT_DESCRIPTION\n","        )\n","        return values\n","\n","    def _run(self, query):\n","        return self.agent(query)\n","\n","    def _arun(self, query):\n","        return NotImplementedError"]},{"cell_type":"markdown","metadata":{"id":"yPiNa0g1xBtZ"},"source":["## Define Agent Toolkits"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":683,"status":"ok","timestamp":1691337694641,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"oOBsFzF5y85V","outputId":"fba87689-8c91-4342-ec08-3c5bd88e4483"},"outputs":[{"name":"stdout","output_type":"stream","text":["Document Store loaded successfully!!!\n"]}],"source":["llm = ChatOpenAI(model_name=\"gpt-4\", temperature=0, max_tokens=256)\n","math_chain = CustomMathChain.from_llm(llm=llm)\n","llm_chain = LLMChain(prompt=ChatPromptTemplate.from_messages([HumanMessagePromptTemplate.from_template(\"Question: {question}\")]), llm=llm)\n","search = GoogleSerperAPIWrapper()\n","\n","tools = [QASearchTool(k=4, llm=llm, docstore=os.path.join(MAIN_DIR, \"vectorstore\"), description = \"Use this tool to search for document and answer questions related to treatment of disease\"),\n","         CSVDatabaseSearchTool(\n","              name = \"Patient Database Search\",\n","              description = \"Use this tool to search for patients information including demographics (Patient ID, age, gender, race) and lab results (Creatinine)\",\n","              llm=llm, data_path=os.path.join(MAIN_DIR, \"database\", \"patients_info.csv\")),\n","         Tool(name='Calculator', func=math_chain.run, description='Useful for when you need to answer questions about math.'),\n","         Tool(name='General Knowledge',func=llm_chain.run, description='Useful for general knowledge question.'),\n","         Tool(name=\"Search\", func=search.run, description=\"Useful for when you need to search for information online, especially math equations.\")\n","         ]"]},{"cell_type":"markdown","metadata":{"id":"47BuerE3xE5y"},"source":["## Define Agent Instructions"]},{"cell_type":"markdown","metadata":{"id":"DmrY6SY_lUfT"},"source":["Let's create a prompt for agent behaviour. We will use **ReAct** framework (Think-Act-Observe) for agent planning/reasoning and **MRKL** framework for agent choice of tools/actions.\n","\n","In Langchain, **MRKL** agent is implemented under ZeroShotAgent module"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":333,"status":"ok","timestamp":1691337699802,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"opqqL7TjxHHw","outputId":"d55e67ab-af24-4fa8-b34d-686e3f332208"},"outputs":[{"name":"stdout","output_type":"stream","text":["You are a helpful physician assistant giving recommendations on treatment for Chronic Kidney Disease (CKD).\n","However, you are terrible at remembering math equations and should always search it somewhere else.\n","\n","For each patient, calculate the eGFR score for the patient provided based on the latest formula.\n","Remember to convert creatinine measure to the unit specified in the equation. Look for additional information of this patient from database search tools if needed.\n","Finally, based on this patient eGFR and demographics (age, gender, race), classify CKD stage and return the recommendations of the suitable treatment for chronic kidney disease.\n","You have access to the following tools:\n","\n","Docsearch QA Tool: Use this tool to search for document and answer questions related to treatment of disease\n","Patient Database Search: Use this tool to search for patients information including demographics (Patient ID, age, gender, race) and lab results (Creatinine)\n","Calculator: Useful for when you need to answer questions about math.\n","General Knowledge: Useful for general knowledge question.\n","Search: Useful for when you need to search for information online, especially math equations.\n","\n","Use the following format:\n","\n","Question: the input question you must answer\n","Thought: you should always think about what to do\n","Action: the action to take, should be one of [Docsearch QA Tool, Patient Database Search, Calculator, General Knowledge, Search]\n","Action Input: the input to the action\n","Observation: the result of the action\n","... (this Thought/Action/Action Input/Observation can repeat N times)\n","Thought: I now know the final answer\n","Final Answer: the final answer to the original input question\n","\n","Begin!\n","\n","Patient Query: {input}\n","Thought: {agent_scratchpad}\n"]}],"source":["prefix = \"\"\"You are a helpful physician assistant giving recommendations on treatment for Chronic Kidney Disease (CKD).\n","However, you are terrible at remembering math equations and should always search it somewhere else.\n","\n","For each patient, calculate the eGFR score for the patient provided based on the latest formula.\n","Remember to convert creatinine measure to the unit specified in the equation. Look for additional information of this patient from database search tools if needed.\n","Finally, based on this patient eGFR and demographics (age, gender, race), classify CKD stage and return the recommendations of the suitable treatment for chronic kidney disease.\n","You have access to the following tools:\"\"\"\n","\n","suffix = \"\"\"Begin!\n","\n","Patient Query: {input}\n","Thought: {agent_scratchpad}\"\"\"\n","\n","AGENT_PROMPT = ZeroShotAgent.create_prompt(tools=tools, prefix=prefix, suffix=suffix)\n","print(AGENT_PROMPT.template)"]},{"cell_type":"markdown","metadata":{"id":"D62uNmpemDts"},"source":["- Let's Initialize our Agent. For tasks required long-term planning and reasoning, it's recommended to use bigger LLM models, as it has been shown that in general, models with more parameters perform much better in Chain-Of-Thought reasoning and Multi-Step Planning.<br><br>\n","- To prevent Agent getting stuck forever & burn tokens, set max_iterations to terminate Agent loop."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"eKxqLcp91hag"},"outputs":[],"source":["agent_llm_chain = LLMChain(llm=ChatOpenAI(model_name=\"gpt-4\", temperature=0, max_tokens=256), prompt=AGENT_PROMPT)\n","tool_names = [tool.name for tool in tools]\n","agent = ZeroShotAgent(llm_chain=agent_llm_chain, allowed_tools=tool_names)\n","\n","agent_args = {\"max_iterations\": 10, \"early_stopping_method\": \"generate\"}\n","agent_executor = AgentExecutor.from_agent_and_tools(agent=agent, tools=tools, verbose=True, **agent_args)"]},{"cell_type":"markdown","metadata":{"id":"yGub_WbFm9Rf"},"source":["Let's check if the Agent can choose the correct tools and perform the simple singular task"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":12203,"status":"ok","timestamp":1690965689838,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"zp8Nfkn2z_pG","outputId":"dfafadc5-23b9-4ad3-9255-9f1c02775ff0"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mThe patient's query is not related to CKD treatment but rather a general knowledge question about the Harry Potter series.\n","Action: General Knowledge\n","Action Input: How many episodes are there in Harry Potter series\u001b[0m\n","Observation: \u001b[36;1m\u001b[1;3mThe Harry Potter series is a set of seven books, not episodes. However, there are eight movies based on these books.\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI now know the final answer\n","Final Answer: The Harry Potter series consists of seven books. However, there are eight movies based on these books.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'input': 'How many episodes are there in Harry Potter series',\n"," 'output': 'The Harry Potter series consists of seven books. However, there are eight movies based on these books.'}"]},"execution_count":59,"metadata":{},"output_type":"execute_result"}],"source":["agent_executor(\"How many episodes are there in Harry Potter series\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":12522,"status":"ok","timestamp":1690965918501,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"3QtRAJGJ4FFR","outputId":"bd19f661-03c4-4edb-c546-6e17ea8aa17d"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mThe patient is asking for a mathematical calculation. I can use the calculator tool to solve this.\n","Action: Calculator\n","Action Input: 300*300*15+40\u001b[0m\n","Observation: \u001b[38;5;200m\u001b[1;3mAnswer: 1350040\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI now know the final answer.\n","Final Answer: The result of the calculation 300 square multiply by 15 plus 40 is 1350040.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'input': 'What is 300 square multiply by 15 plus 40?',\n"," 'output': 'The result of the calculation 300 square multiply by 15 plus 40 is 1350040.'}"]},"execution_count":65,"metadata":{},"output_type":"execute_result"}],"source":["agent_executor(\"What is 300 square multiply by 15 plus 40?\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":12887,"status":"ok","timestamp":1690965943874,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"kuGaaBNqCzlR","outputId":"cdd27d4e-8f2e-45ad-e93b-5f78feb58c1d"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mThis question is about the FIFA ranking in 2023, which is a general knowledge question. I need to use the search tool to find the latest information.\n","Action: Search\n","Action Input: Top 5 countries on Men's FIFA ranking in 2023\u001b[0m\n","Observation: \u001b[33;1m\u001b[1;3m1. Argentina ; 2. France ; 3. Brazil ; 4. England ; 5. Belgium ... FIFA World Rankings ; 3, Brazil (BRA), 1,828.27 ; 4, England (ENG), 1,797.39 ; 5, Belgium (BEL), 1,788.55 ; 6, Croatia (CRO), 1,742.55 ... FIFA Men's World Ranking List ; 53, Canada, 1431.64 ; 54, Slovakia, 1425.58 ; 55, Venezuela, 1406.1 ; 56, Finland, 1405.71 ... June 2023 FIFA World Rankings revealed: Spain, United States, Mexico, England, Argentina, Brazil... 5, Belgium (BEL), 1,788.55, 1,789, 5. 6, Croatia (CRO), 1,742.55, 1,743, 6. 7, Netherlands (NED), 1,731.23, 1,731, 7. 8, Italy (ITA), 1,726.58, 1,727, 8. The rankings were introduced in December 1992, and eight teams (Argentina, Belgium, Brazil, France, Germany, Italy, the Netherlands and Spain) have held the top ... 5, Belgium, 1788.55, -. 6, Croatia, 1742.55, -. 7, Netherlands, 1731.23, -. 8, Italy, 1726.58, -. 9, Portugal, 1718.25, -. Latest FIFA rankings June 2023 · 1. Argentina, CONMEBOL, - · 2. France, UEFA, - · 3. Brazil, CONEMBOL, - · 4. England, UEFA, 5th (+1) · 5. Belgium ... CONCACAF RANKING INDEX ; 1, Mexico, MEX ; 2, United States, USA ; 3, Canada, CAN ; 4, Costa Rica, CRC ... FIFA Rankings ; 3. Brazil. 1828.27 ; 4. England. 1797.39 ; 5. Belgium. 1788.55 ; 6. Croatia. 1742.55.\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI now know the top 5 countries on Men's FIFA ranking in 2023.\n","Final Answer: The top 5 countries on Men's FIFA ranking in 2023 are 1. Argentina, 2. France, 3. Brazil, 4. England, 5. Belgium.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'input': \"What is the top 5 countries on Men's FIFA ranking in 2023?\",\n"," 'output': \"The top 5 countries on Men's FIFA ranking in 2023 are 1. Argentina, 2. France, 3. Brazil, 4. England, 5. Belgium.\"}"]},"execution_count":67,"metadata":{},"output_type":"execute_result"}],"source":["agent_executor(\"What is the top 5 countries on Men's FIFA ranking in 2023?\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":25083,"status":"ok","timestamp":1690965992398,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"JrXK6i0E3yPW","outputId":"d5fe9cfe-a69c-4a49-cbee-55634e429825"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mFirst, I need to find out the recommended colonoscopy screening for Polyp patients with a family history of Colon Cancer Syndrome X.\n","Action: Docsearch QA Tool\n","Action Input: What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\u001b[0m\n","\n","\u001b[1m> Entering new RetrievalQAWithSourcesChain chain...\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","Observation: \u001b[36;1m\u001b[1;3m{'question': 'What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?', 'answer': 'For Polyp patients with a history of family members having Colon Cancer Syndrome X, the recommended screening is a colonoscopy every 3-5 years. This should begin 10 years before the age at diagnosis of the youngest affected relative.\\n', 'sources': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'source_documents': [Document(page_content='Family history\\nRecommended screening\\nLynch Syndrome\\nSee reference 34\\nFamily Colon Cancer Syndrome X\\nColonoscopy every 3-5 years beginning 10 years before the age at\\ndiagnosis of the youngest affected relative\\nColorectal cancer or an advanced adenoma in two\\nﬁrst-degree relatives diagnosed at any age\\nOR colorectal cancer or an advanced adenoma\\nin a single ﬁrst-degree relative at age < 60 years\\nColonoscopy every 5 years beginning 10 years before the age\\nat diagnosis of the youngest affect interval or age 40,\\nwhichever is earlier; for those with a single ﬁrst-degree\\nrelative with colorectal cancer in whom no signiﬁcant\\nneoplasia appears by age 60 years, physicians can offer\\nexpanding the interval between colonoscopies\\nColorectal cancer or an advanced adenoma in a\\nsingle ﬁrst-degree relative diagnosed at\\nage � 60 years\\nBegin screening at age 40 years; tests and intervals are as\\nper the average-risk screening recommendations (Table 4)\\n10\\nRex et al\\nGastroenterology Vol. -, No. -\\nAGA SECTION', metadata={'source': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'file_path': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'page': 9, 'total_pages': 17, 'format': 'PDF 1.7', 'title': 'Colorectal Cancer Screening: Recommendations for Physicians and Patients From the U.S. Multi-Society Task Force on Colorectal Cancer', 'author': 'Douglas K. Rex', 'subject': 'Gastroenterology, Corrected proof. doi:10.1053/j.gastro.2017.05.013', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170601213054+05'30'\", 'modDate': \"D:20170601213137+05'30'\", 'trapped': ''}), Document(page_content='of second-tier tests to patients who decline both colonos-\\ncopy and FIT), or a risk-stratiﬁed approach (colonoscopy is\\noffered to patients with a higher pretest probability of\\nneoplasia, and FIT is used in persons with a lower pretest\\nprobability of neoplasia).\\nPersons with a history of CRC or a documented\\nadvanced adenoma in a ﬁrst-degree relative age <60 years\\nor 2 ﬁrst-degree relatives with these ﬁndings at any age are\\nrecommended to undergo screening by colonoscopy every\\n5 years, beginning 10 years before the age at diagnosis of\\nthe youngest affected relative, or at age 40, whichever is\\nearlier. Persons with a single ﬁrst-degree relative diag-\\nnosed at �60 years with CRC or an advanced adenoma can\\nbe offered average-risk screening options beginning at age\\n40 years.\\nThe incidence of CRC is rising in persons under age 50.\\nPatients under age 50 with bleeding symptoms consistent\\nwith a colorectal source should be aggressively evaluated\\nand treated. We suggest that screening begin at age 45 in\\nAfrican Americans. Discontinuation of screening should be\\nconsidered when patients who are up to date with screening\\nand have had negative screening tests, particularly colo-\\nnoscopy, reach age 75 years or when life expectancy is <10\\nyears. Persons without prior screening should be consid-\\nered for screening up to age 85 years, depending on co-\\nmorbid conditions and life expectancy.\\nReferences\\n1.Levin B, Lieberman DA, McFarland B, et al. Screening\\nand surveillance for the early detection of colorectal\\ncancer\\nand\\nadenomatous\\npolyps.\\nGastroenterology\\n2008;134:1570–1595.\\n2.Kahi CJ, Boland CR, Dominitz JA, et al. Colonoscopy\\nsurveillance after colorectal cancer resection: recom-\\nmendations of the US multi-society task force on colo-\\nrectal cancer. Gastrointest Endosc 2016;83:489–498.\\n3.Lieberman DA, Rex DK, Winawer SJ, et al. Guidelines for\\ncolonoscopy surveillance after screening and poly-\\npectomy. Gastroenterology 2012;143:844–857.', metadata={'source': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'file_path': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'page': 11, 'total_pages': 17, 'format': 'PDF 1.7', 'title': 'Colorectal Cancer Screening: Recommendations for Physicians and Patients From the U.S. Multi-Society Task Force on Colorectal Cancer', 'author': 'Douglas K. Rex', 'subject': 'Gastroenterology, Corrected proof. doi:10.1053/j.gastro.2017.05.013', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170601213054+05'30'\", 'modDate': \"D:20170601213137+05'30'\", 'trapped': ''}), Document(page_content='inherited syndrome caused by mutations in 1 or more\\nmismatch repair genes. Patients in families that meet the\\nclinical criteria for hereditary nonpolyposis CRC but have\\nmicrosatellite-stable CRCs have family colon cancer syn-\\ndrome X, which has not been genetically deﬁned.122 Persons\\nin families with syndrome X should undergo colonoscopy at\\nleast every 3 to 5 years, beginning 10 years before the age at\\ndiagnosis of the youngest affected relative.\\nA family history of CRC in a ﬁrst-degree relative in-\\ncreases the risk of CRC regardless of the age at diagnosis of\\nTable 4.Multi-Society Task Force Ranking of Current\\nColorectal Cancer Screening Tests\\nTier 1\\nColonoscopy every 10 years\\nAnnual fecal immunochemical test\\nTier 2\\nCT colonography every 5 years\\nFIT–fecal DNA every 3 years\\nFlexible sigmoidoscopy every 10 years (or every 5 years)\\nTier 3\\nCapsule colonoscopy every 5 years\\nAvailable tests not currently recommended\\nSeptin 9\\n- 2017\\nMSTF Recommendations for CRC Screening\\n9\\nAGA SECTION', metadata={'source': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'file_path': '/content/iaim_langchain/raw_data/MSTF 2017 CRC Screening.pdf', 'page': 8, 'total_pages': 17, 'format': 'PDF 1.7', 'title': 'Colorectal Cancer Screening: Recommendations for Physicians and Patients From the U.S. Multi-Society Task Force on Colorectal Cancer', 'author': 'Douglas K. Rex', 'subject': 'Gastroenterology, Corrected proof. doi:10.1053/j.gastro.2017.05.013', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170601213054+05'30'\", 'modDate': \"D:20170601213137+05'30'\", 'trapped': ''}), Document(page_content='Hereditary GI Cancer Syndromes\\n© 2015 by the American College of Gastroenterology \\nThe American Journal of GASTROENTEROLOGY\\n \\n229\\nsurveillance of MMR germline mutation-positive patients, consid-\\neration should be given to annual colonoscopy, as several studies \\nhave demonstrated CRC development with surveillance intervals \\nthat are between 1 and 2 years ( 29,47 ). In carriers of deleterious \\n MSH6 and  PMS2 mutations, the risk of CRC is less and age of diag-\\nnosis later ( 37,49 ) than in patients with  MLH1 and MSH2 muta-\\ntions. In these aff ected individuals, consideration could be given \\nto starting surveillance at age 25–30 in  MSH6 and  PMS2 carriers \\n( 24 ), unless an early-onset cancer exists in a given family; however, \\n( 44 ) reported the results of a prospective observational study of \\ncolonoscopy surveillance of members in hereditary nonpolyposis \\ncolorectal cancer or LS families, revealing a 72% decrease in mor-\\ntality from CRC in those undergoing screening. In several stud-\\nies ( 45–48 ), more frequent colonoscopy screening (≤2 years) was \\nassociated with an earlier stage of CRC at diagnosis and less CRC \\nthan less frequent colonoscopy. At least every 2-year colonoscopic \\nsurveillance of LS patients is supported by the data above and the \\nrapid adenoma–carcinoma sequence reported in these patients. In \\n Table 4 .  Amsterdam criteria, revised Bethesda guidelines, and colorectal cancer risk assessment tool \\n Amsterdam criteria I ( 24 ) \\n At least three relatives with colorectal cancer (CRC); all of the following criteria should be present: \\n  One should be a ﬁ rst-degree relative of the other two; \\n  At least two successive generations must be affected; \\n  At least one of the relatives with CRC must have received the diagnosis before the age of 50 years; \\n  Familial adenomatous polyposis should be excluded; \\n  Tumors should be veriﬁ ed by pathologic examination. \\n Amsterdam criteria II ( 24 )', metadata={'source': '/content/iaim_langchain/raw_data/ACG 2015 Hereditary GI Cancers.pdf', 'file_path': '/content/iaim_langchain/raw_data/ACG 2015 Hereditary GI Cancers.pdf', 'page': 6, 'total_pages': 40, 'format': 'PDF 1.6', 'title': 'ACG Clinical Guideline: Genetic Testing and Management of Hereditary Gastrointestinal Cancer Syndromes', 'author': 'Sapna Syngal', 'subject': 'The American Journal of Gastroenterology 110, 223 (2015). doi:10.1038/ajg.2014.435', 'keywords': 'LWW', 'creator': '', 'producer': 'PD4ML 4.0.15fx1', 'creationDate': 'D:20150124121553Z', 'modDate': \"D:20230410134914Z00'00'\", 'trapped': ''})]}\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI now know the recommended colonoscopy screening for Polyp patients with a family history of Colon Cancer Syndrome X.\n","Final Answer: For Polyp patients with a history of family members having Colon Cancer Syndrome X, the recommended screening is a colonoscopy every 3-5 years. This should begin 10 years before the age at diagnosis of the youngest affected relative.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'input': 'What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?',\n"," 'output': 'For Polyp patients with a history of family members having Colon Cancer Syndrome X, the recommended screening is a colonoscopy every 3-5 years. This should begin 10 years before the age at diagnosis of the youngest affected relative.'}"]},"execution_count":69,"metadata":{},"output_type":"execute_result"}],"source":["agent_executor(\"What is the recommended colonoscopy screening for Polyp patients with history of family members having Colon Cancer Syndrome X?\")"]},{"cell_type":"markdown","metadata":{"id":"gpwm4_GY5nNU"},"source":["Awesome. It seems like Agent can choose the correct tool. <br>\n","Now let's allow it to work on the actual task which requires multiple steps."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":109357,"status":"ok","timestamp":1691338349192,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"tkO6iMKOOYEm","outputId":"7f9048e1-abea-4c34-9290-0cbf9446657f"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mFirst, I need to find out the patient's demographic information and lab results.\n","Action: Patient Database Search\n","Action Input: 17238471\u001b[0m\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mThought: The question is asking for the row of the dataframe where the Patient ID is 17238471.\n","Action: I will use the pandas dataframe method `df.loc[]` to locate the row where the Patient ID is 17238471.\n","Action Input: `df.loc[df['Patient ID'] == 17238471]`\u001b[0m\n","Observation: I will use the pandas dataframe method `df.loc[]` to locate the row where the Patient ID is 17238471. is not a valid tool, try another one.\n","Thought:\u001b[32;1m\u001b[1;3mI need to use the python_repl_ast tool to execute the python command.\n","Action: python_repl_ast\n","Action Input: df.loc[df['Patient ID'] == 17238471]\u001b[0m\n","Observation: \u001b[36;1m\u001b[1;3m   No  Name  Patient ID      NRIC       DOB Gender Ethnicity  \\\n","0   1  Andy    17238471  G1234567  1/2/1950   Male     Black   \n","\n","   Creatinine Serum    Unit  \n","0               111  umol/L  \u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI now know the final answer\n","Final Answer: The row in the dataframe where the Patient ID is 17238471 is:\n","\n","|    |   No | Name    |   Patient ID | NRIC     | DOB       | Gender   | Ethnicity   |   Creatinine Serum | Unit   |\n","|---:|-----:|:--------|-------------:|:---------|:----------|:---------|:------------|-------------------:|:-------|\n","|  0 |    1 | Andy    |     17238471 | G1234567 | 1/2/1950  | Male     | Black       |                111 | umol/L |\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","Observation: \u001b[33;1m\u001b[1;3m{'input': '17238471', 'output': 'The row in the dataframe where the Patient ID is 17238471 is:\\n\\n|    |   No | Name    |   Patient ID | NRIC     | DOB       | Gender   | Ethnicity   |   Creatinine Serum | Unit   |\\n|---:|-----:|:--------|-------------:|:---------|:----------|:---------|:------------|-------------------:|:-------|\\n|  0 |    1 | Andy    |     17238471 | G1234567 | 1/2/1950  | Male     | Black       |                111 | umol/L |'}\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe patient is a 72-year-old black male with a serum creatinine level of 111 umol/L. Now, I need to calculate the eGFR using the CKD-EPI Creatinine Equation (2009). However, I don't remember the equation, so I need to search for it.\n","Action: Search\n","Action Input: CKD-EPI Creatinine Equation (2009)\u001b[0m\n","Observation: \u001b[33;1m\u001b[1;3mThe CKD-EPI equation, expressed as a single equation, is GFR = 141 × min(Scr/κ, 1)α × max(Scr/κ, 1)-1.209 × 0.993Age × 1.018 [if female] _ 1.159 [if black], where Scr is serum creatinine, κ is 0.7 for females and 0.9 for males, α is -0.329 for females and -0.411 for males, min indicates the minimum of Scr/κor 1, and ...\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mNow I have the CKD-EPI Creatinine Equation (2009). I need to convert the creatinine measure from umol/L to mg/dL, as the equation uses mg/dL. Then, I can calculate the eGFR.\n","Action: Calculator\n","Action Input: Convert 111 umol/L creatinine to mg/dL\u001b[0m\n","Observation: \u001b[38;5;200m\u001b[1;3mAnswer: 1.2543\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe creatinine level of 111 umol/L is equivalent to 1.2543 mg/dL. Now, I can use the CKD-EPI Creatinine Equation (2009) to calculate the eGFR.\n","Action: Calculator\n","Action Input: Calculate eGFR using CKD-EPI Creatinine Equation (2009) for a 72-year-old black male with a serum creatinine level of 1.2543 mg/dL\u001b[0m\n","Observation: \u001b[38;5;200m\u001b[1;3mAnswer: 65.97210092189849\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe eGFR for the patient is approximately 66 mL/min/1.73m^2. Now, I need to classify the CKD stage based on this eGFR.\n","Action: General Knowledge\n","Action Input: CKD stages based on eGFR\u001b[0m\n","Observation: \u001b[36;1m\u001b[1;3mChronic Kidney Disease (CKD) is classified into five stages based on the estimated Glomerular Filtration Rate (eGFR), which measures how well the kidneys are filtering wastes from the blood. Here are the stages:\n","\n","1. Stage 1: Kidney damage with normal or high eGFR (≥90 mL/min/1.73 m²). Kidney function is normal or high, but evidence of kidney damage is present.\n","\n","2. Stage 2: Kidney damage with mildly decreased eGFR (60-89 mL/min/1.73 m²). Kidney function is slightly lower than normal, and signs of kidney disease are present.\n","\n","3. Stage 3: Moderately decreased eGFR (30-59 mL/min/1.73 m²). This stage is divided into two: a mild-to-moderate decrease in eGFR (Stage 3a: eGFR 45-59 mL/min/1.73 m²) and a moderate-to-severe decrease in eGFR (Stage 3b: eGFR 30-44 mL/min/1.73 m²).\n","\n","4. Stage 4: Severely decreased eGFR (15-29 mL/min/1\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe patient's eGFR of 66 mL/min/1.73m^2 places him in Stage 2 of CKD, which is characterized by kidney damage with mildly decreased eGFR. Now, I need to search for the recommended treatment for Stage 2 CKD.\n","Action: Docsearch QA Tool\n","Action Input: What is the recommended treatment for Stage 2 Chronic Kidney Disease?\u001b[0m\n","\n","\u001b[1m> Entering new RetrievalQAWithSourcesChain chain...\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","Observation: \u001b[36;1m\u001b[1;3m{'question': 'What is the recommended treatment for Stage 2 Chronic Kidney Disease?', 'answer': 'The recommended treatment for Stage 2 Chronic Kidney Disease includes interventions to slow the progression of the disease such as strict glucose control in diabetes, strict blood pressure control, and angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade. It is also recommended to avoid potential nephrotoxins like nonsteroidal anti-inflammatory drugs and to monitor for complications of CKD such as hyperkalemia, metabolic acidosis, hyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia. \\n', 'sources': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf, /content/iaim_langchain/raw_data/nihms-1069237.pdf', 'source_documents': [Document(page_content='Clinical Practice\\nGuidelines\\nFor Chronic\\nKidney Disease:\\nEvaluation,\\nClassification and\\nStratification', metadata={'source': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'file_path': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'page': 1, 'total_pages': 356, 'format': 'PDF 1.6', 'title': '5349$$CHFM', 'author': 'mac20', 'subject': '', 'keywords': '', 'creator': 'SYSTEM400 Rev 16.02', 'producer': 'Acrobat Distiller 4.05 for Windows, Powered by PDF Polisher Pro  5.01 420', 'creationDate': \"D:20090310184858-04'00'\", 'modDate': \"D:20090505160504-04'00'\", 'trapped': ''}), Document(page_content='KDIGO 2012 Clinical Practice Guideline\\nfor the Evaluation and Management of\\nChronic Kidney Disease\\nKDIGO gratefully acknowledges the following consortium of sponsors that make our initiatives possible: Abbott, Amgen,\\nBayer Schering Pharma, Belo Foundation, Bristol-Myers Squibb, Chugai Pharmaceutical, Coca-Cola Company, Dole Food\\nCompany, Fresenius Medical Care, Genzyme, Hoffmann-LaRoche, JC Penney, Kyowa Hakko Kirin, NATCO—The\\nOrganization for Transplant Professionals, NKF-Board of Directors, Novartis, Pharmacosmos, PUMC Pharmaceutical,\\nRobert and Jane Cizik Foundation, Shire, Takeda Pharmaceutical, Transwestern Commercial Services, Vifor Pharma,\\nand Wyeth.\\nSponsorship Statement: KDIGO is supported by a consortium of sponsors and no funding is accepted for the development\\nof specific guidelines.', metadata={'source': '/content/iaim_langchain/raw_data/nihms639366.pdf', 'file_path': '/content/iaim_langchain/raw_data/nihms639366.pdf', 'page': 1, 'total_pages': 163, 'format': 'PDF 1.6', 'title': '3b2b_npg_kisup_3_1_fmiii 3..3', 'author': '', 'subject': '', 'keywords': '', 'creator': '3B2 Total Publishing System 8.07f/W', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20121207113718+05'30'\", 'modDate': \"D:20121227121547-06'00'\", 'trapped': ''}), Document(page_content='Chronic Kidney Disease Diagnosis and Management:\\nA Review\\nTeresa K. Chen, MD, MHS, Daphne H. Knicely, MD, Morgan E. Grams, MD, PhD\\nDivision of Nephrology, Department of Medicine, Johns Hopkins University School of Medicine, \\nBaltimore, Maryland (Chen, Knicely, Grams); Welch Center for Prevention, Epidemiology, and \\nClinical Research, Johns Hopkins Medical Institutions, Baltimore, Maryland (Chen, Grams).\\nAbstract\\nIMPORTANCE—Chronic kidney disease (CKD) is the 16th leading cause of years of life lost \\nworldwide. Appropriate screening, diagnosis, and management by primary care clinicians are \\nnecessary to prevent adverse CKD-associated outcomes, including cardiovascular disease, end-\\nstage kidney disease, and death.\\nOBSERVATIONS—Defined as a persistent abnormality in kidney structure or function (eg, \\nglomerular filtration rate [GFR] <60 mL/min/1.73 m2 or albuminuria ≥30 mg per 24 hours) for \\nmore than 3 months, CKD affects 8% to 16% of the population worldwide. In developed countries, \\nCKD is most commonly attributed to diabetes and hypertension. However, less than 5% of patients \\nwith early CKD report awareness of their disease. Among individuals diagnosed as having CKD, \\nstaging and new risk assessment tools that incorporate GFR and albuminuria can help guide \\ntreatment, monitoring, and referral strategies. Optimal management of CKD includes \\ncardiovascular risk reduction (eg, statins and blood pressure management), treatment of \\nalbuminuria (eg, angiotensin-converting enzyme inhibitors or angiotensin II receptor blockers), \\navoidance of potential nephrotoxins (eg, nonsteroidal anti-inflammatory drugs), and adjustments \\nto drug dosing (eg, many antibiotics and oral hypoglycemic agents). Patients also require \\nmonitoring for complications of CKD, such as hyperkalemia, metabolic acidosis, \\nhyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia. Those at', metadata={'source': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'file_path': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'page': 0, 'total_pages': 22, 'format': 'PDF 1.5', 'title': 'Chronic Kidney Disease Diagnosis and Management', 'author': 'Teresa K. Chen, Daphne H. Knicely, Morgan E. Grams', 'subject': '', 'keywords': '', 'creator': 'AH XSL Formatter V6.3 R1 for Linux64 : 6.3.1.23495 (2016/02/03 14:39JST)', 'producer': 'Antenna House PDF Output Library 6.3.752 (Linux64)', 'creationDate': \"D:20200212183609-05'00'\", 'modDate': \"D:20200212183609-05'00'\", 'trapped': ''}), Document(page_content='• Interventions to slow the progression of kidney disease should be considered in all\\npatients with chronic kidney disease.\\n• Interventions that have been proven to be effective include:\\n(1) Strict glucose control in diabetes;\\n(2) Strict blood pressure control;\\n(3) Angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade.\\n• Interventions that have been studied, but the results of which are inconclusive, include:\\n(1) Dietary protein restriction;\\n(2) Lipid-lowering therapy;\\n(3) Partial correction of anemia.\\n• Attempts should be made to prevent and correct acute decline in GFR. Frequent causes\\nof acute decline in GFR include:\\n• Volume depletion;\\n• Intravenous radiographic contrast;\\n• Selected antimicrobial agents (for example, aminoglycosides and amphotericin B);\\n• Nonsteroidal anti-inflammatory agents; including cyclo-oxygenase type 2 inhibitors;\\n• Angiotensin-converting enzyme inhibition and angiotensin-2 receptor blockers;\\n• Cyclosporine and tacrolimus;\\n• Obstruction of the urinary tract.\\n• Measurements of serum creatinine for estimation of GFR should be obtained at least\\nyearly in patients with chronic kidney disease and more often in patients with:\\n• GFR �60 mL/min/1.73 m2;\\n• Fast GFR decline in the past (�4 mL/min/1.73 m2 per year);\\n• Risk factors for faster progression;\\n• Ongoing treatment to slow progression;\\n• Exposure to risk factors for acute GFR decline.\\nGUIDELINE 14.\\nAssociation of Chronic Kidney Disease With Diabetic\\nComplications (p. 230)\\nThe risk of cardiovascular disease, retinopathy, and other diabetic complications is higher in\\npatients with diabetic kidney disease than in diabetic patients without kidney disease.\\n• Prevention, detection, evaluation, and treatment of diabetic complications in patients\\nwith chronic kidney disease should follow published guidelines and position statements.\\n• Guidelines regarding angiotensin-converting enzyme inhibitors or angiotensin-receptor', metadata={'source': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'file_path': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'page': 46, 'total_pages': 356, 'format': 'PDF 1.6', 'title': '5349$$CHFM', 'author': 'mac20', 'subject': '', 'keywords': '', 'creator': 'SYSTEM400 Rev 16.02', 'producer': 'Acrobat Distiller 4.05 for Windows, Powered by PDF Polisher Pro  5.01 420', 'creationDate': \"D:20090310184858-04'00'\", 'modDate': \"D:20090505160504-04'00'\", 'trapped': ''})]}\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI now know the recommended treatment for Stage 2 Chronic Kidney Disease.\n","Final Answer: The recommended treatment for Andy, who is at Stage 2 Chronic Kidney Disease, includes interventions to slow the progression of the disease such as strict glucose control in diabetes, strict blood pressure control, and angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade. It is also recommended to avoid potential nephrotoxins like nonsteroidal anti-inflammatory drugs and to monitor for complications of CKD such as hyperkalemia, metabolic acidosis, hyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'input': 'Recommend Treatment for patient Andy. Patient ID is 17238471',\n"," 'output': 'The recommended treatment for Andy, who is at Stage 2 Chronic Kidney Disease, includes interventions to slow the progression of the disease such as strict glucose control in diabetes, strict blood pressure control, and angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade. It is also recommended to avoid potential nephrotoxins like nonsteroidal anti-inflammatory drugs and to monitor for complications of CKD such as hyperkalemia, metabolic acidosis, hyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia.'}"]},"execution_count":50,"metadata":{},"output_type":"execute_result"}],"source":["agent_executor(\"Recommend Treatment for patient Andy. Patient ID is 17238471\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":107731,"status":"ok","timestamp":1691254403449,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"cmclvJIhB7aK","outputId":"bf108fc0-1403-439a-d975-5c7fd482b478"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mI need to find out more about Bobby's health status, especially his kidney function. \n","Action: Patient Database Search\n","Action Input: Patient ID 12374872\u001b[0m\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mThought: I need to find the row in the dataframe where the Patient ID is 12374872.\n","Action: python_repl_ast\n","Action Input: df[df['Patient ID'] == 12374872]\u001b[0m\n","Observation: \u001b[36;1m\u001b[1;3m   No   Name  Patient ID      NRIC       DOB Gender Ethnicity  \\\n","1   2  Bobby    12374872  S2123871  8/3/1944   Male     Black   \n","\n","   Creatinine Serum    Unit  \n","1               120  umol/L  \u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI now know the final answer\n","Final Answer: The patient with ID 12374872 is Bobby, a male of Black ethnicity born on 8/3/1944. His Creatinine Serum level is 120 umol/L.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","Observation: \u001b[33;1m\u001b[1;3m{'input': 'Patient ID 12374872', 'output': 'The patient with ID 12374872 is Bobby, a male of Black ethnicity born on 8/3/1944. His Creatinine Serum level is 120 umol/L.'}\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI need to calculate Bobby's estimated Glomerular Filtration Rate (eGFR) to determine his CKD stage. The eGFR is calculated using the patient's age, gender, race, and serum creatinine level. Since I don't remember the formula, I need to search for it.\n","Action: Search\n","Action Input: eGFR calculation formula\u001b[0m\n","Observation: \u001b[33;1m\u001b[1;3mThe National Kidney Foundation recommends using the CKD-EPI Creatinine Equation (2021) to estimate GFR. Estimates glomerular filtration rate based on creatinine and patient characteristics. This calculator includes inputs based on race, which may or may not ... The CKD-EPI Equation ; White or other, Female, > 0.7, GFR = 144 × (Scr/0.7)-1.209 × (0.993) ; White or other, Male, ≤ 0.9, GFR = 141 × (Scr/0.9)-0.411 × (0.993) ... Determine your GFR value using the GFR Calculator to help measure kidney function. Your Glomerular Filtration Rate value determines your stage of kidney ... The CKD-EPI equation, expressed as a single equation, is: GFR = 141 * min(Scr/κ,1)α * max(Scr/κ, 1)-1.209 ... eGFR Calculator. Serum Creatinine: mg/dL. μmol/L. Serum Cystatin C (optional):. mg/L. Age: Years. Sex: Male. Female. Race (optional):. Non black. 3A 3B: 45-59; 30-44. GFR is estimated by an equation developed by the Chronic Kidney Disease Epidemiology Collaboration. · GFR = 141 * min(Serum_creatinine/kappa, 1)alpha * max( ... Race: White or Other (1). Sex: Female Male. Estimating GFR begins with a simple serum creatinine blood test. If you're using our GFR calculator, your eGFR is calculated using the results of your blood ... STAGE 1: Kidney damage with normal kidney function. STAGE 4: Severe loss of kidney function. STAGE 2: Mild loss of kidney function. STAGE 3A: Mild to moderate loss of kidney function. eGFR stands for estimated Glomerular Filtration Rate. eGFR is calculated from blood test results using a formula that looks at the level of creatinine (a ...\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mNow that I have the formula, I can calculate Bobby's eGFR. The formula for a black male is: GFR = 141 * min(Scr/κ,1)α * max(Scr/κ, 1)-1.209 * 0.993^Age * 1.159. Where Scr is serum creatinine, κ is 0.9 for males, α is -0.411, min indicates the minimum of Scr/κ or 1, and max indicates the maximum of Scr/κ or 1. Bobby's age is 77, and his serum creatinine level is 120 umol/L, which needs to be converted to mg/dL.\n","Action: Search\n","Action Input: Convert 120 umol/L creatinine to mg/dL\u001b[0m\n","Observation: \u001b[33;1m\u001b[1;3mUNIT CONVERSION PANEL · Result · nl values: Creatinine (S)= 0.5 - 1.5 mg/dL (44.21 - 132.63 umol/L) CF: 88.42 · CF= conversion factor. Multiply C Units -first ... Creatinine Unit Conversion · Result : · Creatinine Normal Values (S): 0.5 - 1.5 mg/dL (44.21 - 132.63 umol/L) CF: 88.42 · Molecular Weight: 113.1 · CF= conversion ... Creatinine conversion to mmol/l, µmol/l, mg/dl, mg/100ml, mg%, mg/l, µg/ml. Online calculator. Reference Intervals for serum, plasma, urine. Creatinine unit converter calculator - Calculators Convert creatinine concentration from mg/dL to mmol/L and μmol/L. Calculate it! The formula used to convert µmol/L to Milligram per Deciliter is 1 Micromole per Liter = 0.0113096584483149 Milligram per Deciliter. Measurement is one of the ... Creatinine mg/dL values to µmol/L converter. Formula to convert mg/dL values to µmol/L Creatinine , is 1 mg/dL is equivalent to 88.4 µmol/L . Missing: 120 | Show results with:120. TEST. UNITS CONVERSION. Acetaminophen mg/L x 6.62 = umol/L. Adrenaline nmol/L x 0.183 = ug/L. Adrenocorticotrophic Hormone pmol/L x 4.55 = pg/mL. Albumin. Serum creatinine : eGFR conversion table for women ... creatinine : eGFR conversion table for men. Women. Age, years. Serum creatinine. µmol/L ... 120-129. Creatinine Clearance = [[140 - age(yr)]*weight(kg)]/[72*serum Cr(mg/dL)] (multiply by 0.85 for women). The results you receive from this tool are for ... Average creatinine levels are 0.7–1.2 milligrams per deciliter (mg/dL) for males and 0.5–1.0 for females. High levels may indicate a problem with the ...\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe conversion factor from µmol/L to mg/dL for creatinine is 0.0113096584483149. I can use this to convert Bobby's serum creatinine level to mg/dL.\n","Action: Calculator\n","Action Input: 120 * 0.0113096584483149\u001b[0m\n","Observation: \u001b[38;5;200m\u001b[1;3mAnswer: 1.357159013797788\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mBobby's serum creatinine level in mg/dL is approximately 1.36. Now I can use this value, along with his age, to calculate his eGFR using the CKD-EPI equation.\n","Action: Calculator\n","Action Input: 141 * min(1.36/0.9, 1)**-0.411 * max(1.36/0.9, 1)**-1.209 * 0.993**77 * 1.159\u001b[0m\n","Observation: \u001b[38;5;200m\u001b[1;3mAnswer: 57.75975420742635\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mBobby's eGFR is approximately 57.76 mL/min/1.73m^2. Now I need to classify his CKD stage based on this eGFR value.\n","Action: General Knowledge\n","Action Input: CKD stages based on eGFR\u001b[0m\n","Observation: \u001b[36;1m\u001b[1;3mChronic Kidney Disease (CKD) is classified into five stages based on the estimated Glomerular Filtration Rate (eGFR), which measures how well the kidneys are filtering wastes from the blood. Here are the stages:\n","\n","1. Stage 1: Kidney damage with normal or high eGFR (>90 mL/min per 1.73 m2). Kidney function is normal or high, but evidence of kidney damage is present.\n","\n","2. Stage 2: Kidney damage with mildly decreased eGFR (60-89 mL/min per 1.73 m2). Kidney function is slightly lower than normal, and signs of kidney disease are present.\n","\n","3. Stage 3: Moderately decreased eGFR (30-59 mL/min per 1.73 m2). This stage is divided into two: a mild-to-moderate decrease in eGFR (Stage 3a: eGFR 45-59 mL/min per 1.73 m2) and a moderate-to-severe decrease in eGFR (Stage 3b: eGFR 30-44 mL/min per 1.73 m2).\n","\n","4. Stage 4: Severely decreased eGFR (15-29 mL\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mBobby's eGFR of approximately 57.76 mL/min/1.73m^2 places him in Stage 3a of Chronic Kidney Disease. Now I need to find the suitable treatment recommendations for this stage of CKD.\n","Action: Docsearch QA Tool\n","Action Input: What is the recommended treatment for Stage 3a Chronic Kidney Disease?\u001b[0m\n","\n","\u001b[1m> Entering new RetrievalQAWithSourcesChain chain...\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","Observation: \u001b[36;1m\u001b[1;3m{'question': 'What is the recommended treatment for Stage 3a Chronic Kidney Disease?', 'answer': 'The recommended treatment for Stage 3a Chronic Kidney Disease includes management of hypertension and diabetes. For hypertension, blockade of the renin-angiotensin-aldosterone system with either an angiotensin-converting enzyme inhibitor (ACE-I) or an angiotensin II receptor blocker (ARB) is recommended. Dual therapy with an ACE-I and an ARB is generally avoided due to associated risks. For diabetes, glycemic control may delay progression of CKD, with most guidelines recommending a goal hemoglobin A1c of ~ 7.0%. Dose adjustments in oral hypoglycemic agents may be necessary.\\n', 'sources': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'source_documents': [Document(page_content='Clinical Practice\\nGuidelines\\nFor Chronic\\nKidney Disease:\\nEvaluation,\\nClassification and\\nStratification', metadata={'source': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'file_path': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'page': 1, 'total_pages': 356, 'format': 'PDF 1.6', 'title': '5349$$CHFM', 'author': 'mac20', 'subject': '', 'keywords': '', 'creator': 'SYSTEM400 Rev 16.02', 'producer': 'Acrobat Distiller 4.05 for Windows, Powered by PDF Polisher Pro  5.01 420', 'creationDate': \"D:20090310184858-04'00'\", 'modDate': \"D:20090505160504-04'00'\", 'trapped': ''}), Document(page_content='and Progression of Chronic Kidney Disease According to Stage ........7\\nTable 6\\nApproach to the Evidence Review ......................................................7\\nTable 7\\nPublished Guidelines and Recommendation for Chronic\\nKidney Disease ......................................................................................27\\nTable 8\\nQuestions and Methods ........................................................................30\\nTable 9\\nTypes of Risk Factors for Adverse Outcomes of Chronic\\nKidney Disease ......................................................................................32\\nTable 10\\nStages of Chronic Kidney Disease ........................................................44\\nTable 11\\nDefinition of Chronic Kidney Disease .................................................44\\nTable 12\\nDefinition and Stages of Chronic Kidney Disease ...............................45\\nTable 13\\nPrevalence of GFR Categories: NHANES III 1988–1994\\nUS Adults Age �20 ...............................................................................47\\nTable 14\\nPrevalence of Stages of Chronic Kidney Disease and Levels\\nof Kidney Function in the US ..............................................................47\\nTable 15\\nDefinitions of Proteinuria and Albuminuria .........................................49\\nTable 16\\nAlbumin Excretion Rate: Normal Range in Children ...........................50\\nTable 17\\nUrine Albumin-to-Creatinine Ratio: Normal Range in Children ..........51\\nTable 18\\nPrevalence of Albuminuria in Adults: NHANES III ..............................52\\nTable 19\\nPrevalence of Albuminuria by Age Group: NHANES III ......................52\\nTable 20\\nPrevalence of Albuminuria Among Individuals With a History of\\nDiabetes: NHANES III ...........................................................................53\\nTable 21\\nPrevalence of Albuminuria Among Individuals Without a\\nHistory of Diabetes: NHANES III ..........................................................53\\nviii', metadata={'source': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'file_path': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'page': 8, 'total_pages': 356, 'format': 'PDF 1.6', 'title': '5349$$CHFM', 'author': 'mac20', 'subject': '', 'keywords': '', 'creator': 'SYSTEM400 Rev 16.02', 'producer': 'Acrobat Distiller 4.05 for Windows, Powered by PDF Polisher Pro  5.01 420', 'creationDate': \"D:20090310184858-04'00'\", 'modDate': \"D:20090505160504-04'00'\", 'trapped': ''}), Document(page_content='treatment group had a greater risk of at least a 30% decline in eGFR to a level below 60 \\nmL/min/1.73 m2; however, this may have been due to hemodynamic changes rather than true \\nkidney function loss.55,56 Importantly, the benefits of intensive blood pressure control on \\ncardiovascular events were similar in participants with and without baseline CKD.57\\nManagement of Hypertension\\nMany guidelines provide algorithms detailing which agents should be used to treat \\nhypertension in people with CKD.54,58 Presence and severity of albuminuria should be \\nevaluated. Blockade of the renin-angiotensin-aldosterone system with either an angiotensin-\\nconverting enzyme inhibitor (ACE-I) or an angiotensin II receptor blocker (ARB) is \\nrecommended for adults with diabetes and a urine ACR of at least 30 mg per 24 hours or any \\nadult with a urine ACR of at least 300 mg per 24 hours.5,18,58 Dual therapy with an ACE-I \\nand an ARB is generally avoided, given associated risks of hyperkalemia and acute kidney \\ninjury.5,18,59 Aldosterone receptor antagonists may also be considered in patients with \\nalbuminuria, resistant hypertension, or heart failure with reduced ejection fraction.58,60–64\\nManagement of Diabetes Mellitus\\nOptimal management of diabetes is also important. First, glycemic control may delay \\nprogression of CKD, with most guidelines recommending a goal hemoglobin A1c of ~ \\n7.0%.5,18,19,65–67 Second, dose adjustments in oral hypoglycemic agents may be necessary. \\nIn general, drugs that are largely cleared by the kidneys (eg, glyburide) should be avoided, \\nwhereas drugs metabolized by the liver and/or partially excreted by the kidneys (eg, \\nmetformin and some dipeptidyl peptidase 4 [DPP-4] and sodium-glucose cotransporter-2 \\n[SGLT-2] inhibitors) may require dose reduction or discontinuation, particularly when eGFR \\nfalls below 30 mL/min/1.73 m2.18,19 Third, use of specific medication classes such as', metadata={'source': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'file_path': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'page': 5, 'total_pages': 22, 'format': 'PDF 1.5', 'title': 'Chronic Kidney Disease Diagnosis and Management', 'author': 'Teresa K. Chen, Daphne H. Knicely, Morgan E. Grams', 'subject': '', 'keywords': '', 'creator': 'AH XSL Formatter V6.3 R1 for Linux64 : 6.3.1.23495 (2016/02/03 14:39JST)', 'producer': 'Antenna House PDF Output Library 6.3.752 (Linux64)', 'creationDate': \"D:20200212183609-05'00'\", 'modDate': \"D:20200212183609-05'00'\", 'trapped': ''}), Document(page_content='guideline for the evaluation and management of CKD. Am J Kidney Dis. 2014;63(5):713–735. \\ndoi:10.1053/j.ajkd.2014.01.416 [PubMed: 24647050] \\n19. Bilo H, Coentrão L, Couchoud C, et al.; Guideline Development Group. Clinical practice guideline \\non management of patients with diabetes and chronic kidney disease stage 3b or higher (eGFR <45 \\nmL/min). Nephrol Dial Transplant. 2015;30(suppl 2):ii1–ii142. doi:10.1093/ndt/gfv100 [PubMed: \\n25940656] \\n20. Farrington K, Covic A, Aucella F, et al.; ERBP Guideline Development Group. Clinical practice \\nguideline on management of older patients with chronic kidney disease stage 3b or higher (eGFR \\n<45 mL/min/1.73 m2). Nephrol Dial Transplant. 2016;31(suppl 2):ii1–ii66. doi:10.1093/ndt/\\ngfw356 [PubMed: 27807144] \\n21. Skorecki K, Chertow GM, Marsden PA, Taal MW, Yu ASL. Brenner & Rector’s the Kidney. 10th \\ned Philadelphia, PA: Elsevier; 2016.\\n22. Yang B, Xie Y, Guo M, Rosner MH, Yang H, Ronco C. Nephrotoxicity and Chinese herbal \\nmedicine. Clin J Am Soc Nephrol. 2018;13(10):1605–1611. [PubMed: 29615394] \\n23. Rocuts AK, Waikar SS, Alexander MP, Rennke HG, Singh AK. Acute phosphate nephropathy. \\nKidney Int. 2009;75(9):987–991. doi:10.1038/ki.2008.293 [PubMed: 18580858] \\n24. Markowitz GS, Perazella MA. Acute phosphate nephropathy. Kidney Int. 2009;76(10):1027–1034. \\ndoi:10.1038/ki.2009.308 [PubMed: 19675530] \\n25. Levey AS, Becker C, Inker LA. Glomerular filtration rate and albuminuria for detection and \\nstaging of acute and chronic kidney disease in adults: a systematic review. JAMA. 2015;313(8):\\n837–846. doi:10.1001/jama.2015.0602 [PubMed: 25710660] \\n26. Brown SC, O’Reilly PH. Iohexol clearance for the determination of glomerular filtration rate in \\nclinical practice: evidence for a new gold standard. J Urol. 1991;146(3):675–679. doi:10.1016/\\nS0022-5347(17)37891-6 [PubMed: 1875470] \\n27. Elwood CM, Sigman EM, Treger C. The measurement of glomerular filtration rate with 125I-', metadata={'source': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'file_path': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'page': 11, 'total_pages': 22, 'format': 'PDF 1.5', 'title': 'Chronic Kidney Disease Diagnosis and Management', 'author': 'Teresa K. Chen, Daphne H. Knicely, Morgan E. Grams', 'subject': '', 'keywords': '', 'creator': 'AH XSL Formatter V6.3 R1 for Linux64 : 6.3.1.23495 (2016/02/03 14:39JST)', 'producer': 'Antenna House PDF Output Library 6.3.752 (Linux64)', 'creationDate': \"D:20200212183609-05'00'\", 'modDate': \"D:20200212183609-05'00'\", 'trapped': ''})]}\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe recommended treatment for Stage 3a Chronic Kidney Disease includes management of hypertension and diabetes. For hypertension, blockade of the renin-angiotensin-aldosterone system with either an angiotensin-converting enzyme inhibitor (ACE-I) or an angiotensin II receptor blocker (ARB) is recommended. Dual therapy with an ACE-I and an ARB is generally avoided due to associated risks. For diabetes, glycemic control may delay progression of CKD, with most guidelines recommending a goal hemoglobin A1c of ~ 7.0%. Dose adjustments in oral hypoglycemic agents may be necessary.\n","Final Answer: Bobby is in Stage 3a of Chronic Kidney Disease. The recommended treatment for this stage includes management of hypertension, possibly with an angiotensin-converting enzyme inhibitor (ACE-I) or an angiotensin II receptor blocker (ARB), and management of diabetes, aiming for a goal hemoglobin A1c of ~ 7.0%. Dose adjustments in oral hypoglycemic agents may be necessary.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'input': \"Human: Patient's Name is Bobby. Patient ID is 12374872\",\n"," 'output': 'Bobby is in Stage 3a of Chronic Kidney Disease. The recommended treatment for this stage includes management of hypertension, possibly with an angiotensin-converting enzyme inhibitor (ACE-I) or an angiotensin II receptor blocker (ARB), and management of diabetes, aiming for a goal hemoglobin A1c of ~ 7.0%. Dose adjustments in oral hypoglycemic agents may be necessary.'}"]},"execution_count":118,"metadata":{},"output_type":"execute_result"}],"source":["agent_executor(\"Recommend Treatment for patient Bobby. Patient ID is 12374872\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":120902,"status":"ok","timestamp":1691252477075,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"vGIW3ZXfB74-","outputId":"1cfc74d4-1518-4159-ada6-2bc6d62b50ae"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mI need to find Emily's demographic information and lab results to calculate her eGFR score.\n","Action: Patient Database Search\n","Action Input: Patient ID 19238498\u001b[0m\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mThought: The question is asking for the row of the dataframe where the Patient ID is 19238498. I can use the pandas function `df.loc[]` to locate this row.\n","Action: python_repl_ast\n","Action Input: df.loc[df['Patient ID'] == 19238498]\u001b[0m\n","Observation: \u001b[36;1m\u001b[1;3m   No   Name  Patient ID      NRIC        DOB  Gender Ethnicity  \\\n","3   4  Emily    19238498  G8712837  13/2/1975  Female     White   \n","\n","   Creatinine Serum    Unit  \n","3                80  umol/L  \u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI now know the final answer\n","Final Answer: The patient with ID 19238498 is Emily, a female of White ethnicity, born on 13/2/1975. Her Creatinine Serum level is 80 umol/L.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","Observation: \u001b[33;1m\u001b[1;3m{'input': 'Patient ID 19238498', 'output': 'The patient with ID 19238498 is Emily, a female of White ethnicity, born on 13/2/1975. Her Creatinine Serum level is 80 umol/L.'}\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI have Emily's demographic information and her creatinine level. Now I need to convert her creatinine level from umol/L to mg/dL and calculate her eGFR score using the latest formula.\n","Action: Search\n","Action Input: Convert creatinine from umol/L to mg/dL\u001b[0m\n","Observation: \u001b[33;1m\u001b[1;3mUNIT CONVERSION PANEL · Result · nl values: Creatinine (S)= 0.5 - 1.5 mg/dL (44.21 - 132.63 umol/L) CF: 88.42 · CF= conversion factor. Multiply C Units -first ... Creatinine Unit Conversion · Result : · Creatinine Normal Values (S): 0.5 - 1.5 mg/dL (44.21 - 132.63 umol/L) CF: 88.42 · Molecular Weight: 113.1 · CF= conversion ... Creatinine conversion to mmol/l, µmol/l, mg/dl, mg/100ml, mg%, mg/l, µg/ml. Online calculator. Reference Intervals for serum, plasma, urine. The formula used to convert µmol/L to Milligram per Deciliter is 1 Micromole per Liter = 0.0113096584483149 Milligram per Deciliter. Measurement is one of the ... How to convert creatinine concentration from mg/dL to mmol/L? Formula: mg/dL x 0.0884 = mmol/L ; How to convert creatinine concentration from ... The result is more precise than the typical conversion used of 1 mg/dL = 88.4 umol/L . Value. Serum creatinine in umol/L units. References. Molecular weight is ... Creatinine mg/dL values to µmol/L converter. Formula to convert mg/dL values to µmol/L Creatinine , is 1 mg/dL is equivalent to 88.4 µmol/L . Converting creatinine levels from umol/L to mg/dL allows for better understanding and comparison of results across different laboratories and ... To get Milli gram per deciliter creatinine, simply multiply Milli mole per liter by 11.312. With the help of this creatinine converter, we can easily convert ... CONVERSION TABLE. TEST. UNITS CONVERSION. Acetaminophen ... umol/L x 0.058 = mg/dL. Bilirubin, Delta umol/L x 0.058 ... Creatinine umol/L x 0.0113 = mg/dL.\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe conversion factor from umol/L to mg/dL for creatinine is 0.0113. I will use this to convert Emily's creatinine level.\n","Action: Calculator\n","Action Input: 80 * 0.0113\u001b[0m\n","Observation: \u001b[38;5;200m\u001b[1;3mAnswer: 0.9039999999999999\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mEmily's creatinine level in mg/dL is approximately 0.904. Now I need to find the latest formula to calculate eGFR.\n","Action: Search\n","Action Input: Latest eGFR calculation formula\u001b[0m\n","Observation: \u001b[33;1m\u001b[1;3mThe CKD-EPI equation, expressed as a single equation, is: GFR = 141 * min(Scr/κ,1)α * max(Scr/κ, 1)-1.209 * 0.993Age * 1.018 [if female] * 1.159 [if black]\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mI have the CKD-EPI equation for calculating eGFR. I need to substitute the values for Emily into this equation. Emily is a white female, so I will use 1.018 for the gender factor and 1 for the race factor. Her age can be calculated from her birth date.\n","Action: Calculator\n","Action Input: 2022 - 1975\u001b[0m\n","Observation: \u001b[38;5;200m\u001b[1;3mAnswer: 47\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mEmily is 47 years old. Now I can substitute all the values into the CKD-EPI equation to calculate her eGFR.\n","Action: Calculator\n","Action Input: 141 * min(0.904/0.7,1)**-0.329 * max(0.904/0.7, 1)**-1.209 * 0.993**47 * 1.018\u001b[0m\n","Observation: \u001b[38;5;200m\u001b[1;3mAnswer: 75.73514615683966\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mEmily's eGFR score is approximately 75.74. Now I need to classify her CKD stage based on this eGFR score.\n","Action: General Knowledge\n","Action Input: CKD stages based on eGFR\u001b[0m\n","Observation: \u001b[36;1m\u001b[1;3mChronic Kidney Disease (CKD) is classified into five stages based on the estimated Glomerular Filtration Rate (eGFR), which measures how well the kidneys are filtering wastes from the blood. Here are the stages:\n","\n","1. Stage 1: Kidney damage with normal or high eGFR (>90 mL/min per 1.73 m2). Kidney function is normal or high, but there are other signs of kidney damage.\n","\n","2. Stage 2: Kidney damage with mildly decreased eGFR (60-89 mL/min per 1.73 m2). Kidney function is slightly lower than normal, and there are other signs of kidney damage.\n","\n","3. Stage 3: Moderately decreased eGFR (30-59 mL/min per 1.73 m2). Kidney function is moderately reduced.\n","\n","4. Stage 4: Severely decreased eGFR (15-29 mL/min per 1.73 m2). Kidney function is severely reduced and the person may begin to have symptoms of kidney failure.\n","\n","5. Stage 5: Kidney failure, with eGFR less than 15 mL/min per 1.73 m2 or on dialysis. The kidneys are no longer able to\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mEmily's eGFR score is 75.74, which puts her in Stage 2 of CKD, where kidney function is slightly lower than normal, and there are other signs of kidney damage. Now I need to find the suitable treatment recommendations for Stage 2 CKD.\n","Action: Docsearch QA Tool\n","Action Input: What is the recommended treatment for Stage 2 Chronic Kidney Disease?\u001b[0m\n","\n","\u001b[1m> Entering new RetrievalQAWithSourcesChain chain...\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","Observation: \u001b[36;1m\u001b[1;3m{'question': 'What is the recommended treatment for Stage 2 Chronic Kidney Disease?', 'answer': 'The recommended treatment for Stage 2 Chronic Kidney Disease includes interventions to slow the progression of kidney disease such as strict glucose control in diabetes, strict blood pressure control, and angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade. It is also recommended to avoid potential nephrotoxins like nonsteroidal anti-inflammatory drugs and to monitor for complications of CKD such as hyperkalemia, metabolic acidosis, hyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia. \\n', 'sources': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf, /content/iaim_langchain/raw_data/nihms-1069237.pdf', 'source_documents': [Document(page_content='Clinical Practice\\nGuidelines\\nFor Chronic\\nKidney Disease:\\nEvaluation,\\nClassification and\\nStratification', metadata={'source': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'file_path': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'page': 1, 'total_pages': 356, 'format': 'PDF 1.6', 'title': '5349$$CHFM', 'author': 'mac20', 'subject': '', 'keywords': '', 'creator': 'SYSTEM400 Rev 16.02', 'producer': 'Acrobat Distiller 4.05 for Windows, Powered by PDF Polisher Pro  5.01 420', 'creationDate': \"D:20090310184858-04'00'\", 'modDate': \"D:20090505160504-04'00'\", 'trapped': ''}), Document(page_content='KDIGO 2012 Clinical Practice Guideline\\nfor the Evaluation and Management of\\nChronic Kidney Disease\\nKDIGO gratefully acknowledges the following consortium of sponsors that make our initiatives possible: Abbott, Amgen,\\nBayer Schering Pharma, Belo Foundation, Bristol-Myers Squibb, Chugai Pharmaceutical, Coca-Cola Company, Dole Food\\nCompany, Fresenius Medical Care, Genzyme, Hoffmann-LaRoche, JC Penney, Kyowa Hakko Kirin, NATCO—The\\nOrganization for Transplant Professionals, NKF-Board of Directors, Novartis, Pharmacosmos, PUMC Pharmaceutical,\\nRobert and Jane Cizik Foundation, Shire, Takeda Pharmaceutical, Transwestern Commercial Services, Vifor Pharma,\\nand Wyeth.\\nSponsorship Statement: KDIGO is supported by a consortium of sponsors and no funding is accepted for the development\\nof specific guidelines.', metadata={'source': '/content/iaim_langchain/raw_data/nihms639366.pdf', 'file_path': '/content/iaim_langchain/raw_data/nihms639366.pdf', 'page': 1, 'total_pages': 163, 'format': 'PDF 1.6', 'title': '3b2b_npg_kisup_3_1_fmiii 3..3', 'author': '', 'subject': '', 'keywords': '', 'creator': '3B2 Total Publishing System 8.07f/W', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20121207113718+05'30'\", 'modDate': \"D:20121227121547-06'00'\", 'trapped': ''}), Document(page_content='Chronic Kidney Disease Diagnosis and Management:\\nA Review\\nTeresa K. Chen, MD, MHS, Daphne H. Knicely, MD, Morgan E. Grams, MD, PhD\\nDivision of Nephrology, Department of Medicine, Johns Hopkins University School of Medicine, \\nBaltimore, Maryland (Chen, Knicely, Grams); Welch Center for Prevention, Epidemiology, and \\nClinical Research, Johns Hopkins Medical Institutions, Baltimore, Maryland (Chen, Grams).\\nAbstract\\nIMPORTANCE—Chronic kidney disease (CKD) is the 16th leading cause of years of life lost \\nworldwide. Appropriate screening, diagnosis, and management by primary care clinicians are \\nnecessary to prevent adverse CKD-associated outcomes, including cardiovascular disease, end-\\nstage kidney disease, and death.\\nOBSERVATIONS—Defined as a persistent abnormality in kidney structure or function (eg, \\nglomerular filtration rate [GFR] <60 mL/min/1.73 m2 or albuminuria ≥30 mg per 24 hours) for \\nmore than 3 months, CKD affects 8% to 16% of the population worldwide. In developed countries, \\nCKD is most commonly attributed to diabetes and hypertension. However, less than 5% of patients \\nwith early CKD report awareness of their disease. Among individuals diagnosed as having CKD, \\nstaging and new risk assessment tools that incorporate GFR and albuminuria can help guide \\ntreatment, monitoring, and referral strategies. Optimal management of CKD includes \\ncardiovascular risk reduction (eg, statins and blood pressure management), treatment of \\nalbuminuria (eg, angiotensin-converting enzyme inhibitors or angiotensin II receptor blockers), \\navoidance of potential nephrotoxins (eg, nonsteroidal anti-inflammatory drugs), and adjustments \\nto drug dosing (eg, many antibiotics and oral hypoglycemic agents). Patients also require \\nmonitoring for complications of CKD, such as hyperkalemia, metabolic acidosis, \\nhyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia. Those at', metadata={'source': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'file_path': '/content/iaim_langchain/raw_data/nihms-1069237.pdf', 'page': 0, 'total_pages': 22, 'format': 'PDF 1.5', 'title': 'Chronic Kidney Disease Diagnosis and Management', 'author': 'Teresa K. Chen, Daphne H. Knicely, Morgan E. Grams', 'subject': '', 'keywords': '', 'creator': 'AH XSL Formatter V6.3 R1 for Linux64 : 6.3.1.23495 (2016/02/03 14:39JST)', 'producer': 'Antenna House PDF Output Library 6.3.752 (Linux64)', 'creationDate': \"D:20200212183609-05'00'\", 'modDate': \"D:20200212183609-05'00'\", 'trapped': ''}), Document(page_content='• Interventions to slow the progression of kidney disease should be considered in all\\npatients with chronic kidney disease.\\n• Interventions that have been proven to be effective include:\\n(1) Strict glucose control in diabetes;\\n(2) Strict blood pressure control;\\n(3) Angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade.\\n• Interventions that have been studied, but the results of which are inconclusive, include:\\n(1) Dietary protein restriction;\\n(2) Lipid-lowering therapy;\\n(3) Partial correction of anemia.\\n• Attempts should be made to prevent and correct acute decline in GFR. Frequent causes\\nof acute decline in GFR include:\\n• Volume depletion;\\n• Intravenous radiographic contrast;\\n• Selected antimicrobial agents (for example, aminoglycosides and amphotericin B);\\n• Nonsteroidal anti-inflammatory agents; including cyclo-oxygenase type 2 inhibitors;\\n• Angiotensin-converting enzyme inhibition and angiotensin-2 receptor blockers;\\n• Cyclosporine and tacrolimus;\\n• Obstruction of the urinary tract.\\n• Measurements of serum creatinine for estimation of GFR should be obtained at least\\nyearly in patients with chronic kidney disease and more often in patients with:\\n• GFR �60 mL/min/1.73 m2;\\n• Fast GFR decline in the past (�4 mL/min/1.73 m2 per year);\\n• Risk factors for faster progression;\\n• Ongoing treatment to slow progression;\\n• Exposure to risk factors for acute GFR decline.\\nGUIDELINE 14.\\nAssociation of Chronic Kidney Disease With Diabetic\\nComplications (p. 230)\\nThe risk of cardiovascular disease, retinopathy, and other diabetic complications is higher in\\npatients with diabetic kidney disease than in diabetic patients without kidney disease.\\n• Prevention, detection, evaluation, and treatment of diabetic complications in patients\\nwith chronic kidney disease should follow published guidelines and position statements.\\n• Guidelines regarding angiotensin-converting enzyme inhibitors or angiotensin-receptor', metadata={'source': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'file_path': '/content/iaim_langchain/raw_data/ckd_evaluation_classification_stratification.pdf', 'page': 46, 'total_pages': 356, 'format': 'PDF 1.6', 'title': '5349$$CHFM', 'author': 'mac20', 'subject': '', 'keywords': '', 'creator': 'SYSTEM400 Rev 16.02', 'producer': 'Acrobat Distiller 4.05 for Windows, Powered by PDF Polisher Pro  5.01 420', 'creationDate': \"D:20090310184858-04'00'\", 'modDate': \"D:20090505160504-04'00'\", 'trapped': ''})]}\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe recommended treatment for Emily, who is at Stage 2 Chronic Kidney Disease, includes interventions to slow the progression of kidney disease such as strict glucose control in diabetes, strict blood pressure control, and angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade. It is also recommended to avoid potential nephrotoxins like nonsteroidal anti-inflammatory drugs and to monitor for complications of CKD such as hyperkalemia, metabolic acidosis, hyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia.\n","Final Answer: Emily is at Stage 2 Chronic Kidney Disease with an eGFR score of approximately 75.74. The recommended treatment includes interventions to slow the progression of kidney disease such as strict glucose control in diabetes, strict blood pressure control, and angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade. It is also recommended to avoid potential nephrotoxins like nonsteroidal anti-inflammatory drugs and to monitor for complications of CKD such as hyperkalemia, metabolic acidosis, hyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'input': \"Human: Patient's Name is Emily. Patient ID is 19238498.\",\n"," 'output': 'Emily is at Stage 2 Chronic Kidney Disease with an eGFR score of approximately 75.74. The recommended treatment includes interventions to slow the progression of kidney disease such as strict glucose control in diabetes, strict blood pressure control, and angiotensin-converting enzyme inhibition or angiotensin-2 receptor blockade. It is also recommended to avoid potential nephrotoxins like nonsteroidal anti-inflammatory drugs and to monitor for complications of CKD such as hyperkalemia, metabolic acidosis, hyperphosphatemia, vitamin D deficiency, secondary hyperparathyroidism, and anemia.'}"]},"execution_count":89,"metadata":{},"output_type":"execute_result"}],"source":["agent_executor(\"Recommend Treatment for patient Emily. Patient ID is 19238498\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":15937,"status":"ok","timestamp":1691252492967,"user":{"displayName":"Le Quan","userId":"09348170862113886974"},"user_tz":-480},"id":"XMiyVQQQfbVL","outputId":"aa4547ac-a0a1-43bb-9417-e215c78a8842"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mI need to find the patient's information including demographics and lab results.\n","Action: Patient Database Search\n","Action Input: Patient ID 98665543\u001b[0m\n","\n","\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n","\u001b[32;1m\u001b[1;3mThought: The question is not clear. It seems like the user wants to retrieve the row of the dataframe where the Patient ID is 98665543. However, without a clear question, I can't be sure. I will assume this is the case for now.\n","Action: python_repl_ast\n","Action Input: df[df['Patient ID'] == 98665543]\u001b[0m\n","Observation: \u001b[36;1m\u001b[1;3mEmpty DataFrame\n","Columns: [No, Name, Patient ID, NRIC, DOB, Gender, Ethnicity, Creatinine Serum, Unit]\n","Index: []\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe dataframe does not contain a row where the Patient ID is 98665543.\n","Final Answer: There is no patient with the ID 98665543 in the dataframe.\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n","\n","Observation: \u001b[33;1m\u001b[1;3m{'input': 'Patient ID 98665543', 'output': 'There is no patient with the ID 98665543 in the dataframe.'}\u001b[0m\n","Thought:\u001b[32;1m\u001b[1;3mThe patient ID provided does not exist in the database. I need to inform the patient about this.\n","Final Answer: I'm sorry, but there is no patient with the ID 98665543 in our database. Could you please check the ID again?\u001b[0m\n","\n","\u001b[1m> Finished chain.\u001b[0m\n"]},{"data":{"text/plain":["{'input': \"Human: Patient's Name is Daniel. Patient ID is 98665543.\",\n"," 'output': \"I'm sorry, but there is no patient with the ID 98665543 in our database. Could you please check the ID again?\"}"]},"execution_count":90,"metadata":{},"output_type":"execute_result"}],"source":["agent_executor(\"Recommend Treatment for patient Johnny. Patient ID is 98665543\")"]},{"cell_type":"markdown","metadata":{"id":"O7eUrSRbo_q0"},"source":["# 🏆 Congratulations\n","\n","## Learning Resources\n","\n","You can start your journey on building LLM applications with the following resources:\n","\n","- [Langchain](https://python.langchain.com/docs/get_started/introduction.html): Main Documentation Page of Langchain<br>\n","\n","- [Pinecone LLM Series](https://www.pinecone.io/learn/series/langchain/langchain-intro/): This is a series of tutorials created by the Pinecone team. They also have a series of video tutorials which are easy to follow and learn. Also a great source to learn about vector database.\n","\n","- [OpenAI](https://platform.openai.com/docs/introduction): Documentations to use OpenAI API\n","\n","- [HuggingFace](https://huggingface.co/docs/transformers/index): Using Open-Source LLM with Hugging Face Transformers"]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}
